{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 2025.05.21 전처리"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from pathlib import Path\n",
    "from tqdm import tqdm\n",
    "\n",
    "data = pd.read_csv('data/AF_data_info.csv')\n",
    "data.head(1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Read PDB File"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_pdb(pdb_path):\n",
    "    records = []\n",
    "    with open(pdb_path, 'r') as f:\n",
    "        for line in f:\n",
    "            if line.startswith(('ATOM', 'HETATM')):\n",
    "                record = {\n",
    "                    'record': line[0:6].strip(),\n",
    "                    'atom_serial': int(line[6:11]),\n",
    "                    'atom_name': line[12:16].strip(),\n",
    "                    'alt_loc': line[16],\n",
    "                    'res_name': line[17:20].strip(),\n",
    "                    'chain_id': line[21],\n",
    "                    'res_seq': int(line[22:26]),\n",
    "                    'i_code': line[26],\n",
    "                    'x': float(line[30:38]),\n",
    "                    'y': float(line[38:46]),\n",
    "                    'z': float(line[46:54]),\n",
    "                    'occupancy': float(line[54:60]),\n",
    "                    'b_factor': float(line[60:66]),\n",
    "                    'element': line[76:78].strip(),\n",
    "                    'charge': line[78:80].strip()\n",
    "                }\n",
    "                records.append(record)\n",
    "    records = pd.DataFrame(records)\n",
    "    records = records[records['element'] != 'H']\n",
    "    return records"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Read Mol2 to PDB format"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_mol2(mol2_path, ligand=False):\n",
    "    with open(mol2_path) as f:\n",
    "        lines = f.readlines()\n",
    "\n",
    "    if ligand:\n",
    "        atom_start = next(i for i, l in enumerate(lines) if l.startswith('@<TRIPOS>ATOM'))\n",
    "        bond_start = next(i for i, l in enumerate(lines) if l.startswith('@<TRIPOS>BOND'))\n",
    "        atom_end = bond_start\n",
    "        atom_lines = lines[atom_start+1:atom_end]\n",
    "        return ligand_reader(atom_lines)\n",
    "    \n",
    "    else:\n",
    "        atom_start = next(i for i, l in enumerate(lines) if l.startswith('@<TRIPOS>ATOM'))\n",
    "        bond_start = next(i for i, l in enumerate(lines) if l.startswith('@<TRIPOS>BOND'))\n",
    "        chain_start = next(i for i, l in enumerate(lines) if l.startswith('@<TRIPOS>SUBSTRUCTURE'))\n",
    "        set_start = next(i for i, l in enumerate(lines) if l.startswith('@<TRIPOS>SET'))\n",
    "\n",
    "        atom_end = bond_start\n",
    "        bond_end = chain_start\n",
    "        chain_end = set_start\n",
    "\n",
    "        atom_lines = lines[atom_start+1:atom_end]\n",
    "        bond_lines = lines[bond_start+1:bond_end]\n",
    "        chain_lines = lines[chain_start+1:chain_end]\n",
    "        return to_pdb_format(atom_lines, bond_lines, chain_lines)\n",
    "\n",
    "\n",
    "def ligand_reader(atom):\n",
    "    records = []\n",
    "    for idx, line in enumerate(atom):\n",
    "        fields = line.split()\n",
    "        atom_id = int(fields[0])\n",
    "        atom_name = fields[1]\n",
    "        x, y, z = map(float, fields[2:5])\n",
    "        atom_type = fields[5]\n",
    "        resid = int(fields[6])\n",
    "        resname_full = fields[7]\n",
    "        res_name = resname_full[:3]\n",
    "        occupancy = 1.0\n",
    "        b_factor = 0.0\n",
    "        element = atom_type.split('.')[0][0].upper()\n",
    "        charge = fields[8] if len(fields) > 8 and fields[8].replace('.', '', 1).replace('-', '', 1).isdigit() else ''\n",
    "        record = {\n",
    "            'record': 'ATOM',\n",
    "            'atom_serial': atom_id,\n",
    "            'atom_name': atom_name,\n",
    "            'alt_loc': '',\n",
    "            'res_name': res_name,\n",
    "            'chain_id': '',\n",
    "            'res_seq': resid,\n",
    "            'i_code': '',\n",
    "            'x': x,\n",
    "            'y': y,\n",
    "            'z': z,\n",
    "            'atom_type': atom_type,\n",
    "            'occupancy': occupancy,\n",
    "            'b_factor': b_factor,\n",
    "            'element': element,\n",
    "            'charge': charge\n",
    "        }\n",
    "        records.append(record)\n",
    "    records = pd.DataFrame(records)\n",
    "    records = records[records['element'] != 'H']       \n",
    "    return records.reset_index(drop=True)\n",
    "\n",
    "\n",
    "def parse_substructure(lines):\n",
    "    substructures = []\n",
    "    for line in lines:\n",
    "        parts = line.strip().split()\n",
    "        sub = {\n",
    "            'id': int(parts[0]),\n",
    "            'name': parts[1],\n",
    "            'root_atom': int(parts[2]),\n",
    "            'type': parts[3],\n",
    "            'dict_type': parts[4],\n",
    "            'chain': parts[5],\n",
    "            'residue': parts[6],\n",
    "            'unk': parts[7],\n",
    "            'status': parts[8],\n",
    "        }\n",
    "        substructures.append(sub)\n",
    "    return pd.DataFrame(substructures)\n",
    "\n",
    "\n",
    "def to_pdb_format(atom, bond, chain):\n",
    "    records = []\n",
    "    chains = parse_substructure(chain)\n",
    "    \n",
    "    box = {}\n",
    "    prev = None\n",
    "    root_atom_chain = None\n",
    "    chain_atoms = [1]\n",
    "    for idx, line in enumerate(atom):\n",
    "        fields = line.split()\n",
    "        atom_id = int(fields[0])\n",
    "        atom_name = fields[1]\n",
    "        x, y, z = map(float, fields[2:5])\n",
    "        atom_type = fields[5]\n",
    "        resid = int(fields[6])\n",
    "        \n",
    "        resname_full = fields[7]\n",
    "        res_name = resname_full[:3]\n",
    "        \n",
    "        if idx == 0:\n",
    "            prev = resid\n",
    "        else:\n",
    "            if resid == prev:\n",
    "                chain_atoms.append(atom_id)\n",
    "            else:\n",
    "                for a in chain_atoms:\n",
    "                    box[a] = root_atom_chain\n",
    "                prev = resid\n",
    "                chain_atoms = [atom_id]\n",
    "\n",
    "        if atom_id in chains['root_atom'].values:\n",
    "            chain_id = chains[chains['root_atom'] == atom_id]['chain'].values[0]\n",
    "            root_atom_chain = chain_id\n",
    "        else:\n",
    "            chain_id = None\n",
    "            \n",
    "        occupancy = 1.0\n",
    "        b_factor = 0.0\n",
    "        element = atom_type.split('.')[0][0].upper()\n",
    "        charge = fields[8] if len(fields) > 8 and fields[8].replace('.', '', 1).replace('-', '', 1).isdigit() else ''\n",
    "        record = {\n",
    "            'record': 'ATOM',\n",
    "            'atom_serial': atom_id,\n",
    "            'atom_name': atom_name,\n",
    "            'alt_loc': '',\n",
    "            'res_name': res_name,\n",
    "            'chain_id': chain_id,\n",
    "            'res_seq': resid,\n",
    "            'i_code': '',\n",
    "            'x': x,\n",
    "            'y': y,\n",
    "            'z': z,\n",
    "            'atom_type': atom_type,\n",
    "            'occupancy': occupancy,\n",
    "            'b_factor': b_factor,\n",
    "            'element': element,\n",
    "            'charge': charge\n",
    "        }\n",
    "        records.append(record)\n",
    "    \n",
    "    records = pd.DataFrame(records)\n",
    "    records = records[records['element'] != 'H']       \n",
    "    records['chain_id'] = records['atom_serial'].map(box)\n",
    "    return records.reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Parsing Pocket"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "amino_acids = {\n",
    "    'ALA': 'A', 'ARG': 'R', 'ASN': 'N', 'ASP': 'D',\n",
    "    'CYS': 'C', 'GLU': 'E', 'GLN': 'Q', 'GLY': 'G',\n",
    "    'HIS': 'H', 'ILE': 'I', 'LEU': 'L', 'LYS': 'K',\n",
    "    'MET': 'M', 'PHE': 'F', 'PRO': 'P', 'SER': 'S',\n",
    "    'THR': 'T', 'TRP': 'W', 'TYR': 'Y', 'VAL': 'V'\n",
    "    }\n",
    "\n",
    "rev_amino_acids = {v: k for k, v in amino_acids.items()}\n",
    "\n",
    "modified_residue_map = {\n",
    "    \"MSE\": \"M\",  # Selenomethionine → Methionine\n",
    "    \"SEP\": \"S\",  # Phosphoserine → Serine\n",
    "    \"TPO\": \"T\",  # Phosphothreonine → Threonine\n",
    "    \"PTR\": \"Y\",  # Phosphotyrosine → Tyrosine\n",
    "    \"HYP\": \"P\",  # Hydroxyproline → Proline\n",
    "    \"KCX\": \"K\",  # Carboxylysine → Lysine\n",
    "    \"CSO\": \"C\",  # Oxidized cysteine → Cysteine\n",
    "    \"CGU\": \"E\",  # γ-carboxy-glutamate → Glutamic Acid\n",
    "    \"F2F\": \"F\",  # Fluorophenylalanine → Phenylalanine\n",
    "    \"ASH\": \"D\",  # Protonated Aspartic Acid → Aspartic Acid\n",
    "    \"GLH\": \"E\",  # Protonated Glutamic Acid → Glutamic Acid\n",
    "    \"CYX\": \"C\",  # Disulfide-bonded Cysteine → Cysteine\n",
    "    \"HID\": \"H\",  # Neutral Histidine (delta-protonated)\n",
    "    \"HIE\": \"H\",  # Neutral Histidine (epsilon-protonated)\n",
    "    \"HIP\": \"H\",  # Positively charged Histidine\n",
    "}\n",
    "def parse_residue_and_align(lst, chain=True):\n",
    "    result = []\n",
    "    prev = None\n",
    "    seen_counts = {}\n",
    "    \n",
    "    if chain:\n",
    "        \n",
    "        for num, chain in lst:\n",
    "            if num == prev:\n",
    "                continue\n",
    "            count = seen_counts.get(num, 0) + 1\n",
    "            seen_counts[num] = count\n",
    "            key = (f\"{num}\" if count == 1 else f\"{num}-{count}\", chain)\n",
    "            result.append(key)\n",
    "            prev = num\n",
    "        result = {k: v for k, v in result}\n",
    "    \n",
    "    else:    \n",
    "        for num in lst:\n",
    "            if num == prev:\n",
    "                result.append(key)\n",
    "                continue \n",
    "            count = seen_counts.get(num, 0) + 1\n",
    "            seen_counts[num] = count\n",
    "            key = f\"{num}\" if count == 1 else f\"{num}-{count}\"\n",
    "            result.append(key)\n",
    "            prev = num\n",
    "            \n",
    "    return result\n",
    "\n",
    "\n",
    "def read_residue(res):\n",
    "    global amino_acids\n",
    "\n",
    "    init = 0\n",
    "    result = []\n",
    "    res_dict = {}\n",
    "    already = set()\n",
    "    for r, c, idx, ic in res:\n",
    "        if r == 'HOH':\n",
    "            continue\n",
    "        if r not in amino_acids:\n",
    "            conv_r = 'X'\n",
    "        else:\n",
    "            conv_r = amino_acids[r]\n",
    "        key = (c, idx, ic)\n",
    "        if key in already:\n",
    "            continue\n",
    "        already.add(key)\n",
    "        result.append(conv_r)\n",
    "        res_dict[key] = init\n",
    "        init += 1\n",
    "    rev_res_dict = {v: k for k, v in res_dict.items()}\n",
    "    return ''.join(result), res_dict, rev_res_dict"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## PyMOL Modules"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "import os\n",
    "import pymol\n",
    "from pymol import cmd\n",
    "from pathlib import Path\n",
    "from tqdm import tqdm\n",
    "\n",
    "# pymol.finish_launching() # if you see the work in pymol, you need to run this\n",
    "\n",
    "\n",
    "def align_and_save(apo_path, holo_path, output_path):\n",
    "    cmd.reinitialize()  # Clear previous structures\n",
    "    cmd.load(str(apo_path), 'apo')\n",
    "    cmd.load(str(holo_path), 'holo')\n",
    "    alignment_info = cmd.super('apo', 'holo')\n",
    "    cmd.save(str(output_path), 'apo')\n",
    "    rmsd = alignment_info[0]  # RMSD\n",
    "    return rmsd\n",
    "\n",
    "\n",
    "def point_mutation(apo_path, output_path, mutations):\n",
    "    global rev_amino_acids\n",
    "    cmd.reinitialize()  # Clear previous structures\n",
    "    cmd.load(apo_path, 'apo')\n",
    "    cmd.wizard('mutagenesis')\n",
    "    for row in mutations:\n",
    "        chain, resid, ap_res, ho_res = map(str, row)\n",
    "        # PyMOL residue selection: /object//chain/resi\n",
    "        if ho_res == 'X':\n",
    "            continue\n",
    "        selection = f'/apo//{chain}/{resid}/'\n",
    "        print(f'selcction: {selection}')\n",
    "        print(f'apo: {ap_res} -> holo: {ho_res}')\n",
    "        cmd.get_wizard().set_mode(rev_amino_acids[ho_res])\n",
    "        cmd.get_wizard().do_select(selection)\n",
    "        cmd.get_wizard().apply()\n",
    "    cmd.set_wizard()\n",
    "    cmd.save(output_path, 'apo')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pathlib import Path\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from scipy.spatial.distance import cdist\n",
    "\n",
    "def find_pocket_residues(complex_df: pd.DataFrame, ligand_df: pd.DataFrame, cutoff: float = 10.0):\n",
    "    ligand_df = ligand_df.reset_index(drop=True)\n",
    "    complex_df = complex_df.reset_index(drop=True)\n",
    "    ligand_coords = ligand_df[['x', 'y', 'z']].to_numpy()\n",
    "    protein_df = complex_df[~complex_df['res_name'].isin(ligand_df['res_name'].unique())].copy()\n",
    "    protein_df = protein_df.reset_index(drop=True)\n",
    "    protein_coords = protein_df[['x', 'y', 'z']].to_numpy()\n",
    "    \n",
    "    # calculate distance between ligand and protein\n",
    "    dist_matrix = cdist(ligand_coords, protein_coords)\n",
    "    \n",
    "    # extract indices of atoms within cutoff\n",
    "    close_idx = np.where(dist_matrix <= cutoff)[1]\n",
    "    close_atoms = protein_df.iloc[close_idx]\n",
    "\n",
    "    # extract pocket residues\n",
    "    pocket_residues = {\n",
    "        (row['chain_id'], row['res_seq'], row['i_code'])\n",
    "        for _, row in close_atoms.iterrows()\n",
    "    }\n",
    "    pocket_df = find_atom(complex_df, pocket_residues)\n",
    "    return pocket_df\n",
    "\n",
    "\n",
    "def write_pocket_to_mol2(pocket_df: pd.DataFrame, output_path: str):\n",
    "    pocket_df = pocket_df.reset_index(drop=True)\n",
    "    pocket_df['atom_id'] = np.arange(1, len(pocket_df) + 1)\n",
    "    pocket_df[['x', 'y', 'z', 'charge']] = pocket_df[['x', 'y', 'z', 'charge']].astype(float)\n",
    "\n",
    "    with open(output_path, 'w') as f:\n",
    "        # MOLECULE section\n",
    "        f.write(\"@<TRIPOS>MOLECULE\\n\")\n",
    "        f.write(\"POCKET\\n\")\n",
    "        f.write(f\"{len(pocket_df)} 0 1\\n\")\n",
    "        f.write(\"PROTEIN\\n\")\n",
    "        f.write(\"USER_CHARGES\\n\\n\")\n",
    "\n",
    "        # ATOM section\n",
    "        f.write(\"@<TRIPOS>ATOM\\n\")\n",
    "        for _, row in pocket_df.iterrows():\n",
    "            f.write(\n",
    "                f\"{row['atom_id']:>7} {row['atom_name']:<8} \"\n",
    "                f\"{row['x']:>10.4f} {row['y']:>10.4f} {row['z']:>10.4f} \"\n",
    "                f\"{row['atom_type']:<6} {row['res_seq']:>5} {row['res_name']:<8} \"\n",
    "                f\"{row.get('charge', 0.0):>9.4f}\\n\"\n",
    "            )\n",
    "\n",
    "        # BOND section (empty)\n",
    "        f.write(\"@<TRIPOS>BOND\\n\")\n",
    "\n",
    "\n",
    "def read_csar(mol2_path):\n",
    "    mol2_path = Path(mol2_path)\n",
    "    output_path = mol2_path.parent / f'{mol2_path.parent.stem}_ligand.mol2'\n",
    "\n",
    "    with open(mol2_path) as f:\n",
    "        lines = f.readlines()\n",
    "\n",
    "    atom_start = next(i for i, l in enumerate(lines) if l.startswith('@<TRIPOS>ATOM'))\n",
    "    bond_start = next(i for i, l in enumerate(lines) if l.startswith('@<TRIPOS>BOND'))\n",
    "    chain_start = next(i for i, l in enumerate(lines) if l.startswith('@<TRIPOS>SUBSTRUCTURE'))\n",
    "    set_start = next((i for i, l in enumerate(lines) if l.startswith('@<TRIPOS>SET')), len(lines))\n",
    "\n",
    "    atom_lines = lines[atom_start+1:bond_start]\n",
    "    bond_lines = lines[bond_start+1:chain_start]\n",
    "    chain_lines = lines[chain_start+1:set_start]\n",
    "\n",
    "    # --- parse ligand definition ---\n",
    "    ligand_line_idx = next(i for i, l in enumerate(lines) if l.strip().startswith(\"LIGAND\"))\n",
    "\n",
    "    ligand_atom_ids = []\n",
    "    i = ligand_line_idx + 1\n",
    "    while i < len(lines):\n",
    "        line = lines[i].strip()\n",
    "        if not line or not line.split()[0].isdigit():\n",
    "            break\n",
    "        ligand_atom_ids.extend(int(token) for token in line.split() if token.isdigit())\n",
    "        i += 1\n",
    "\n",
    "    ligand_atom_ids_set = set(ligand_atom_ids[1:])\n",
    "    ligand_atom_lines = [l for l in atom_lines if int(l.split()[0]) in ligand_atom_ids_set]\n",
    "    ligand_atom_id_map = {int(l.split()[0]): idx+1 for idx, l in enumerate(ligand_atom_lines)}\n",
    "\n",
    "    # --- extract ligand bond ---\n",
    "    ligand_bond_lines = []\n",
    "    bond_num = 1\n",
    "    for l in bond_lines:\n",
    "        fields = l.split()\n",
    "        if len(fields) < 4: # skip protein bonds\n",
    "            continue\n",
    "        a1, a2 = int(fields[1]), int(fields[2])\n",
    "        if a1 in ligand_atom_ids_set and a2 in ligand_atom_ids_set:\n",
    "            new_a1 = ligand_atom_id_map[a1]\n",
    "            new_a2 = ligand_atom_id_map[a2]\n",
    "            new_line = f\"{bond_num:>6} {new_a1:>5} {new_a2:>5} {fields[3]}\"\n",
    "            if len(fields) > 4:\n",
    "                new_line += \" \" + \" \".join(fields[4:])\n",
    "            ligand_bond_lines.append(new_line + \"\\n\")\n",
    "            bond_num += 1\n",
    "\n",
    "    # --- Save ligand mol2 ---\n",
    "    with open(output_path, 'w') as f:\n",
    "        f.write(\"@<TRIPOS>MOLECULE\\n\")\n",
    "        f.write(f\"{mol2_path.parent.stem}\\n\")\n",
    "        f.write(f\"{len(ligand_atom_lines)} {len(ligand_bond_lines)} 1\\n\")\n",
    "        f.write(\"SMALL\\n\")\n",
    "        f.write(\"USER_CHARGES\\n\\n\")\n",
    "\n",
    "        f.write(\"@<TRIPOS>ATOM\\n\")\n",
    "        for idx, line in enumerate(ligand_atom_lines, 1):\n",
    "            tokens = line.split()\n",
    "            atom_id = idx\n",
    "            atom_name = tokens[1]\n",
    "            x, y, z = map(float, tokens[2:5])\n",
    "            atom_type = tokens[5]\n",
    "            resid = tokens[6]\n",
    "            resname = tokens[7]\n",
    "            charge = float(tokens[8]) if len(tokens) > 8 else 0.0\n",
    "\n",
    "            formatted = f\"{atom_id:>7} {atom_name:<8} {x:>10.4f} {y:>10.4f} {z:>10.4f} \"\n",
    "            formatted += f\"{atom_type:<6} {resid:>5} {resname:<8} {charge:>9.4f}\\n\"\n",
    "            f.write(formatted)\n",
    "\n",
    "        f.write(\"@<TRIPOS>BOND\\n\")\n",
    "        for line in ligand_bond_lines:\n",
    "            f.write(line)\n",
    "    \n",
    "    # print(f\"Ligand saved: {output_path}\")\n",
    "    complex = to_pdb_format(atom_lines, bond_lines, chain_lines)\n",
    "    ligand =  ligand_reader(ligand_atom_lines)\n",
    "    pocket = find_pocket_residues(complex, ligand)\n",
    "    write_pocket_to_mol2(pocket, output_path.parent / 'pocket_atoms.mol2')\n",
    "    print('Pocket saved: ', output_path.parent / 'pocket_atoms.mol2')\n",
    "    return pocket\n",
    "\n",
    "# pk = read_csar(holo)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## APO Align"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle\n",
    "from pathlib import Path\n",
    "import logging\n",
    "\n",
    "# 로그 설정\n",
    "out = Path('data/raw/')\n",
    "out.mkdir(parents=True, exist_ok=True)\n",
    "log_path = out / \"processing.log\"\n",
    "logging.basicConfig(\n",
    "    level=logging.INFO,\n",
    "    format='%(asctime)s %(levelname)s: %(message)s',\n",
    "    handlers=[\n",
    "        logging.FileHandler(log_path, mode='w'),\n",
    "        logging.StreamHandler()\n",
    "    ]\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "no_pdbbind = ['1mh5', '1d6v', '1ct8', '1i7z', '1a4k', '1a0q', '1i6v', '1c12', '2qhr', '4nyi', '4nyj', '4nym', '2hrp', '3eql', '4kmu', '4kn4', '4kn7', '2mpa', '1kcs', '1zyr']\n",
    "no_csar = ['1gz9', '2v7t', '2j4k', '2v7u', '1swk', '2vhw', '1gzc', '1bcj', '2hr6']\n",
    "no_rdkit = ['5tmp', '1ksn', '4kcx', '4lhm', '4xtw', '4xtx', '1sl3', '1mue', '1a7x', '3zjt', '4ob0', '4ob1', '4ob2', '4y63', '2fm5', '2fou', '2fov', '2foy', '3bho', '3v9b', '1nu1', '4u5t', '4i60', '3whw', '3cst', '3fxz', '3fy0', '4ie2', '4ie3', '3e6k', '3vjs', '3vjt', '2vr0', '3bwf', '3zju', '4lv1', '4wkv', '4wku', '3zp9', '4hww', '2pll', '2q2n', '3kck', '2z3z', '4l6q', '1qpf', '3e9b', '3egk', '4inr', '4inu', '4ixu', '4ixv', '1h07', '2ci9', '3wax', '3wd2', '4hze', '4i06', '4wvs', '4no1', '2eep', '4wkt', '1epq', '4rlp', '4hxq', '4bcb']\n",
    "no_match = no_pdbbind + no_csar + no_rdkit # 20 + 9 + 66\n",
    "print(len(no_match)) # no match pocket"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Final"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "## 72mins\n",
    "final_data = []\n",
    "import pandas as pd\n",
    "from rdkit import Chem\n",
    "from Bio import pairwise2\n",
    "from Bio.pairwise2 import format_alignment\n",
    "\n",
    "def sequence_reader(df):\n",
    "    atom_end = df[df['record'] == 'ATOM'].index[-1]\n",
    "    main_structure = df[:atom_end]\n",
    "    read, ridx, r_ridx = read_residue(df[:atom_end][['res_name', 'chain_id', 'res_seq', 'i_code']].values)\n",
    "    return main_structure, read, ridx, r_ridx\n",
    "\n",
    "\n",
    "def find_root_atom(df):\n",
    "    return df[df['atom_name'] == 'CA'][['chain_id', 'res_seq', 'i_code']].values\n",
    "\n",
    "\n",
    "def find_atom(df, keys):\n",
    "    result = []\n",
    "    for k in keys:\n",
    "        result.append(df[(df['chain_id'] == k[0]) & (df['res_seq'] == k[1]) & (df['i_code'] == k[2])])\n",
    "    result = pd.concat(result).reset_index(drop=True)\n",
    "    return result\n",
    "\n",
    "\n",
    "count = 0\n",
    "max_mutation = 0\n",
    "\n",
    "rmsd_results = {}\n",
    "apo_path = Path('data/raw/afdb')\n",
    "holo_path = Path('data/raw/pdb')\n",
    "\n",
    "# output path\n",
    "apo_align_path = Path('data/raw/af_align')\n",
    "apo_align_path.mkdir(exist_ok=True, parents=True)\n",
    "\n",
    "mutant_path = Path('data/raw/mutant')\n",
    "mutant_path.mkdir(exist_ok=True, parents=True)\n",
    "\n",
    "a2h_path = Path('data/preprocessed/a2h')\n",
    "a2h_path.mkdir(exist_ok=True, parents=True)\n",
    "\n",
    "# main loop\n",
    "num = 0\n",
    "for idx, row in data.iterrows():\n",
    "    pdb = row['PDB']\n",
    "    uni = row['UniProt']\n",
    "    st = row['SET']\n",
    "\n",
    "    if pdb in no_match:\n",
    "        continue\n",
    "\n",
    "    if st == 'PDB':\n",
    "        continue\n",
    "    \n",
    "    elif st != 'CSAR':\n",
    "        # path\n",
    "        apo = apo_path / f\"AF-{uni}-F1-model_v4.pdb\"    \n",
    "        holo = holo_path / st / pdb / f\"{pdb}_protein.pdb\"\n",
    "        pocket = holo_path / st / pdb / f\"{pdb}_pocket.pdb\"\n",
    "        ligand = holo_path / st / pdb / f\"{pdb}_ligand.mol2\"\n",
    "        \n",
    "        # read pdb format\n",
    "        ap = read_pdb(apo)\n",
    "        ho = read_pdb(holo)\n",
    "        pk = read_pdb(pocket)\n",
    "        \n",
    "        # read ligand\n",
    "        li = Chem.MolFromMol2File(ligand)\n",
    "        smi = Chem.CanonSmiles(Chem.MolToSmiles(li))\n",
    "        row['Ligand'] = smi\n",
    "        row['Ligand_Len'] = len(smi)\n",
    "        final_data.append(row.to_dict())\n",
    "        \n",
    "        # read pocket and parse chain_id from holo\n",
    "        ho_keys = parse_residue_and_align(ho[['res_seq', 'chain_id']].values)\n",
    "        pk['c_key'] = parse_residue_and_align(pk['res_seq'].values, chain=False)\n",
    "        pk['chain_id'] = pk['c_key'].map(ho_keys)\n",
    "        pk = pk.drop(columns=['c_key']).reset_index(drop=True)\n",
    "        pk.loc[pk['res_name'] == 'HOH', 'chain_id'] = ' '\n",
    "\n",
    "    else: # CSAR\n",
    "        \n",
    "        # path\n",
    "        apo = apo_path / f\"AF-{uni}-F1-model_v4.pdb\"\n",
    "        holo = [f for f in(holo_path / st / pdb).glob('*complex.mol2')][0]\n",
    "        \n",
    "        # read for pdb format\n",
    "        ap = read_pdb(apo)\n",
    "        ho = read_mol2(holo)\n",
    "        pk = read_csar(holo)\n",
    "        \n",
    "        # fix csar sequence\n",
    "        pk_seq = sequence_reader(pk)[1]\n",
    "        try:\n",
    "            li = Chem.MolFromMol2File(holo.parent / f'{pdb}_ligand.mol2')\n",
    "            smi = Chem.MolToSmiles(li)\n",
    "        except:\n",
    "            li = Chem.MolFromMol2File(holo.parent / f'{pdb}_ligand_rm_mg.mol2')\n",
    "            smi = Chem.MolToSmiles(li)\n",
    "        smi = Chem.CanonSmiles(smi)\n",
    "        row['Pocket'] = pk_seq\n",
    "        row['Pocket_Len'] = len(pk_seq)\n",
    "        row['Ligand'] = smi\n",
    "        row['Ligand_Len'] = len(smi)\n",
    "        final_data.append(row.to_dict())\n",
    "        \n",
    "        # remove ligand\n",
    "        ho = ho[ho['res_name'] != 'INH'].reset_index(drop=True)\n",
    "\n",
    "    num += 1\n",
    "    if num % 1000 == 0:\n",
    "        print(num)\n",
    "\n",
    "    logging.info(f\"\\n[{idx+1}/{len(data)}] Processing code: {pdb}\")\n",
    "\n",
    "    # # common process\n",
    "    # extract amino acid sequence\n",
    "    ho_main, ho_read, ho_ridx, ho_r_ridx = sequence_reader(ho)\n",
    "    pk_main, pk_read, pk_ridx, pk_r_ridx = sequence_reader(pk)\n",
    "    ap_main, ap_read, ap_ridx, ap_r_ridx = sequence_reader(ap)\n",
    "    # print(len(ho_ridx), ho_read)\n",
    "    # print(len(pk_ridx), pk_read)\n",
    "    # print(len(ap_ridx), ap_read)\n",
    "\n",
    "    # search pocket in holo\n",
    "    pk_ca = find_root_atom(pk_main)\n",
    "    ho_pk = find_atom(ho_main, pk_ca)\n",
    "    ho_ca = find_root_atom(ho_pk)\n",
    "    ho_pk_idx = sorted([ho_ridx[(idx[0], idx[1], idx[2])] for idx in ho_ca])\n",
    "    \n",
    "    # extract pocket amino acids in holo (save position)\n",
    "    ho_pk_read = ho_read[ho_pk_idx[0]:ho_pk_idx[-1]+1]\n",
    "    ho_pk_res_dict = {ho_pk_idx[0] + i:i  for i in range(len(ho_pk_read))}\n",
    "    ho_pk_res_idx = [ho_pk_res_dict[i] for i in ho_pk_idx] # for mutation position\n",
    "    \n",
    "    # sequence alignment (score params: match, mismatch, gap open, gap extension)\n",
    "    alignments = pairwise2.align.localms(ho_pk_read, ap_read, 2, -1, -5, -1)\n",
    "    ho_align, ap_align = alignments[0][:2]\n",
    "\n",
    "    # mapping position & check mutation\n",
    "    mapping_align = {}\n",
    "    align_position = {}\n",
    "    mutations = []\n",
    "    st_pos = ho_pos = ap_pos = -1\n",
    "\n",
    "    for a, b in zip(ho_align, ap_align):\n",
    "        if a != '-':\n",
    "            ho_pos += 1\n",
    "            st_pos += 1\n",
    "        if b != '-':\n",
    "            ap_pos += 1\n",
    "            st_pos += 1\n",
    "        if a != '-' and b != '-':\n",
    "            st_pos += 1\n",
    "            align_position[st_pos] = ho_pos\n",
    "            mapping_align[ho_pos] = ap_pos\n",
    "            if a != b and ho_pos in ho_pk_res_idx:\n",
    "                mutations.append((ap_pos, b, a))\n",
    "                \n",
    "    # check mutation\n",
    "    if mutations:\n",
    "        max_mutation = max(max_mutation, len(mutations))\n",
    "\n",
    "        transform = []\n",
    "        for mut in mutations:\n",
    "            ap_pk_idx, ap_res, ho_res = mut\n",
    "            ap_chain, ap_resid, _ = ap_r_ridx[ap_pk_idx]\n",
    "            transform.append((ap_chain, ap_resid, ap_res, ho_res))\n",
    "        \n",
    "        mut_out = mutant_path / f\"AF-{uni}-mut-{pdb}.pdb\"    \n",
    "        point_mutation(apo, mut_out, transform)\n",
    "\n",
    "        # reload\n",
    "        apo = mut_out\n",
    "        ap = read_pdb(apo)\n",
    "        ap_main, ap_read, ap_ridx, ap_r_ridx = sequence_reader(ap)\n",
    "    \n",
    "    # find pocket in apo (ho_pk_idx -> ho_pk_res_dict -> mapping align)\n",
    "    ap_ca = []\n",
    "    rm_ca = []\n",
    "    for i in ho_pk_idx:\n",
    "        hp_key = ho_pk_res_dict[i]  # atom within pocket\n",
    "        if hp_key in mapping_align.keys(): # atom within mapping & pocket\n",
    "            ap_ca.append(ap_r_ridx[mapping_align[hp_key]]) # find \n",
    "        else:\n",
    "            rm_ca.append(i)\n",
    "    if rm_ca: # select atom atoms only in apo\n",
    "        ho_pk = find_atom(ho_main, [ho_r_ridx[i] for i in ho_pk_idx if not i in rm_ca])\n",
    "    ap_pk = find_atom(ap_main, ap_ca) # find pocket in apo\n",
    "\n",
    "    # superimpose\n",
    "    align_out = apo_align_path / f\"AF-{uni}-align-{pdb}.pdb\"\n",
    "    rmsd = align_and_save(apo, holo, align_out)\n",
    "    rmsd_results[pdb] = rmsd\n",
    "    \n",
    "    # mutated & aligned apo structure\n",
    "    aa = read_pdb(align_out)\n",
    "    aa_main, aa_read, aa_ridx, aa_r_ridx = sequence_reader(aa)\n",
    "    aa_pk = find_atom(aa_main, ap_ca)\n",
    "    \n",
    "    # save\n",
    "    a2h_out = a2h_path / f\"{pdb}_a2h.pkl\"\n",
    "    a2h = {'APO': aa_pk, 'HOLO': ho_pk}\n",
    "    with open(a2h_out, 'wb') as f:\n",
    "        pd.to_pickle(a2h, f)\n",
    "\n",
    "print('done')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# from rdkit import Chem, RDLogger\n",
    "# RDLogger.DisableLog('rdApp.*')\n",
    "\n",
    "# for _, row in data.iterrows():\n",
    "#     pdb = row['PDB']\n",
    "#     st = row['SET']\n",
    "    \n",
    "#     if pdb in no_match:\n",
    "#         continue\n",
    "    \n",
    "#     if st == 'PDB':\n",
    "#         continue\n",
    "#     path = f'data/raw/pdb/{st}/{pdb}/{pdb}_ligand.mol2'\n",
    "#     mol = Chem.MolFromMol2File(path)\n",
    "#     if mol:\n",
    "#         smi = Chem.MolToSmiles(mol)\n",
    "#         smi = Chem.CanonSmiles(smi)\n",
    "#     else:\n",
    "#         print(pdb, st, smi)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('data/raw/rmsd_results.txt', 'w') as f:\n",
    "    for k, v in rmsd_results.items():\n",
    "        f.write(f\"{k}: {v}\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(a2h_out, 'rb') as f:\n",
    "    loaded_bundle = pd.read_pickle(f)\n",
    "loaded_bundle['APO']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "final = pd.DataFrame(final_data)\n",
    "final.to_csv('data/final.csv', index=False, header=True)\n",
    "final"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "final = final[~final['PDB'].isin(no_match)]\n",
    "final['SET'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "final[final['SET'] == 'Refined']['UniProt'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "align_pdbs = [f.stem.split('-')[3] for f in apo_align_path.glob('*.pdb')]\n",
    "len(align_pdbs) # exclude 29 pdbs (no_match)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "import warnings\n",
    "warnings.filterwarnings('ignore', category=DeprecationWarning)\n",
    "\n",
    "general = final[final['SET'] == 'General']\n",
    "refined = final[final['SET'] == 'Refined']\n",
    "\n",
    "seeds = [42, 100, 123, 456, 789]\n",
    "for idx, seed in enumerate(seeds):\n",
    "    # minimum sample per UniProt\n",
    "    num_uni = refined['UniProt'].nunique()\n",
    "    samples_per_uni = 1000 // num_uni\n",
    "\n",
    "    # sampling each UniProt (maximum samples_per_uni)\n",
    "    sampled = refined.groupby('UniProt', group_keys=False).apply(\n",
    "        lambda x: x.sample(n=min(samples_per_uni, len(x)), random_state=seed)\n",
    "    )\n",
    "\n",
    "    # remaining samples\n",
    "    remaining = 1000 - len(sampled)\n",
    "    if remaining > 0:\n",
    "        additional = refined[~refined.index.isin(sampled.index)].sample(n=remaining, random_state=seed)\n",
    "        valid = pd.concat([sampled, additional])\n",
    "\n",
    "    #check \n",
    "    valid['UniProt'].value_counts()\n",
    "\n",
    "    # create training set\n",
    "    remain_refined = refined[~refined.index.isin(valid.index)]\n",
    "    train = pd.concat([general, remain_refined]).reset_index(drop=True)\n",
    "\n",
    "    # extract pdb list\n",
    "    train_pdb = train['PDB'].tolist()\n",
    "    valid_pdb = valid['PDB'].tolist()\n",
    "    fold_pdb = {'SEED': seed, 'TRN': train_pdb, 'VAL': valid_pdb}\n",
    "\n",
    "    # save\n",
    "    with open(f'cache/fold_{idx+1}.json', 'w') as f:\n",
    "        json.dump(fold_pdb, f)\n",
    "    \n",
    "    # check\n",
    "    print(f'Fold {idx+1}: Seed {seed}')\n",
    "    print(len(train), len(valid))\n",
    "    print()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "num_uni"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print('done')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from rdkit import RDLogger\n",
    "RDLogger.DisableLog('rdApp.*')\n",
    "count = 0\n",
    "box = {}\n",
    "no_rdkit = []\n",
    "for _, row in final.iterrows():\n",
    "    pdb = row['PDB']\n",
    "    st = row['SET']\n",
    "    # smi = row['Ligand']\n",
    "    # mol = Chem.MolFromSmiles(smi)\n",
    "    # if mol is None:\n",
    "    #     count += 1\n",
    "    #     box[st] = box.get(st, 0) + 1\n",
    "    # else:\n",
    "    #     smi = Chem.CanonSmiles(Chem.MolToSmiles(mol))\n",
    "    \n",
    "    \n",
    "    path = f'data/raw/pdb/{st}/{pdb}/{pdb}_ligand.mol2'\n",
    "    mol = Chem.MolFromMol2File(path)\n",
    "    if mol is None:\n",
    "        count += 1\n",
    "        box[st] = box.get(st, 0) + 1\n",
    "        no_rdkit.append(pdb)\n",
    "        print(pdb, st, row['Ligand'])\n",
    "    else:\n",
    "        smi = Chem.CanonSmiles(Chem.MolToSmiles(mol))            \n",
    "print(box, count)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ff = final[final['PDB'] != '2qvu'].reset_index(drop=True)\n",
    "len(ff), len(ff['UniProt'].unique())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "ff.to_csv('data/data_info.csv', index=False, header=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Dummy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# search largest vector\n",
    "import pickle\n",
    "from pathlib import Path\n",
    "\n",
    "def find_largest_vector(directory):\n",
    "    directory = Path(directory)\n",
    "    max_count = 0\n",
    "    max_ca = 0\n",
    "    max_file = None\n",
    "    max_ca_file = None\n",
    "\n",
    "    with open(a2h_path / f\"{intereset}_a2h.pkl\", 'rb') as f:\n",
    "        lb = pd.read_pickle(f)\n",
    "        count = len(data['APO'])\n",
    "        ca = pd.DataFrame(data['APO'])\n",
    "        ca = ca[ca['atom_name'] == 'CA']\n",
    "        if count > max_count:\n",
    "            max_count = count\n",
    "            max_file = pkl_file\n",
    "        \n",
    "        ca_count = len(ca)\n",
    "        if ca_count > max_ca:\n",
    "            max_ca = ca_count\n",
    "            max_ca_file = pkl_file\n",
    "\n",
    "    print(f\"Largest vector: {max_file.name} with {max_count} entries\")\n",
    "    print(f\"Largest CA vector: {max_ca_file.name} with {max_ca} entries\")\n",
    "\n",
    "find_largest_vector(\"data/raw/af_align\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle\n",
    "from pathlib import Path\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "def find_files_with_coord_diff(directory, threshold=1e-3):\n",
    "    directory = Path(directory)\n",
    "    files_with_diff = []\n",
    "\n",
    "    for pkl_file in directory.glob(\"*.pkl\"):\n",
    "        with open(pkl_file, \"rb\") as f:\n",
    "            data = pickle.load(f)\n",
    "        apo = pd.DataFrame(data['APO'])\n",
    "        holo = pd.DataFrame(data['HOLO'])\n",
    "\n",
    "        # 기준: res_seq, atom_name이 같은 원자끼리 좌표 비교\n",
    "        merged = pd.merge(\n",
    "            apo, holo,\n",
    "            on=['res_seq', 'atom_name'],\n",
    "            suffixes=('_apo', '_holo')\n",
    "        )\n",
    "\n",
    "        # coord 컬럼에서 좌표 추출\n",
    "        def extract_coords(row, prefix):\n",
    "            return np.array(row[f'coord_{prefix}'])\n",
    "\n",
    "        # 좌표 차이 계산\n",
    "        merged['coord_diff'] = merged.apply(\n",
    "            lambda row: np.linalg.norm(\n",
    "                extract_coords(row, 'apo') - extract_coords(row, 'holo')\n",
    "            ),\n",
    "            axis=1\n",
    "        )\n",
    "\n",
    "        # 임계값(threshold) 이상 차이가 있는 경우\n",
    "        if (merged['coord_diff'] > threshold).any():\n",
    "            files_with_diff.append(pkl_file.name)\n",
    "\n",
    "    print(f\"Files with coordinate differences (>{threshold} Å):\")\n",
    "    for fname in files_with_diff:\n",
    "        print(fname)\n",
    "\n",
    "find_files_with_coord_diff(\"APO_to_Holo\", threshold=0.1)  # 0.1Å 이상 차이"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3sjt General C[C@]([NH3+])(CCC[CH2]~B(O)(O)O)C(=O)[O-]\n",
      "3skk General [NH3+][C@@](CCC[CH2]~B(O)(O)O)(C(=O)[O-])C(F)F\n",
      "4kii General CCCCCCCCCCCC(=O)[N+]12c3cccc[n]3~[Rh+2]1~[n]1ccccc12\n",
      "1ijr General CC(=O)N[C@@H](Cc1ccc(OCC(=O)[O-])[c](~[P+](=O)([O-])[O-])c1)C(=O)N[C@@H](C)c1ccc(OCC2CCCCC2)c(C(N)=O)c1\n",
      "4z46 General C1CC[C@H]2[NH2+]~[Pt+]~[NH2+][C@@H]2C1\n"
     ]
    }
   ],
   "source": [
    "for _, row in ff.iterrows():\n",
    "    if '~' in row['Ligand']:\n",
    "        print(row['PDB'], row['SET'], row['Ligand'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 256,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['1uld', '1q0y', '1xw6']"
      ]
     },
     "execution_count": 256,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sse_path = Path('data/preprocessed/sse')\n",
    "no_sse_pdb = []\n",
    "for _, row in ff.iterrows():\n",
    "    pdb = row['PDB']\n",
    "    sse_global = sse_path / 'global' / f'{pdb}.csv'\n",
    "    sse_pocket = sse_path / 'pocket' / f'{pdb}.csv'\n",
    "    if not sse_global.exists() or not sse_pocket.exists():\n",
    "        no_sse_pdb.append(pdb)\n",
    "    # if '~' in row['Ligand']:\n",
    "    #     print(row['PDB'], row['SET'], row['Ligand'])\n",
    "no_sse_pdb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "done\n"
     ]
    }
   ],
   "source": [
    "import shutil\n",
    "cc = Path('data/raw/copy_csar')\n",
    "cc.mkdir(parents=True, exist_ok=True)\n",
    "csar_path = Path('data/raw/pdb/CSAR')\n",
    "for pdb in csar_path.glob('*'):\n",
    "    file = [f for f in pdb.glob('*complex.mol2')][0]\n",
    "    shutil.copy(file, cc / f'{pdb.stem}_protein.mol2')\n",
    "print('done')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 250,
   "metadata": {},
   "outputs": [],
   "source": [
    "def dssp_to_ss8(dssp_path):\n",
    "    ss8_seq = []\n",
    "    with open(dssp_path, 'r') as f:\n",
    "        parsing = False\n",
    "        for line in f:\n",
    "            if not parsing:\n",
    "                if line.startswith(\"  #  RESIDUE AA\"):\n",
    "                    parsing = True\n",
    "                continue\n",
    "\n",
    "            if len(line) < 17:\n",
    "                print(\"over\")\n",
    "                continue\n",
    "            \n",
    "            ss_char = line[16]\n",
    "            # print(line.strip())\n",
    "            if ss_char == ' ':\n",
    "                ss_char = 'C'  # loop\n",
    "            ss8_seq.append(ss_char)\n",
    "\n",
    "    return ''.join(ss8_seq)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 251,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'CEEEEEEESSCCTTCSSEEEEETCEECTTSCEESSCBCTTSPBPSSCEEEEEESSCEECBCTTTCCBCEEEEEEEEECCCCCSSSCCCEEEEEEEECTTCCCCBCGGGTTTBSSSSCCGGGCCEEEEEECSCCTTSCSCSSEEEEEESSSSCSEEEECCCCTTCCEEEEEEEETTTTEEEEEEEETTTTEEEEEEEECCGGGTSCSEEEEEEEEEECSSTTCCCCCEEEEEEEEEEECC'"
      ]
     },
     "execution_count": 251,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tmp = r\"data\\preprocessed\\CSAR_DSSP\\1ax1.dssp\"\n",
    "dssp_to_ss8(tmp)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [],
   "source": [
    "ss8_out = Path('data/preprocessed/csar.out.ss8')\n",
    "dssps = Path('data/preprocessed/CSAR_DSSP').glob('*.dssp')\n",
    "\n",
    "with open(ss8_out, 'w') as f:\n",
    "    for dssp in dssps:\n",
    "        ss8 = dssp_to_ss8(dssp)\n",
    "        f.write(f'>{dssp.stem}\\n{ss8}\\n')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# SSE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [],
   "source": [
    "csar_ff = ff[ff['SET'] == 'CSAR'].reset_index(drop=True)\n",
    "pocs = []\n",
    "glos = []\n",
    "for _, row in csar_ff.iterrows():\n",
    "    pdb = row['PDB']\n",
    "    poc = row['Pocket']\n",
    "    glo = row['Global']\n",
    "    pocs.append({'id': pdb, 'seq': poc})\n",
    "    glos.append({'id': pdb, 'seq': glo})\n",
    "\n",
    "pocs = pd.DataFrame(pocs)\n",
    "glos = pd.DataFrame(glos)\n",
    "pocs.to_csv('data/preprocessed/csar_pockets.csv', index=False)\n",
    "glos.to_csv('data/preprocessed/csar_proteins.csv', index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 155,
   "metadata": {},
   "outputs": [],
   "source": [
    "a2h_path = Path('data/preprocessed/a2h')\n",
    "pocket_inform = {}\n",
    "for _, row in csar_ff.iterrows():\n",
    "    pdb = row['PDB']\n",
    "    with open(a2h_path / f'{pdb}_a2h.pkl', 'rb') as f:\n",
    "        data = pickle.load(f)['HOLO']\n",
    "    cas = data[data['atom_name'] == 'CA'][['res_seq', 'chain_id']].values\n",
    "    pocket_inform[pdb] = [f\"{v}-{k}\" for k, v in cas]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 257,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>record</th>\n",
       "      <th>atom_serial</th>\n",
       "      <th>atom_name</th>\n",
       "      <th>alt_loc</th>\n",
       "      <th>res_name</th>\n",
       "      <th>chain_id</th>\n",
       "      <th>res_seq</th>\n",
       "      <th>i_code</th>\n",
       "      <th>x</th>\n",
       "      <th>y</th>\n",
       "      <th>z</th>\n",
       "      <th>atom_type</th>\n",
       "      <th>occupancy</th>\n",
       "      <th>b_factor</th>\n",
       "      <th>element</th>\n",
       "      <th>charge</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ATOM</td>\n",
       "      <td>3255</td>\n",
       "      <td>CA</td>\n",
       "      <td></td>\n",
       "      <td>GLU</td>\n",
       "      <td>B</td>\n",
       "      <td>208</td>\n",
       "      <td></td>\n",
       "      <td>37.173</td>\n",
       "      <td>13.994</td>\n",
       "      <td>29.573</td>\n",
       "      <td>C.3</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>C</td>\n",
       "      <td>0.0397</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ATOM</td>\n",
       "      <td>4363</td>\n",
       "      <td>CA</td>\n",
       "      <td></td>\n",
       "      <td>ASN</td>\n",
       "      <td>B</td>\n",
       "      <td>276</td>\n",
       "      <td></td>\n",
       "      <td>46.293</td>\n",
       "      <td>22.036</td>\n",
       "      <td>41.599</td>\n",
       "      <td>C.3</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>C</td>\n",
       "      <td>0.0143</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>ATOM</td>\n",
       "      <td>4189</td>\n",
       "      <td>CA</td>\n",
       "      <td></td>\n",
       "      <td>ARG</td>\n",
       "      <td>B</td>\n",
       "      <td>265</td>\n",
       "      <td></td>\n",
       "      <td>28.335</td>\n",
       "      <td>17.624</td>\n",
       "      <td>52.664</td>\n",
       "      <td>C.3</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>C</td>\n",
       "      <td>-0.2637</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>ATOM</td>\n",
       "      <td>2941</td>\n",
       "      <td>CA</td>\n",
       "      <td></td>\n",
       "      <td>VAL</td>\n",
       "      <td>B</td>\n",
       "      <td>188</td>\n",
       "      <td></td>\n",
       "      <td>41.524</td>\n",
       "      <td>21.656</td>\n",
       "      <td>40.304</td>\n",
       "      <td>C.3</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>C</td>\n",
       "      <td>-0.0875</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>ATOM</td>\n",
       "      <td>3284</td>\n",
       "      <td>CA</td>\n",
       "      <td></td>\n",
       "      <td>VAL</td>\n",
       "      <td>B</td>\n",
       "      <td>210</td>\n",
       "      <td></td>\n",
       "      <td>33.537</td>\n",
       "      <td>17.256</td>\n",
       "      <td>32.289</td>\n",
       "      <td>C.3</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>C</td>\n",
       "      <td>-0.0875</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>ATOM</td>\n",
       "      <td>3466</td>\n",
       "      <td>CA</td>\n",
       "      <td></td>\n",
       "      <td>ALA</td>\n",
       "      <td>B</td>\n",
       "      <td>221</td>\n",
       "      <td></td>\n",
       "      <td>37.590</td>\n",
       "      <td>8.616</td>\n",
       "      <td>50.321</td>\n",
       "      <td>C.3</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>C</td>\n",
       "      <td>0.0337</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>ATOM</td>\n",
       "      <td>2977</td>\n",
       "      <td>CA</td>\n",
       "      <td></td>\n",
       "      <td>ASN</td>\n",
       "      <td>B</td>\n",
       "      <td>190</td>\n",
       "      <td></td>\n",
       "      <td>39.600</td>\n",
       "      <td>22.991</td>\n",
       "      <td>46.478</td>\n",
       "      <td>C.3</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>C</td>\n",
       "      <td>0.0143</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>ATOM</td>\n",
       "      <td>3500</td>\n",
       "      <td>CA</td>\n",
       "      <td></td>\n",
       "      <td>GLY</td>\n",
       "      <td>B</td>\n",
       "      <td>223</td>\n",
       "      <td></td>\n",
       "      <td>32.013</td>\n",
       "      <td>10.789</td>\n",
       "      <td>47.963</td>\n",
       "      <td>C.3</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>C</td>\n",
       "      <td>-0.0252</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>ATOM</td>\n",
       "      <td>3319</td>\n",
       "      <td>CA</td>\n",
       "      <td></td>\n",
       "      <td>VAL</td>\n",
       "      <td>B</td>\n",
       "      <td>212</td>\n",
       "      <td></td>\n",
       "      <td>33.541</td>\n",
       "      <td>17.011</td>\n",
       "      <td>38.645</td>\n",
       "      <td>C.3</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>C</td>\n",
       "      <td>-0.0875</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67</th>\n",
       "      <td>ATOM</td>\n",
       "      <td>3180</td>\n",
       "      <td>CA</td>\n",
       "      <td></td>\n",
       "      <td>SER</td>\n",
       "      <td>B</td>\n",
       "      <td>203</td>\n",
       "      <td></td>\n",
       "      <td>37.984</td>\n",
       "      <td>19.221</td>\n",
       "      <td>38.963</td>\n",
       "      <td>C.3</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>C</td>\n",
       "      <td>-0.0249</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73</th>\n",
       "      <td>ATOM</td>\n",
       "      <td>3144</td>\n",
       "      <td>CA</td>\n",
       "      <td></td>\n",
       "      <td>HIS</td>\n",
       "      <td>B</td>\n",
       "      <td>201</td>\n",
       "      <td></td>\n",
       "      <td>36.426</td>\n",
       "      <td>19.920</td>\n",
       "      <td>45.121</td>\n",
       "      <td>C.3</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>C</td>\n",
       "      <td>0.0188</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83</th>\n",
       "      <td>ATOM</td>\n",
       "      <td>3355</td>\n",
       "      <td>CA</td>\n",
       "      <td></td>\n",
       "      <td>ASN</td>\n",
       "      <td>B</td>\n",
       "      <td>214</td>\n",
       "      <td></td>\n",
       "      <td>32.947</td>\n",
       "      <td>16.309</td>\n",
       "      <td>45.148</td>\n",
       "      <td>C.3</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>C</td>\n",
       "      <td>0.0143</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>91</th>\n",
       "      <td>ATOM</td>\n",
       "      <td>3521</td>\n",
       "      <td>CA</td>\n",
       "      <td></td>\n",
       "      <td>GLU</td>\n",
       "      <td>B</td>\n",
       "      <td>225</td>\n",
       "      <td></td>\n",
       "      <td>30.571</td>\n",
       "      <td>12.595</td>\n",
       "      <td>42.104</td>\n",
       "      <td>C.3</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>C</td>\n",
       "      <td>0.0397</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100</th>\n",
       "      <td>ATOM</td>\n",
       "      <td>3210</td>\n",
       "      <td>CA</td>\n",
       "      <td></td>\n",
       "      <td>ARG</td>\n",
       "      <td>B</td>\n",
       "      <td>205</td>\n",
       "      <td></td>\n",
       "      <td>38.819</td>\n",
       "      <td>18.109</td>\n",
       "      <td>32.927</td>\n",
       "      <td>C.3</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>C</td>\n",
       "      <td>-0.2637</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>111</th>\n",
       "      <td>ATOM</td>\n",
       "      <td>3380</td>\n",
       "      <td>CA</td>\n",
       "      <td></td>\n",
       "      <td>ARG</td>\n",
       "      <td>B</td>\n",
       "      <td>216</td>\n",
       "      <td></td>\n",
       "      <td>36.070</td>\n",
       "      <td>14.277</td>\n",
       "      <td>51.087</td>\n",
       "      <td>C.3</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>C</td>\n",
       "      <td>-0.2637</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>122</th>\n",
       "      <td>ATOM</td>\n",
       "      <td>3551</td>\n",
       "      <td>CA</td>\n",
       "      <td></td>\n",
       "      <td>ARG</td>\n",
       "      <td>B</td>\n",
       "      <td>227</td>\n",
       "      <td></td>\n",
       "      <td>29.437</td>\n",
       "      <td>16.051</td>\n",
       "      <td>37.225</td>\n",
       "      <td>C.3</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>C</td>\n",
       "      <td>-0.2637</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>133</th>\n",
       "      <td>ATOM</td>\n",
       "      <td>4132</td>\n",
       "      <td>CA</td>\n",
       "      <td></td>\n",
       "      <td>TYR</td>\n",
       "      <td>B</td>\n",
       "      <td>262</td>\n",
       "      <td></td>\n",
       "      <td>26.370</td>\n",
       "      <td>21.864</td>\n",
       "      <td>45.199</td>\n",
       "      <td>C.3</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>C</td>\n",
       "      <td>-0.0014</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>145</th>\n",
       "      <td>ATOM</td>\n",
       "      <td>3248</td>\n",
       "      <td>CA</td>\n",
       "      <td></td>\n",
       "      <td>GLY</td>\n",
       "      <td>B</td>\n",
       "      <td>207</td>\n",
       "      <td></td>\n",
       "      <td>40.168</td>\n",
       "      <td>15.721</td>\n",
       "      <td>28.033</td>\n",
       "      <td>C.3</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>C</td>\n",
       "      <td>-0.0252</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>149</th>\n",
       "      <td>ATOM</td>\n",
       "      <td>3087</td>\n",
       "      <td>CA</td>\n",
       "      <td></td>\n",
       "      <td>ILE</td>\n",
       "      <td>B</td>\n",
       "      <td>198</td>\n",
       "      <td></td>\n",
       "      <td>37.812</td>\n",
       "      <td>19.009</td>\n",
       "      <td>51.210</td>\n",
       "      <td>C.3</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>C</td>\n",
       "      <td>-0.0597</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>157</th>\n",
       "      <td>ATOM</td>\n",
       "      <td>2927</td>\n",
       "      <td>CA</td>\n",
       "      <td></td>\n",
       "      <td>THR</td>\n",
       "      <td>B</td>\n",
       "      <td>187</td>\n",
       "      <td></td>\n",
       "      <td>42.805</td>\n",
       "      <td>21.836</td>\n",
       "      <td>36.726</td>\n",
       "      <td>C.3</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>C</td>\n",
       "      <td>-0.0389</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>164</th>\n",
       "      <td>ATOM</td>\n",
       "      <td>3459</td>\n",
       "      <td>CA</td>\n",
       "      <td></td>\n",
       "      <td>GLY</td>\n",
       "      <td>B</td>\n",
       "      <td>220</td>\n",
       "      <td></td>\n",
       "      <td>38.711</td>\n",
       "      <td>9.713</td>\n",
       "      <td>53.774</td>\n",
       "      <td>C.3</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>C</td>\n",
       "      <td>-0.0252</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>168</th>\n",
       "      <td>ATOM</td>\n",
       "      <td>3507</td>\n",
       "      <td>CA</td>\n",
       "      <td></td>\n",
       "      <td>PRO</td>\n",
       "      <td>B</td>\n",
       "      <td>224</td>\n",
       "      <td></td>\n",
       "      <td>29.606</td>\n",
       "      <td>10.384</td>\n",
       "      <td>45.032</td>\n",
       "      <td>C.3</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>C</td>\n",
       "      <td>-0.0266</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>175</th>\n",
       "      <td>ATOM</td>\n",
       "      <td>2957</td>\n",
       "      <td>CA</td>\n",
       "      <td></td>\n",
       "      <td>PHE</td>\n",
       "      <td>B</td>\n",
       "      <td>189</td>\n",
       "      <td></td>\n",
       "      <td>40.366</td>\n",
       "      <td>24.143</td>\n",
       "      <td>42.940</td>\n",
       "      <td>C.3</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>C</td>\n",
       "      <td>-0.0024</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>186</th>\n",
       "      <td>ATOM</td>\n",
       "      <td>3476</td>\n",
       "      <td>CA</td>\n",
       "      <td></td>\n",
       "      <td>TRP</td>\n",
       "      <td>B</td>\n",
       "      <td>222</td>\n",
       "      <td></td>\n",
       "      <td>35.792</td>\n",
       "      <td>10.740</td>\n",
       "      <td>47.724</td>\n",
       "      <td>C.3</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>C</td>\n",
       "      <td>-0.0275</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>200</th>\n",
       "      <td>ATOM</td>\n",
       "      <td>3125</td>\n",
       "      <td>CA</td>\n",
       "      <td></td>\n",
       "      <td>LEU</td>\n",
       "      <td>B</td>\n",
       "      <td>200</td>\n",
       "      <td></td>\n",
       "      <td>34.270</td>\n",
       "      <td>21.566</td>\n",
       "      <td>47.792</td>\n",
       "      <td>C.3</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>C</td>\n",
       "      <td>-0.0518</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>208</th>\n",
       "      <td>ATOM</td>\n",
       "      <td>3300</td>\n",
       "      <td>CA</td>\n",
       "      <td></td>\n",
       "      <td>ILE</td>\n",
       "      <td>B</td>\n",
       "      <td>211</td>\n",
       "      <td></td>\n",
       "      <td>33.692</td>\n",
       "      <td>19.168</td>\n",
       "      <td>35.547</td>\n",
       "      <td>C.3</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>C</td>\n",
       "      <td>-0.0597</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>216</th>\n",
       "      <td>ATOM</td>\n",
       "      <td>3369</td>\n",
       "      <td>CA</td>\n",
       "      <td></td>\n",
       "      <td>SER</td>\n",
       "      <td>B</td>\n",
       "      <td>215</td>\n",
       "      <td></td>\n",
       "      <td>33.892</td>\n",
       "      <td>16.449</td>\n",
       "      <td>48.837</td>\n",
       "      <td>C.3</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>C</td>\n",
       "      <td>-0.0249</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>222</th>\n",
       "      <td>ATOM</td>\n",
       "      <td>3161</td>\n",
       "      <td>CA</td>\n",
       "      <td></td>\n",
       "      <td>ILE</td>\n",
       "      <td>B</td>\n",
       "      <td>202</td>\n",
       "      <td></td>\n",
       "      <td>36.207</td>\n",
       "      <td>21.311</td>\n",
       "      <td>41.593</td>\n",
       "      <td>C.3</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>C</td>\n",
       "      <td>-0.0597</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>230</th>\n",
       "      <td>ATOM</td>\n",
       "      <td>4377</td>\n",
       "      <td>CA</td>\n",
       "      <td></td>\n",
       "      <td>ALA</td>\n",
       "      <td>B</td>\n",
       "      <td>277</td>\n",
       "      <td></td>\n",
       "      <td>47.369</td>\n",
       "      <td>21.313</td>\n",
       "      <td>38.008</td>\n",
       "      <td>C.3</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>C</td>\n",
       "      <td>0.0337</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>235</th>\n",
       "      <td>ATOM</td>\n",
       "      <td>3335</td>\n",
       "      <td>CA</td>\n",
       "      <td></td>\n",
       "      <td>PHE</td>\n",
       "      <td>B</td>\n",
       "      <td>213</td>\n",
       "      <td></td>\n",
       "      <td>32.381</td>\n",
       "      <td>18.356</td>\n",
       "      <td>41.992</td>\n",
       "      <td>C.3</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>C</td>\n",
       "      <td>-0.0024</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>246</th>\n",
       "      <td>ATOM</td>\n",
       "      <td>3191</td>\n",
       "      <td>CA</td>\n",
       "      <td></td>\n",
       "      <td>ILE</td>\n",
       "      <td>B</td>\n",
       "      <td>204</td>\n",
       "      <td></td>\n",
       "      <td>37.888</td>\n",
       "      <td>20.778</td>\n",
       "      <td>35.482</td>\n",
       "      <td>C.3</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>C</td>\n",
       "      <td>-0.0597</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>254</th>\n",
       "      <td>ATOM</td>\n",
       "      <td>3536</td>\n",
       "      <td>CA</td>\n",
       "      <td></td>\n",
       "      <td>GLU</td>\n",
       "      <td>B</td>\n",
       "      <td>226</td>\n",
       "      <td></td>\n",
       "      <td>28.391</td>\n",
       "      <td>15.404</td>\n",
       "      <td>40.800</td>\n",
       "      <td>C.3</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>C</td>\n",
       "      <td>0.0397</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>263</th>\n",
       "      <td>ATOM</td>\n",
       "      <td>2913</td>\n",
       "      <td>CA</td>\n",
       "      <td></td>\n",
       "      <td>THR</td>\n",
       "      <td>B</td>\n",
       "      <td>186</td>\n",
       "      <td></td>\n",
       "      <td>42.898</td>\n",
       "      <td>18.886</td>\n",
       "      <td>34.333</td>\n",
       "      <td>C.3</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>C</td>\n",
       "      <td>-0.0389</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    record  atom_serial atom_name alt_loc res_name chain_id  res_seq i_code  \\\n",
       "1     ATOM         3255        CA              GLU        B      208          \n",
       "10    ATOM         4363        CA              ASN        B      276          \n",
       "18    ATOM         4189        CA              ARG        B      265          \n",
       "29    ATOM         2941        CA              VAL        B      188          \n",
       "36    ATOM         3284        CA              VAL        B      210          \n",
       "43    ATOM         3466        CA              ALA        B      221          \n",
       "48    ATOM         2977        CA              ASN        B      190          \n",
       "56    ATOM         3500        CA              GLY        B      223          \n",
       "60    ATOM         3319        CA              VAL        B      212          \n",
       "67    ATOM         3180        CA              SER        B      203          \n",
       "73    ATOM         3144        CA              HIS        B      201          \n",
       "83    ATOM         3355        CA              ASN        B      214          \n",
       "91    ATOM         3521        CA              GLU        B      225          \n",
       "100   ATOM         3210        CA              ARG        B      205          \n",
       "111   ATOM         3380        CA              ARG        B      216          \n",
       "122   ATOM         3551        CA              ARG        B      227          \n",
       "133   ATOM         4132        CA              TYR        B      262          \n",
       "145   ATOM         3248        CA              GLY        B      207          \n",
       "149   ATOM         3087        CA              ILE        B      198          \n",
       "157   ATOM         2927        CA              THR        B      187          \n",
       "164   ATOM         3459        CA              GLY        B      220          \n",
       "168   ATOM         3507        CA              PRO        B      224          \n",
       "175   ATOM         2957        CA              PHE        B      189          \n",
       "186   ATOM         3476        CA              TRP        B      222          \n",
       "200   ATOM         3125        CA              LEU        B      200          \n",
       "208   ATOM         3300        CA              ILE        B      211          \n",
       "216   ATOM         3369        CA              SER        B      215          \n",
       "222   ATOM         3161        CA              ILE        B      202          \n",
       "230   ATOM         4377        CA              ALA        B      277          \n",
       "235   ATOM         3335        CA              PHE        B      213          \n",
       "246   ATOM         3191        CA              ILE        B      204          \n",
       "254   ATOM         3536        CA              GLU        B      226          \n",
       "263   ATOM         2913        CA              THR        B      186          \n",
       "\n",
       "          x       y       z atom_type  occupancy  b_factor element   charge  \n",
       "1    37.173  13.994  29.573       C.3        1.0       0.0       C   0.0397  \n",
       "10   46.293  22.036  41.599       C.3        1.0       0.0       C   0.0143  \n",
       "18   28.335  17.624  52.664       C.3        1.0       0.0       C  -0.2637  \n",
       "29   41.524  21.656  40.304       C.3        1.0       0.0       C  -0.0875  \n",
       "36   33.537  17.256  32.289       C.3        1.0       0.0       C  -0.0875  \n",
       "43   37.590   8.616  50.321       C.3        1.0       0.0       C   0.0337  \n",
       "48   39.600  22.991  46.478       C.3        1.0       0.0       C   0.0143  \n",
       "56   32.013  10.789  47.963       C.3        1.0       0.0       C  -0.0252  \n",
       "60   33.541  17.011  38.645       C.3        1.0       0.0       C  -0.0875  \n",
       "67   37.984  19.221  38.963       C.3        1.0       0.0       C  -0.0249  \n",
       "73   36.426  19.920  45.121       C.3        1.0       0.0       C   0.0188  \n",
       "83   32.947  16.309  45.148       C.3        1.0       0.0       C   0.0143  \n",
       "91   30.571  12.595  42.104       C.3        1.0       0.0       C   0.0397  \n",
       "100  38.819  18.109  32.927       C.3        1.0       0.0       C  -0.2637  \n",
       "111  36.070  14.277  51.087       C.3        1.0       0.0       C  -0.2637  \n",
       "122  29.437  16.051  37.225       C.3        1.0       0.0       C  -0.2637  \n",
       "133  26.370  21.864  45.199       C.3        1.0       0.0       C  -0.0014  \n",
       "145  40.168  15.721  28.033       C.3        1.0       0.0       C  -0.0252  \n",
       "149  37.812  19.009  51.210       C.3        1.0       0.0       C  -0.0597  \n",
       "157  42.805  21.836  36.726       C.3        1.0       0.0       C  -0.0389  \n",
       "164  38.711   9.713  53.774       C.3        1.0       0.0       C  -0.0252  \n",
       "168  29.606  10.384  45.032       C.3        1.0       0.0       C  -0.0266  \n",
       "175  40.366  24.143  42.940       C.3        1.0       0.0       C  -0.0024  \n",
       "186  35.792  10.740  47.724       C.3        1.0       0.0       C  -0.0275  \n",
       "200  34.270  21.566  47.792       C.3        1.0       0.0       C  -0.0518  \n",
       "208  33.692  19.168  35.547       C.3        1.0       0.0       C  -0.0597  \n",
       "216  33.892  16.449  48.837       C.3        1.0       0.0       C  -0.0249  \n",
       "222  36.207  21.311  41.593       C.3        1.0       0.0       C  -0.0597  \n",
       "230  47.369  21.313  38.008       C.3        1.0       0.0       C   0.0337  \n",
       "235  32.381  18.356  41.992       C.3        1.0       0.0       C  -0.0024  \n",
       "246  37.888  20.778  35.482       C.3        1.0       0.0       C  -0.0597  \n",
       "254  28.391  15.404  40.800       C.3        1.0       0.0       C   0.0397  \n",
       "263  42.898  18.886  34.333       C.3        1.0       0.0       C  -0.0389  "
      ]
     },
     "execution_count": 257,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "with open(a2h_path / f'1uld_a2h.pkl', 'rb') as f:\n",
    "    data = pickle.load(f)['HOLO']\n",
    "data[data['atom_name'] == 'CA']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 260,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['==== Secondary Structure Definition by the program DSSP, NKI version 4.5.1                         ==== DATE=2025-05-13        .\\n', 'REFERENCE W. KABSCH AND C.SANDER, BIOPOLYMERS 22 (1983) 2577-2637                                                              .\\n', 'HEADER    SUGAR BINDING PROTEIN                   12-SEP-03   1ULD                                                             .\\n', 'COMPND    MOL_ID: 1; MOLECULE: galectin-2; CHAIN: A, B, C, D; SYNONYM: CGL2; ENGINEERED: YES                                   .\\n', 'SOURCE    MOL_ID: 1; GENE: cgl2; ORGANISM_SCIENTIFIC: Coprinopsis cinerea; ORGANISM_TAXID: 5346; EXPRESSION_SYSTEM: Sacchar... .\\n', 'AUTHOR    P.J.Walser; P.W.Haebel; M.Kuenzler; U.Kues; M.Aebi; N.Ban                                                            .\\n', '  600  4  0  0  0 TOTAL NUMBER OF RESIDUES, NUMBER OF CHAINS, NUMBER OF SS-BRIDGES(TOTAL,INTRACHAIN,INTERCHAIN)                .\\n', ' 27366.9   ACCESSIBLE SURFACE OF PROTEIN (ANGSTROM**2)                                                                         .\\n', '  396 66.0   TOTAL NUMBER OF HYDROGEN BONDS OF TYPE O(I)-->H-N(J)  , SAME NUMBER PER 100 RESIDUES                              .\\n', '    0  0.0   TOTAL NUMBER OF HYDROGEN BONDS IN     PARALLEL BRIDGES, SAME NUMBER PER 100 RESIDUES                              .\\n', '  265 44.2   TOTAL NUMBER OF HYDROGEN BONDS IN ANTIPARALLEL BRIDGES, SAME NUMBER PER 100 RESIDUES                              .\\n', '    4  0.7   TOTAL NUMBER OF HYDROGEN BONDS OF TYPE O(I)-->H-N(I-5), SAME NUMBER PER 100 RESIDUES                              .\\n', '    4  0.7   TOTAL NUMBER OF HYDROGEN BONDS OF TYPE O(I)-->H-N(I-4), SAME NUMBER PER 100 RESIDUES                              .\\n', '    4  0.7   TOTAL NUMBER OF HYDROGEN BONDS OF TYPE O(I)-->H-N(I-3), SAME NUMBER PER 100 RESIDUES                              .\\n', '    1  0.2   TOTAL NUMBER OF HYDROGEN BONDS OF TYPE O(I)-->H-N(I-2), SAME NUMBER PER 100 RESIDUES                              .\\n', '    0  0.0   TOTAL NUMBER OF HYDROGEN BONDS OF TYPE O(I)-->H-N(I-1), SAME NUMBER PER 100 RESIDUES                              .\\n', '    0  0.0   TOTAL NUMBER OF HYDROGEN BONDS OF TYPE O(I)-->H-N(I+0), SAME NUMBER PER 100 RESIDUES                              .\\n', '    0  0.0   TOTAL NUMBER OF HYDROGEN BONDS OF TYPE O(I)-->H-N(I+1), SAME NUMBER PER 100 RESIDUES                              .\\n', '   79 13.2   TOTAL NUMBER OF HYDROGEN BONDS OF TYPE O(I)-->H-N(I+2), SAME NUMBER PER 100 RESIDUES                              .\\n', '   28  4.7   TOTAL NUMBER OF HYDROGEN BONDS OF TYPE O(I)-->H-N(I+3), SAME NUMBER PER 100 RESIDUES                              .\\n', '    3  0.5   TOTAL NUMBER OF HYDROGEN BONDS OF TYPE O(I)-->H-N(I+4), SAME NUMBER PER 100 RESIDUES                              .\\n', '    4  0.7   TOTAL NUMBER OF HYDROGEN BONDS OF TYPE O(I)-->H-N(I+5), SAME NUMBER PER 100 RESIDUES                              .\\n', '  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30     *** HISTOGRAMS OF ***           .\\n', '  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0    RESIDUES PER ALPHA HELIX         .\\n', '  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0    PARALLEL BRIDGES PER LADDER      .\\n', '  3  0  0 16 13 11  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0    ANTIPARALLEL BRIDGES PER LADDER  .\\n', '  0  0  0  0  8  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0    LADDERS PER SHEET                .\\n', '  #  RESIDUE AA STRUCTURE BP1 BP2  ACC     N-H-->O    O-->H-N    N-H-->O    O-->H-N    TCO  KAPPA ALPHA  PHI   PSI    X-CA   Y-CA   Z-CA\\n', '    1    1 A M              0   0  107      0, 0.0   139,-2.9     0, 0.0     2,-0.4   0.000 360.0 360.0 360.0 131.4   18.4   47.6   33.2\\n', '    2    2 A L  E     +A  139   0A  87    137,-0.2     2,-0.3   135,-0.0   137,-0.2  -0.771 360.0 176.8 -97.7 136.0   17.3   50.6   35.3\\n', '    3    3 A Y  E     -A  138   0A  28    135,-2.6   135,-2.5    -2,-0.4     2,-0.6  -0.950  26.3-140.5-134.0 153.4   13.8   51.2   36.4\\n', '    4    4 A H  E     -A  137   0A 101     -2,-0.3     2,-0.7   133,-0.2   133,-0.2  -0.942  22.0-163.5-116.2 105.4   12.1   53.8   38.7\\n', '    5    5 A L  E     -A  136   0A   0    131,-2.0   131,-1.9    -2,-0.6     2,-0.2  -0.818   6.4-153.9 -95.7 110.4    9.5   52.0   40.7\\n', '    6    6 A F    >   -     0   0  102     -2,-0.7     3,-1.0   129,-0.2   119,-0.3  -0.521  29.2 -92.1 -84.1 151.3    7.0   54.4   42.3\\n', '    7    7 A V  T 3  S+     0   0   41      1,-0.3   119,-0.2    -2,-0.2     3,-0.1  -0.298 111.0  14.9 -63.0 141.8    5.1   53.6   45.4\\n', '    8    8 A N  T 3  S+     0   0   93    117,-1.9    -1,-0.3     1,-0.3     2,-0.2   0.850 121.6  77.7  59.6  36.6    1.6   52.0   44.9\\n', '    9    9 A N  E <   -F  125   0B  93    116,-1.3   116,-2.7    -3,-1.0     2,-0.3  -0.743  66.7-141.6-149.6-165.8    2.7   51.3   41.3\\n', '   10   10 A Q  E     -F  124   0B  91    114,-0.3     2,-0.4    -2,-0.2   114,-0.2  -0.957  10.8-122.8-159.1 172.5    4.9   48.8   39.4\\n', '   11   11 A V  E     -F  123   0B  17    112,-2.1   112,-1.6    -2,-0.3     2,-0.6  -0.996  19.1-133.7-129.0 133.3    7.3   48.5   36.6\\n', '   12   12 A K  E     -F  122   0B 119     -2,-0.4   110,-0.2   110,-0.2     2,-0.1  -0.775  32.0-126.0 -84.0 121.6    6.9   46.2   33.6\\n', '   13   13 A L  E     -     0   0B   3    108,-2.5     3,-0.1    -2,-0.6    -1,-0.0  -0.436   2.3-136.0 -72.2 142.2   10.2   44.4   33.1\\n', '   14   14 A Q  E    S+     0   0B 124     -2,-0.1     2,-0.3     1,-0.1    -1,-0.1   0.866  94.6  10.7 -61.5 -39.9   11.9   44.6   29.6\\n', '   15   15 A N  E    S-     0   0B 123    105,-0.0    -1,-0.1     3,-0.0     2,-0.1  -0.922 103.8 -82.5-135.0 155.8   12.7   40.9   29.9\\n', '   16   16 A D  E     -     0   0B  65     -2,-0.3     2,-0.8     1,-0.1   105,-0.2  -0.343  35.2-120.2 -69.0 141.8   11.4   38.4   32.4\\n', '   17   17 A F  E     +F  120   0B   2    103,-2.7   103,-2.0    -2,-0.1    -1,-0.1  -0.718  59.0 144.5 -78.1 112.7   13.1   38.1   35.8\\n', '   18   18 A K    >   -     0   0   84     -2,-0.8     3,-3.0   101,-0.2    77,-0.3  -0.756  49.8 -25.8-140.9-174.6   14.2   34.5   35.8\\n', '   19   19 A P  T 3  S+     0   0   58      0, 0.0    77,-0.2     0, 0.0    -1,-0.2  -0.073 129.8  16.8 -42.5 131.6   17.0   32.1   36.9\\n', '   20   20 A E  T 3  S+     0   0   32     75,-3.1   124,-0.2     1,-0.3    76,-0.1   0.061  95.1 134.9  89.2 -19.8   20.3   34.0   37.3\\n', '   21   21 A S    <   -     0   0    3     -3,-3.0    74,-2.4   122,-0.1     2,-0.4  -0.267  38.3-161.4 -62.6 145.2   18.5   37.4   37.4\\n', '   22   22 A V  E     -BC  94 141A   1    119,-2.1   119,-3.4    72,-0.2     2,-0.5  -0.996   5.3-166.3-131.5 125.8   19.4   39.9   40.0\\n', '   23   23 A A  E     -BC  93 140A   0     70,-3.0    70,-2.6    -2,-0.4     2,-0.7  -0.961   3.0-165.2-116.6 117.5   17.3   42.8   41.1\\n', '   24   24 A A  E     -BC  92 139A  12    115,-3.3   115,-2.4    -2,-0.5     2,-0.5  -0.902   7.2-169.1-104.9 108.7   19.0   45.5   43.2\\n', '   25   25 A I  E     -BC  91 138A   4     66,-3.3    66,-2.7    -2,-0.7     2,-0.4  -0.865   7.4-178.0-100.0 125.3   16.4   47.8   44.8\\n', '   26   26 A R  E     -BC  90 137A 124    111,-3.3   111,-3.0    -2,-0.5     2,-0.3  -0.982  14.7-150.2-131.9 132.2   17.7   51.0   46.4\\n', '   27   27 A S        -     0   0    7     62,-2.6   109,-0.1    -2,-0.4     3,-0.0  -0.741  11.2-150.1 -97.6 144.2   15.9   53.7   48.4\\n', '   28   28 A S  S    S+     0   0   88    107,-0.4     2,-0.3    -2,-0.3    -1,-0.1   0.489  84.2  24.6 -86.7  -4.1   17.0   57.4   48.4\\n', '   29   29 A A        -     0   0   30     60,-0.1   104,-0.5    57,-0.0     2,-0.4  -0.961  61.9-142.5-152.9 167.9   15.5   57.6   51.9\\n', '   30   30 A F  B     -K  132   0C  54     -2,-0.3     2,-1.1   102,-0.2   102,-0.2  -0.938  11.1-160.6-138.2 112.4   14.5   55.8   55.0\\n', '   31   31 A N    >   -     0   0   68    100,-2.9     3,-3.0    -2,-0.4   101,-0.0  -0.763   3.8-168.9 -97.4  91.1   11.4   57.0   56.8\\n', '   32   32 A S  T 3  S+     0   0   86     -2,-1.1    -1,-0.2     1,-0.3    99,-0.1   0.761  88.0  52.6 -49.1 -31.2   11.8   55.5   60.3\\n', '   33   33 A K  T 3  S+     0   0  191     -3,-0.1    -1,-0.3    97,-0.1    98,-0.1   0.201  87.9 120.8 -92.5  16.0    8.2   56.4   61.1\\n', '   34   34 A G    <   -     0   0    8     -3,-3.0    96,-0.2     1,-0.1    97,-0.2  -0.293  53.5-134.7 -76.5 164.3    6.9   54.6   58.0\\n', '   35   35 A G        -     0   0   25     94,-3.3    96,-0.2     1,-0.2     2,-0.1  -0.059  51.3 -29.6-101.2-155.9    4.5   51.7   58.0\\n', '   36   36 A T        -     0   0   71     19,-0.1    92,-0.5    94,-0.1     2,-0.4  -0.379  61.9-166.1 -61.2 127.5    4.2   48.3   56.4\\n', '   37   37 A T  E     -GH  54 127B   0     17,-2.4    17,-3.1    90,-0.1     2,-0.4  -0.975   1.5-161.6-124.2 133.8    5.9   48.4   53.0\\n', '   38   38 A V  E     -GH  53 126B  18     88,-3.0    88,-2.6    -2,-0.4     2,-0.4  -0.907  15.3-173.9-123.5 145.2    5.5   45.7   50.3\\n', '   39   39 A F  E     -GH  52 125B   0     13,-1.6    13,-2.0    -2,-0.4     2,-0.4  -1.000  14.1-170.1-129.9 127.8    7.3   44.6   47.1\\n', '   40   40 A N  E     -GH  51 124B  13     84,-2.1    84,-3.2    -2,-0.4     2,-0.7  -0.980  17.4-148.6-124.1 132.5    5.8   41.9   44.9\\n', '   41   41 A F  E     -GH  50 123B   0      9,-2.5     8,-1.8    -2,-0.4     9,-1.2  -0.894  29.7-165.2 -97.2 116.7    7.4   40.1   42.0\\n', '   42   42 A L  E     -GH  48 122B  18     80,-2.8    79,-2.9    -2,-0.7    80,-2.2  -0.826  14.4-136.7-110.2 145.7    4.5   39.3   39.6\\n', '   43   43 A S    >   -     0   0    5      4,-1.7     3,-2.2    -2,-0.3    77,-0.2  -0.325  41.9 -94.4 -85.9 174.6    4.2   37.0   36.6\\n', '   44   44 A A  T 3  S+     0   0   86     75,-0.5    -1,-0.1     1,-0.3    76,-0.1   0.857 130.5  53.6 -58.7 -34.7    2.5   38.0   33.4\\n', '   45   45 A G  T 3  S-     0   0   38      2,-0.0    -1,-0.3     1,-0.0    23,-0.2   0.292 121.7-110.9 -84.0  10.7   -0.7   36.5   34.7\\n', '   46   46 A E    <   +     0   0  105     -3,-2.2     2,-0.3     1,-0.2    -2,-0.1   0.720  63.0 155.1  68.2  23.8   -0.3   38.6   37.9\\n', '   47   47 A N        -     0   0   24     20,-0.1    -4,-1.7     1,-0.1     2,-1.3  -0.614  49.5-120.6 -79.7 143.5    0.5   35.8   40.2\\n', '   48   48 A I  E     +GI  42  66B  13     18,-2.7    18,-2.0    -2,-0.3    -6,-0.2  -0.715  37.7 171.8 -88.8  95.4    2.4   36.9   43.2\\n', '   49   49 A L  E    S+     0   0B   3     -8,-1.8     2,-0.4    -2,-1.3    66,-0.4   0.841  73.0  16.1 -70.6 -38.5    5.6   34.8   42.9\\n', '   50   50 A L  E     -G   41   0B   0     -9,-1.2    -9,-2.5    -3,-0.2     2,-0.5  -0.943  67.5-172.3-144.8 118.6    7.4   36.6   45.8\\n', '   51   51 A H  E     -GI  40  64B  20     13,-2.5    13,-1.4    -2,-0.4     2,-0.5  -0.942   1.9-171.3-111.4 128.0    5.7   38.8   48.3\\n', '   52   52 A I  E     -GI  39  63B   0    -13,-2.0   -13,-1.6    -2,-0.5     2,-0.7  -0.937   2.0-173.4-124.9 106.2    7.9   40.8   50.7\\n', '   53   53 A S  E     -GI  38  62B   4      9,-3.4     9,-3.0    -2,-0.5     2,-0.8  -0.878   7.3-161.5-104.8 106.7    6.0   42.5   53.5\\n', '   54   54 A I  E     -GI  37  61B   0    -17,-3.1   -17,-2.4    -2,-0.7     7,-0.2  -0.794  11.5-177.6 -90.6 112.3    8.2   44.7   55.6\\n', '   55   55 A R  E  >> + I   0  60B  47      5,-3.1     5,-2.4    -2,-0.8     4,-0.8  -0.805  14.9 172.9-117.8  91.5    6.5   45.3   58.9\\n', '   56   56 A P  T >45S+     0   0   29      0, 0.0     3,-0.7     0, 0.0    -1,-0.2   0.918  82.5  54.9 -62.1 -43.1    8.3   47.7   61.3\\n', '   57   57 A G  T 345S+     0   0   86      1,-0.3    -2,-0.1     2,-0.1   -22,-0.0   0.772 116.8  37.4 -61.7 -25.7    5.4   47.8   63.8\\n', '   58   58 A E  T 345S-     0   0   93     -3,-0.5    -1,-0.3     2,-0.2     3,-0.1   0.463 107.1-129.6-101.9  -6.4    5.5   44.0   64.1\\n', '   59   59 A N  T <<5S+     0   0   78     -4,-0.8    21,-1.5    -3,-0.7     2,-0.3   0.894  70.7 106.7  57.0  43.6    9.3   43.9   63.8\\n', '   60   60 A V  E   < -IJ  55  79B  11     -5,-2.4    -5,-3.1    19,-0.3     2,-0.5  -0.980  60.7-145.8-148.8 158.9    9.1   41.3   61.1\\n', '   61   61 A I  E     -IJ  54  78B   0     17,-2.0    17,-1.1    -2,-0.3     2,-0.4  -0.995  25.0-155.7-125.9 119.3    9.5   40.8   57.3\\n', '   62   62 A V  E     -IJ  53  77B   1     -9,-3.0    -9,-3.4    -2,-0.5     2,-0.4  -0.853   7.3-165.7-105.6 133.4    7.1   38.3   55.8\\n', '   63   63 A F  E     +IJ  52  76B   0     13,-3.0    13,-1.9    -2,-0.4     2,-0.3  -0.899  26.1 133.9-112.9 140.9    7.7   36.4   52.6\\n', '   64   64 A N  E     -I   51   0B   2    -13,-1.4   -13,-2.5    -2,-0.4     2,-0.3  -0.930  31.9-142.9-165.1-174.2    4.9   34.5   50.8\\n', '   65   65 A S  E     +     0   0B   4     -2,-0.3     8,-0.9   -15,-0.2     2,-0.2  -0.945  15.9 167.8-155.8 169.1    3.3   33.7   47.4\\n', '   66   66 A R  E     -I   48   0B  36    -18,-2.0   -18,-2.7    -2,-0.3     6,-0.1  -0.827  37.3 -98.0 175.5 147.3   -0.1   33.2   45.9\\n', '   67   67 A L    >   -     0   0   92      4,-0.3     3,-1.8    -2,-0.2   -20,-0.1  -0.288  51.4 -94.8 -70.2 161.8   -1.9   33.0   42.5\\n', '   68   68 A K  T 3  S+     0   0   96      1,-0.3    -1,-0.1   -23,-0.2   -22,-0.1   0.853 127.1  17.1 -44.7 -46.6   -3.7   36.0   41.2\\n', '   69   69 A N  T 3  S+     0   0   58      2,-0.1    -1,-0.3   482,-0.0   483,-0.1   0.048 115.2  98.9-116.5  23.2   -7.0   34.8   42.6\\n', '   70   70 A G  S <  S-     0   0   10     -3,-1.8     2,-0.3     1,-0.1    -4,-0.0  -0.231  72.7 -95.1 -99.0-170.6   -5.5   32.3   45.0\\n', '   71   71 A A        -     0   0   21     -2,-0.1    -4,-0.3     1,-0.1   500,-0.2  -0.755  36.8-100.2-109.7 155.3   -4.6   32.1   48.7\\n', '   72   72 A W        -     0   0   65    498,-3.2   500,-0.3    -2,-0.3    -6,-0.2  -0.321  36.3-142.5 -65.6 154.5   -1.3   32.9   50.6\\n', '   73   73 A G        -     0   0   36     -8,-0.9    -1,-0.1     1,-0.3     2,-0.1  -0.085  50.4 -42.9-100.0-156.0    0.6   29.8   51.6\\n', '   74   74 A P        -     0   0   99      0, 0.0    -1,-0.3     0, 0.0     2,-0.2  -0.435  68.6-116.3 -70.9 151.9    2.5   29.4   54.8\\n', '   75   75 A E        -     0   0   84     -3,-0.1     2,-0.5   -11,-0.1   -11,-0.2  -0.567  16.8-153.4 -88.8 152.7    4.5   32.5   55.8\\n', '   76   76 A E  E     -J   63   0B  65    -13,-1.9   -13,-3.0    -2,-0.2     2,-0.3  -0.998  24.5-173.7-125.0 123.9    8.3   32.7   56.1\\n', '   77   77 A R  E     +J   62   0B 116     -2,-0.5   -15,-0.2   -15,-0.2   -17,-0.0  -0.823  20.7 178.7-124.3 161.2    9.5   35.3   58.5\\n', '   78   78 A I  E     -J   61   0B  31    -17,-1.1   -17,-2.0    -2,-0.3    -2,-0.0  -0.969  45.8 -85.8-152.1 149.8   12.6   37.0   59.8\\n', '   79   79 A P  E     -J   60   0B  78      0, 0.0   -19,-0.3     0, 0.0   -20,-0.1  -0.341  28.5-145.5 -61.5 140.2   13.1   39.8   62.4\\n', '   80   80 A Y    >   +     0   0   34    -21,-1.5     3,-0.6   -25,-0.2     4,-0.4   0.875  68.6  99.0 -72.7 -40.2   12.8   43.3   60.9\\n', '   81   81 A A  T 3  S+     0   0   46    -22,-0.3     0, 0.0     1,-0.2     0, 0.0  -0.076  87.3  11.4 -53.1 144.6   15.4   44.9   63.2\\n', '   82   82 A E  T 3  S+     0   0  171      2,-0.1    -1,-0.2     1,-0.1     3,-0.1   0.862  97.3 105.3  55.2  42.5   19.0   45.4   62.1\\n', '   83   83 A K    <   +     0   0   28     -3,-0.6    23,-2.5     1,-0.1    22,-1.1   0.715  67.0  54.3-114.8 -41.0   18.2   44.7   58.4\\n', '   84   84 A F        -     0   0   17     -4,-0.4    -1,-0.1    21,-0.2    -2,-0.1  -0.673  63.9-145.7-103.1 153.7   18.3   48.1   56.7\\n', '   85   85 A R        -     0   0  123     -2,-0.3     3,-0.4     3,-0.1     0, 0.0  -0.944  34.4-110.6-118.6 111.8   20.9   50.8   56.6\\n', '   86   86 A P  S    S+     0   0   76      0, 0.0   -57,-0.0     0, 0.0   -58,-0.0  -0.419  86.6  21.9 -76.8 159.9   19.7   54.4   56.5\\n', '   87   87 A P  S    S+     0   0   92      0, 0.0   -56,-0.0     0, 0.0    -2,-0.0  -0.978 135.9  27.2 -91.2   1.9   19.5   56.8   54.6\\n', '   88   88 A N  S    S-     0   0   92     -3,-0.4   -60,-0.1   -60,-0.1    -3,-0.1  -0.618  76.8-179.8-120.5  70.5   19.8   54.5   51.6\\n', '   89   89 A P        -     0   0    1      0, 0.0   -62,-2.6     0, 0.0     2,-0.3  -0.310  10.6-160.6 -67.9 156.1   18.4   51.1   52.6\\n', '   90   90 A S  E     -B   26   0A  24    -64,-0.2     2,-0.4     2,-0.0    15,-0.4  -0.951  20.5-156.5-140.4 162.1   18.4   48.4   49.9\\n', '   91   91 A I  E     -B   25   0A   3    -66,-2.7   -66,-3.3    -2,-0.3     2,-0.5  -0.989  11.5-166.6-136.7 123.7   16.8   45.0   48.9\\n', '   92   92 A T  E     -BD  24 103A  27     11,-3.3    11,-3.0    -2,-0.4     2,-0.5  -0.960   4.3-171.6-117.5 124.2   18.6   42.6   46.7\\n', '   93   93 A V  E     -BD  23 102A   0    -70,-2.6   -70,-3.0    -2,-0.5     2,-0.4  -0.959   1.5-172.9-117.1 127.6   16.6   39.7   45.2\\n', '   94   94 A I  E     -BD  22 101A  11      7,-2.9     7,-2.4    -2,-0.5     2,-0.8  -0.978  19.0-139.9-125.0 130.4   18.3   36.9   43.2\\n', '   95   95 A D  E     + D   0 100A   2    -74,-2.4   -75,-3.1    -2,-0.4     5,-0.2  -0.786  19.7 177.8 -89.1 110.5   16.6   34.1   41.3\\n', '   96   96 A H        -     0   0    7      3,-2.9     2,-1.7    -2,-0.8     4,-0.1   0.101  48.3-112.6 -98.9  21.4   18.4   30.8   41.9\\n', '   97   97 A G  S    S+     0   0   30      1,-0.2     3,-0.1     2,-0.1    -1,-0.1  -0.448 110.9  40.1  85.6 -67.6   16.0   28.7   39.9\\n', '   98   98 A D  S    S+     0   0   85     -2,-1.7    16,-2.0     1,-0.2     2,-0.3   0.517 130.0   5.6 -92.7  -5.5   14.5   26.6   42.7\\n', '   99   99 A R  E    S- E   0 113A  13     14,-0.3    -3,-2.9    16,-0.0     2,-0.3  -0.987  70.2-111.8-167.5 162.5   14.4   29.5   45.1\\n', '  100  100 A F  E     -DE  95 112A   5     12,-2.6    12,-2.6    -2,-0.3     2,-0.5  -0.759  19.7-148.7-101.8 151.5   14.9   33.2   45.6\\n', '  101  101 A Q  E     -DE  94 111A  10     -7,-2.4    -7,-2.9    -2,-0.3     2,-0.5  -0.987  14.2-166.6-118.2 126.8   17.6   34.7   47.7\\n', '  102  102 A I  E     -DE  93 110A   0      8,-3.0     8,-2.8    -2,-0.5     2,-0.4  -0.972   5.7-176.1-118.0 130.5   16.7   38.0   49.4\\n', '  103  103 A R  E     -D   92   0A  14    -11,-3.0   -11,-3.3    -2,-0.5     2,-0.1  -0.990   4.6-173.2-128.2 134.5   19.4   40.2   51.0\\n', '  104  104 A F        -     0   0    6      2,-0.5   -20,-0.2    -2,-0.4   -13,-0.2  -0.230  50.5 -78.4-104.4-161.4   18.9   43.4   52.9\\n', '  105  105 A D  S    S+     0   0   22    -22,-1.1   -21,-0.2   -15,-0.4     2,-0.2   0.936 110.1  19.9 -69.7 -46.3   21.7   45.8   54.1\\n', '  106  106 A Y  S    S+     0   0   96    -23,-2.5    -2,-0.5     2,-0.1   -21,-0.0  -0.666  99.4  34.3-117.3 174.5   22.7   43.7   57.1\\n', '  107  107 A G  S    S-     0   0   44     -2,-0.2    -3,-0.1    -4,-0.1     2,-0.1  -0.011  90.1 -47.0  75.9 178.1   22.4   40.1   58.3\\n', '  108  108 A T        -     0   0   93      1,-0.1    -4,-0.3    -5,-0.1    -2,-0.1  -0.344  66.5 -89.5 -82.7 165.3   22.3   36.9   56.4\\n', '  109  109 A S        -     0   0   11     -6,-0.1     2,-0.4    -2,-0.1    -6,-0.2  -0.281  30.9-131.6 -72.7 160.2   20.3   36.2   53.3\\n', '  110  110 A I  E     -E  102   0A  16     -8,-2.8    -8,-3.0     2,-0.0     2,-0.5  -0.936  17.2-154.9-111.6 134.5   16.8   34.8   53.3\\n', '  111  111 A Y  E     -E  101   0A  45     -2,-0.4     2,-0.5   -10,-0.2   -10,-0.2  -0.958   8.9-172.0-114.8 129.1   16.1   31.9   51.0\\n', '  112  112 A Y  E     -E  100   0A   0    -12,-2.6   -12,-2.6    -2,-0.5     3,-0.1  -0.934  20.8-133.5-125.1 108.5   12.5   31.3   49.7\\n', '  113  113 A N  E     -E   99   0A  98     -2,-0.5   -14,-0.3   -14,-0.3     2,-0.1  -0.311  33.8-107.4 -57.7 135.4   11.7   28.2   47.8\\n', '  114  114 A K        -     0   0   46    -16,-2.0     3,-0.1     1,-0.1    -1,-0.1  -0.401  34.4-178.5 -66.9 141.9    9.7   28.9   44.7\\n', '  115  115 A R  S    S+     0   0   86    -66,-0.4     2,-0.4     1,-0.3    -1,-0.1   0.540  71.4  48.2-114.9 -16.9    6.0   27.9   44.9\\n', '  116  116 A I  S    S-     0   0   25    -67,-0.2     2,-1.9     2,-0.1    -1,-0.3  -0.975  77.4-139.3-129.6 115.9    4.9   29.0   41.5\\n', '  117  117 A K        +     0   0  201     -2,-0.4     2,-0.3    -3,-0.1   -19,-0.0  -0.520  65.2 101.7 -79.4  81.7    7.0   27.9   38.4\\n', '  118  118 A E        -     0   0   74     -2,-1.9     2,-0.3   -69,-0.0    -2,-0.1  -0.957  67.8-116.1-151.7 160.7    6.8   31.1   36.4\\n', '  119  119 A N        -     0   0   59     -2,-0.3   -75,-0.5  -101,-0.1     2,-0.4  -0.677  36.9-104.0 -99.5 158.6    9.1   34.1   35.8\\n', '  120  120 A A  E     -F   17   0B   2   -103,-2.0  -103,-2.7    -2,-0.3   -77,-0.3  -0.674  29.3-179.5 -87.2 134.9    8.3   37.6   36.9\\n', '  121  121 A A  E    S+     0   0B   6    -79,-2.9  -108,-2.5    -2,-0.4     2,-0.3   0.531  71.5  17.5-104.9 -12.2    7.1   40.1   34.3\\n', '  122  122 A A  E     -FH  12  42B   6    -80,-2.2   -80,-2.8  -110,-0.2     2,-0.4  -0.984  59.3-140.5-156.0 162.5    6.7   43.0   36.8\\n', '  123  123 A I  E     -FH  11  41B   0   -112,-1.6  -112,-2.1    -2,-0.3     2,-0.3  -0.999  22.6-159.9-128.6 126.8    7.7   44.3   40.2\\n', '  124  124 A A  E     -FH  10  40B   5    -84,-3.2   -84,-2.1    -2,-0.4     2,-0.4  -0.796   5.2-159.4-109.7 154.6    5.1   46.1   42.3\\n', '  125  125 A Y  E     +FH   9  39B   0   -116,-2.7  -117,-1.9  -119,-0.3  -116,-1.3  -0.985  14.5 178.3-135.8 118.3    5.6   48.5   45.3\\n', '  126  126 A N  E     + H   0  38B  63    -88,-2.6   -88,-3.0    -2,-0.4     2,-0.3  -0.978  12.3 139.3-129.7 138.3    2.8   49.2   47.8\\n', '  127  127 A A  E     - H   0  37B   8     -2,-0.4     3,-0.4   -90,-0.2   -90,-0.1  -0.867  37.9-154.7-173.6 135.7    2.6   51.3   50.9\\n', '  128  128 A E  S    S+     0   0  195    -92,-0.5     2,-0.3     1,-0.3    -1,-0.1   0.909 109.0   8.7 -76.2 -46.5    0.2   53.7   52.7\\n', '  129  129 A N  S    S-     0   0   93    -93,-0.3   -94,-3.3   -98,-0.0    -1,-0.3  -0.704  97.3-143.8-135.1  78.9    3.2   55.3   54.4\\n', '  130  130 A S        -     0   0   14     -3,-0.4   -94,-0.1    -2,-0.3   -93,-0.1  -0.018  15.9-147.6 -51.7 144.5    6.3   53.9   52.8\\n', '  131  131 A L  S    S+     0   0    0    -96,-0.2  -100,-2.9   -97,-0.2     2,-0.2   0.676  86.4  48.2 -80.1 -21.4    9.5   53.2   54.7\\n', '  132  132 A F  B    S-K   30   0C   8   -102,-0.2  -102,-0.2     1,-0.2     4,-0.1  -0.584  94.1 -84.6-117.3 177.3   11.3   54.0   51.5\\n', '  133  133 A S        -     0   0   14   -104,-0.5    -1,-0.2    -2,-0.2    -2,-0.1  -0.213  46.3 -97.8 -76.3 168.3   11.4   56.7   48.8\\n', '  134  134 A S  S    S+     0   0   60   -128,-0.1  -128,-0.2     1,-0.1     2,-0.2  -0.966 108.3  24.4-135.1 115.9    9.2   57.0   45.8\\n', '  135  135 A P  S    S-     0   0   34      0, 0.0  -107,-0.4     0, 0.0     2,-0.4   0.592  82.6-157.3 -84.0-176.5   10.4   56.0   43.5\\n', '  136  136 A V  E     -A    5   0A   1   -131,-1.9  -131,-2.0    -2,-0.2     2,-0.4  -0.883  27.2-113.0-112.5 148.7   13.0   53.5   44.5\\n', '  137  137 A T  E     -AC   4  26A  39   -111,-3.0  -111,-3.3    -2,-0.4     2,-0.4  -0.705  33.3-168.5 -84.8 134.9   15.9   52.8   42.1\\n', '  138  138 A V  E     -AC   3  25A   0   -135,-2.5  -135,-2.6    -2,-0.4     2,-0.6  -0.969  10.4-173.5-130.8 119.5   15.9   49.3   40.6\\n', '  139  139 A D  E     -AC   2  24A  15   -115,-2.4  -115,-3.3    -2,-0.4     2,-0.5  -0.951  10.3-164.6-110.1 122.2   18.7   47.7   38.6\\n', '  140  140 A V  E     - C   0  23A  17   -139,-2.9     2,-0.3    -2,-0.6  -117,-0.2  -0.937   2.7-165.2-112.0 126.6   17.9   44.3   37.1\\n', '  141  141 A H  E     - C   0  22A  16   -119,-3.4  -119,-2.1    -2,-0.5     3,-0.1  -0.832  23.6-145.3-109.0 146.3   20.7   42.1   35.8\\n', '  142  142 A G  S    S+     0   0   52     -2,-0.3     2,-0.3     1,-0.2    -1,-0.1   0.748  92.0   1.9 -77.5 -23.6   20.3   39.1   33.6\\n', '  143  143 A L  S    S-     0   0   85   -121,-0.1    -1,-0.2  -122,-0.1  -121,-0.1  -0.970 101.3 -70.8-155.3 160.4   23.2   37.6   35.5\\n', '  144  144 A L  P>    -     0   0   40     -2,-0.3     3,-0.1  -124,-0.2  -123,-0.0  -0.359  58.4-133.0 -56.8 126.3   25.5   38.5   38.4\\n', '  145  145 A P  PP    -     0   0    9      0, 0.0    -1,-0.1     0, 0.0    -4,-0.0  -0.271  34.5 -72.8 -78.7 169.1   27.7   41.3   37.1\\n', '  146  146 A P  PP    -     0   0  107      0, 0.0   155,-0.1     0, 0.0    29,-0.0  -0.333  57.7-126.6 -59.1 144.7   31.5   41.5   37.6\\n', '  147  147 A L  PP    -     0   0   39    151,-0.1   151,-0.1    -3,-0.1   143,-0.0  -0.772   8.5-128.0-103.9 139.7   32.3   42.4   41.2\\n', '  148  148 A P  P<    -     0   0   44      0, 0.0    -1,-0.0     0, 0.0   142,-0.0  -0.400  37.1-101.9 -75.5 159.0   34.5   45.3   42.5\\n', '  149  149 A P              0   0   90      0, 0.0   142,-0.1     0, 0.0    -2,-0.0   0.722 360.0 360.0 -52.5 -21.9   37.3   44.4   45.0\\n', '  150  150 A A              0   0  108    147,-0.1   145,-0.0   142,-0.1   142,-0.0  -0.965 360.0 360.0-170.1 360.0   34.9   45.7   47.7\\n', '  151        !*             0   0    0      0, 0.0     0, 0.0     0, 0.0     0, 0.0   0.000 360.0 360.0 360.0 360.0    0.0    0.0    0.0\\n', '  152    1 B M              0   0  109    139,-0.6   139,-2.8   138,-0.1     2,-0.4   0.000 360.0 360.0 360.0 123.2   40.3   40.7   50.5\\n', '  153    2 B L  E     +L  290   0D  90    137,-0.2     2,-0.3   135,-0.0   137,-0.2  -0.757 360.0 175.9 -95.3 133.9   42.7   40.0   47.7\\n', '  154    3 B Y  E     -L  289   0D  31    135,-2.8   135,-2.6    -2,-0.4     2,-0.6  -0.932  26.1-141.0-130.2 155.3   44.9   36.8   47.5\\n', '  155    4 B H  E     -L  288   0D 101     -2,-0.3     2,-0.7   133,-0.2   133,-0.2  -0.946  21.3-161.9-117.8 103.3   47.2   35.5   44.8\\n', '  156    5 B L  E     -L  287   0D   0    131,-2.1   131,-2.3    -2,-0.6     2,-0.2  -0.810   7.5-153.1 -92.1 111.8   46.7   31.8   44.6\\n', '  157    6 B F    >   -     0   0  104     -2,-0.7     3,-1.0   129,-0.3   119,-0.3  -0.512  28.1 -92.7 -84.3 151.1   49.6   30.0   42.9\\n', '  158    7 B V  T 3  S+     0   0   39      1,-0.3   119,-0.2    -2,-0.2     3,-0.1  -0.352 110.8  13.6 -64.5 139.9   49.3   26.7   41.0\\n', '  159    8 B N  T 3  S+     0   0   89    117,-2.0    -1,-0.3     1,-0.3     2,-0.2   0.874 122.0  77.8  60.3  39.0   50.0   23.6   43.1\\n', '  160    9 B N  E <   -Q  276   0E  94    116,-1.3   116,-2.5    -3,-1.0    -1,-0.3  -0.762  66.9-141.1-151.1-165.7   49.8   25.8   46.2\\n', '  161   10 B Q  E     -Q  275   0E  86    114,-0.2     2,-0.4    -2,-0.2    -4,-0.2  -0.950   8.9-124.5-159.0 174.4   47.1   27.4   48.3\\n', '  162   11 B V  E     -Q  274   0E  16    112,-2.2   112,-1.9    -2,-0.3     2,-0.5  -0.999  18.5-134.7-133.9 131.7   46.2   30.5   50.2\\n', '  163   12 B K  E     -Q  273   0E 139     -2,-0.4     2,-0.3   110,-0.2   110,-0.3  -0.751  28.5-126.5 -85.7 125.7   45.1   30.6   53.9\\n', '  164   13 B L  E     -     0   0E   3    108,-2.7     4,-0.1    -2,-0.5     3,-0.0  -0.543   2.2-144.5 -76.3 134.3   42.1   32.8   54.3\\n', '  165   14 B Q  E    S+     0   0E 144     -2,-0.3     2,-0.3     1,-0.1    -1,-0.1   0.803  91.0  28.8 -65.1 -28.3   42.3   35.6   56.8\\n', '  166   15 B N  E    S-     0   0E 118      3,-0.0    -1,-0.1   105,-0.0   105,-0.0  -0.972 100.9 -88.8-134.2 149.7   38.7   35.0   57.6\\n', '  167   16 B D  E     -     0   0E  84     -2,-0.3     2,-1.0     1,-0.1   105,-0.2  -0.240  34.2-128.9 -55.3 134.9   36.5   31.8   57.4\\n', '  168   17 B F  E     +Q  271   0E   0    103,-3.3   103,-2.2     4,-0.1    -1,-0.1  -0.823  57.6 141.4 -86.7 108.9   34.8   31.3   54.0\\n', '  169   18 B K    >   -     0   0   96     -2,-1.0     3,-3.0     1,-0.2    77,-0.3  -0.730  51.7 -27.8-139.2-174.9   31.2   30.9   55.3\\n', '  170   19 B P  T 3  S+     0   0   45      0, 0.0    77,-0.2     0, 0.0    -1,-0.2  -0.091 129.7  19.5 -44.1 134.1   27.6   31.7   54.5\\n', '  171   20 B E  T 3  S+     0   0   18     75,-3.1   124,-0.1     1,-0.3    76,-0.1   0.161  96.1 132.8  87.2 -14.9   27.3   34.9   52.5\\n', '  172   21 B S    <   -     0   0    4     -3,-3.0    74,-2.6    74,-0.2     2,-0.4  -0.385  38.8-162.4 -69.7 145.9   31.0   34.6   51.5\\n', '  173   22 B V  E     -MN 245 292D   1    119,-2.1   119,-3.0    72,-0.2     2,-0.6  -0.996   8.2-163.6-132.7 130.5   32.0   35.1   47.9\\n', '  174   23 B A  E     -MN 244 291D   0     70,-3.2    70,-2.5    -2,-0.4     2,-0.8  -0.958   4.4-166.7-116.2 114.8   35.2   34.0   46.2\\n', '  175   24 B A  E     -MN 243 290D  10    115,-3.2   115,-2.4    -2,-0.6     2,-0.5  -0.875   9.0-169.4-103.3 103.1   35.8   35.7   42.8\\n', '  176   25 B I  E     -MN 242 289D   4     66,-3.4    66,-2.5    -2,-0.8     2,-0.4  -0.815   6.7-178.8 -95.3 128.9   38.6   33.8   41.1\\n', '  177   26 B R  E     -MN 241 288D 125    111,-3.5   111,-3.1    -2,-0.5     2,-0.3  -0.991  13.7-152.3-133.3 135.8   40.1   35.4   38.0\\n', '  178   27 B S        -     0   0    6     62,-2.5   109,-0.1    -2,-0.4   107,-0.0  -0.821  14.3-145.5-105.6 143.8   42.9   34.2   35.6\\n', '  179   28 B S  S    S+     0   0   88    107,-0.4     2,-0.3    -2,-0.3    60,-0.1   0.399  85.2  17.1 -84.7   2.5   45.2   36.4   33.6\\n', '  180   29 B A        -     0   0   32     60,-0.1   104,-0.6    57,-0.0     2,-0.5  -0.989  62.9-134.1-164.1 165.2   45.1   33.9   30.8\\n', '  181   30 B F  B     -V  283   0F  48     -2,-0.3     2,-1.2   102,-0.2   102,-0.2  -0.963  18.6-150.7-130.0 110.6   43.4   30.8   29.3\\n', '  182   31 B N    >   -     0   0   72    100,-3.1     3,-2.3    -2,-0.5    -2,-0.0  -0.728   8.2-160.4 -86.7  98.6   45.8   28.0   28.3\\n', '  183   32 B S  T 3  S+     0   0   85     -2,-1.2    -1,-0.2     1,-0.3    99,-0.1   0.839  88.7  42.6 -43.3 -50.5   43.9   26.4   25.4\\n', '  184   33 B K  T 3  S+     0   0  181     -3,-0.1    -1,-0.3    97,-0.1    98,-0.1   0.039  88.4 134.9 -91.6  27.1   45.8   23.1   25.6\\n', '  185   34 B G    <   -     0   0    7     -3,-2.3    96,-0.2    97,-0.1    97,-0.2  -0.302  43.7-138.5 -73.9 159.8   45.8   23.0   29.4\\n', '  186   35 B G        -     0   0   24     94,-3.3    96,-0.3     1,-0.2    -1,-0.1  -0.094  51.0 -28.2-101.5-158.3   44.9   19.8   31.3\\n', '  187   36 B T        -     0   0   70     19,-0.1    92,-0.5    94,-0.1    93,-0.4  -0.324  61.8-168.8 -58.5 128.2   42.9   18.9   34.3\\n', '  188   37 B T  E     -RS 205 278E   0     17,-2.5    17,-3.1    90,-0.1     2,-0.4  -0.985   0.3-162.4-125.5 133.6   42.8   21.8   36.7\\n', '  189   38 B V  E     -RS 204 277E  16     88,-3.2    88,-2.5    -2,-0.4     2,-0.4  -0.917  13.3-173.1-123.5 143.8   41.5   21.7   40.3\\n', '  190   39 B F  E     -RS 203 276E   0     13,-2.0    13,-2.1    -2,-0.4     2,-0.4  -1.000  13.3-170.0-130.9 127.8   40.4   24.1   42.9\\n', '  191   40 B N  E     -RS 202 275E  11     84,-2.5    84,-3.2    -2,-0.4     2,-0.7  -0.970  18.1-150.7-123.9 136.5   39.6   23.0   46.5\\n', '  192   41 B F  E     -RS 201 274E   0      9,-2.5     8,-1.5    -2,-0.4     9,-1.4  -0.934  30.1-165.1-102.9 115.4   38.0   24.8   49.3\\n', '  193   42 B L  E     -RS 199 273E  17     80,-3.1    79,-3.4    -2,-0.7    80,-1.8  -0.776  17.2-130.4-107.1 148.1   39.5   23.3   52.5\\n', '  194   43 B S    >   -     0   0    5      4,-2.1     3,-2.1    -2,-0.3    77,-0.2  -0.382  39.0-102.6 -82.9 169.8   38.5   23.5   56.1\\n', '  195   44 B A  T 3  S+     0   0   88     75,-0.5    -1,-0.1     1,-0.3    76,-0.1   0.835 127.9  57.6 -63.1 -28.1   41.1   24.4   58.7\\n', '  196   45 B G  T 3  S-     0   0   41      2,-0.1    -1,-0.3     1,-0.0    23,-0.2   0.340 121.7-112.2 -82.6   7.6   41.2   20.6   59.4\\n', '  197   46 B E    <   +     0   0  105     -3,-2.1     2,-0.3     1,-0.2    -2,-0.1   0.768  65.0 149.9  67.8  27.5   42.0   20.0   55.8\\n', '  198   47 B N        -     0   0   23     20,-0.1    -4,-2.1    22,-0.1     2,-1.2  -0.694  53.1-120.2 -84.8 145.2   38.7   18.4   54.9\\n', '  199   48 B I  E     +RT 193 217E   7     18,-3.4    18,-2.0    -2,-0.3     3,-0.2  -0.747  36.7 173.8 -91.1  96.8   37.8   19.0   51.2\\n', '  200   49 B L  E    S+     0   0E   5     -8,-1.5    66,-0.4    -2,-1.2     2,-0.4   0.904  73.9  13.9 -68.2 -42.3   34.5   20.8   51.5\\n', '  201   50 B L  E     -R  192   0E   0     -9,-1.4    -9,-2.5    -3,-0.2     2,-0.5  -0.956  67.7-171.6-142.3 116.6   34.3   21.6   47.8\\n', '  202   51 B H  E     -RT 191 215E  19     13,-2.9    13,-1.6    -2,-0.4     2,-0.5  -0.933   3.0-170.7-109.2 128.3   36.4   19.9   45.1\\n', '  203   52 B I  E     -RT 190 214E   0    -13,-2.1   -13,-2.0    -2,-0.5     2,-0.7  -0.951   3.2-176.8-124.0 107.5   36.2   21.3   41.6\\n', '  204   53 B S  E     -RT 189 213E   4      9,-3.0     9,-2.8    -2,-0.5     2,-0.7  -0.912   7.5-164.1-110.2 107.7   38.0   19.2   39.0\\n', '  205   54 B I  E     -RT 188 212E   0    -17,-3.1   -17,-2.5    -2,-0.7     7,-0.2  -0.827   9.3-172.1 -92.7 114.5   37.9   20.8   35.5\\n', '  206   55 B R  E  >> + T   0 211E  50      5,-2.8     5,-1.7    -2,-0.7     4,-1.2  -0.864  16.5 179.8-117.2 101.3   38.8   18.1   32.9\\n', '  207   56 B P  T  45S+     0   0   34      0, 0.0    -1,-0.2     0, 0.0   -21,-0.1   0.926  84.8  50.3 -61.6 -47.1   39.3   19.2   29.3\\n', '  208   57 B G  T  45S+     0   0   79      1,-0.2    -2,-0.1     2,-0.1   -22,-0.0   0.957 119.1  34.9 -57.4 -54.9   40.2   15.7   28.0\\n', '  209   58 B E  T  45S-     0   0   99      2,-0.1    -1,-0.2     1,-0.0     3,-0.1   0.658 105.4-135.2 -74.3 -15.1   37.2   14.0   29.6\\n', '  210   59 B N  T  <5 +     0   0   78     -4,-1.2    21,-1.4     1,-0.2    22,-0.4   0.927  64.6 106.6  58.1  50.5   35.2   17.2   28.9\\n', '  211   60 B V  E   < -TU 206 230E  20     -5,-1.7    -5,-2.8    19,-0.3     2,-0.5  -0.945  62.4-136.6-147.0 166.6   33.5   17.3   32.3\\n', '  212   61 B I  E     -TU 205 229E   0     17,-2.4    17,-1.4    -2,-0.3     2,-0.5  -0.991  23.3-156.7-129.1 118.7   33.7   19.2   35.5\\n', '  213   62 B V  E     -TU 204 228E   1     -9,-2.8    -9,-3.0    -2,-0.5     2,-0.4  -0.866   6.5-167.5-104.2 129.0   33.5   17.0   38.6\\n', '  214   63 B F  E     +TU 203 227E   2     13,-2.4    13,-2.1    -2,-0.5     2,-0.3  -0.932  24.9 140.9-111.9 136.0   32.4   18.4   42.0\\n', '  215   64 B N  E     -T  202   0E   2    -13,-1.6   -13,-2.9    -2,-0.4     2,-0.3  -0.932  34.0-139.5-159.4 179.0   32.9   16.3   45.1\\n', '  216   65 B S  E     +     0   0E   5     -2,-0.3     8,-1.8   -15,-0.2     2,-0.3  -0.926  18.1 176.2-143.3 162.6   33.9   16.4   48.8\\n', '  217   66 B R  E     -T  199   0E  86    -18,-2.0   -18,-3.4    -2,-0.3     6,-0.1  -0.917  30.5-111.6-171.1 142.8   36.1   14.3   51.1\\n', '  218   67 B L    >   -     0   0   92      4,-0.5     3,-1.9    -2,-0.3   -20,-0.1  -0.411  48.4 -95.2 -75.0 157.9   37.4   14.1   54.6\\n', '  219   68 B K  T 3  S+     0   0  101      1,-0.3    -1,-0.1   -23,-0.2   -22,-0.1   0.858 122.3  14.6 -38.4 -58.2   41.1   14.7   55.2\\n', '  220   69 B N  T 3  S+     0   0  154      2,-0.0    -1,-0.3     0, 0.0     2,-0.1   0.040 113.5 100.2-111.0  26.4   42.0   10.9   55.2\\n', '  221   70 B G  S <  S-     0   0   38     -3,-1.9     2,-0.2     1,-0.1     3,-0.1  -0.375  74.3 -97.0-101.7-177.9   38.7    9.7   53.8\\n', '  222   71 B A        -     0   0   89     -2,-0.1    -4,-0.5     1,-0.1    -1,-0.1  -0.583  46.3 -87.8 -97.4 162.2   37.6    8.6   50.3\\n', '  223   72 B W        -     0   0  108     -2,-0.2    -6,-0.2    -6,-0.1    -1,-0.1  -0.376  42.3-134.1 -66.6 146.9   35.8   10.7   47.7\\n', '  224   73 B G        -     0   0   32     -8,-1.8     2,-0.1     1,-0.2    -1,-0.1  -0.018  50.5 -47.6 -86.2-163.8   32.0   10.8   48.0\\n', '  225   74 B P        -     0   0   99      0, 0.0    -1,-0.2     0, 0.0     2,-0.2  -0.446  67.7-116.4 -68.3 142.3   29.6   10.4   45.0\\n', '  226   75 B E        -     0   0   93     -3,-0.1     2,-0.5    -2,-0.1   -11,-0.2  -0.509  18.0-157.8 -82.1 147.8   30.6   12.6   42.1\\n', '  227   76 B E  E     +U  214   0E  75    -13,-2.1   -13,-2.4    -2,-0.2     2,-0.3  -0.950  28.2 177.0-122.5 103.5   28.4   15.4   40.8\\n', '  228   77 B R  E     +U  213   0E 105     -2,-0.5   -15,-0.2   -15,-0.2   -17,-0.0  -0.760  19.3 178.7-115.0 160.4   29.4   16.1   37.2\\n', '  229   78 B I  E     -U  212   0E  45    -17,-1.4   -17,-2.4    -2,-0.3     2,-0.1  -0.945  44.3 -84.4-145.5 160.4   28.4   18.2   34.3\\n', '  230   79 B P  E     -U  211   0E  81      0, 0.0   -19,-0.3     0, 0.0   -20,-0.1  -0.456  28.6-140.4 -70.4 146.0   29.8   18.6   30.8\\n', '  231   80 B Y        +     0   0   23    -21,-1.4     3,-0.5   -25,-0.2   -20,-0.2   0.874  40.0 161.9 -70.8 -41.1   32.8   21.1   30.6\\n', '  232   81 B A  S    S-     0   0   58    -22,-0.4   -21,-0.0     1,-0.2    -1,-0.0   0.650  72.1 -51.8  21.5  64.0   31.5   22.5   27.3\\n', '  233   82 B E  S    S+     0   0  163      1,-0.1    -1,-0.2    23,-0.0    25,-0.1   0.944  74.8 161.3  51.1  64.7   33.4   25.8   26.9\\n', '  234   83 B K        +     0   0   19     -3,-0.5    23,-1.7    22,-0.1    22,-1.4   0.652  50.0  77.8 -89.7 -15.4   32.6   27.4   30.3\\n', '  235   84 B F        -     0   0    7     21,-0.2     7,-0.1    20,-0.1   -54,-0.0  -0.799  65.5-145.0-108.6 139.9   35.5   29.9   30.2\\n', '  236   85 B R    >   -     0   0  114     -2,-0.4     3,-0.5     3,-0.1    -2,-0.0  -0.832  40.2-108.8 -97.1 107.2   35.9   33.2   28.4\\n', '  237   86 B P  T 3  S+     0   0   77      0, 0.0   -58,-0.0     0, 0.0   -57,-0.0  -0.372  87.3  22.4 -70.8 158.1   39.5   33.8   27.2\\n', '  238   87 B P  T 3  S+     0   0   89      0, 0.0    -2,-0.0     0, 0.0   -56,-0.0  -0.991 134.3  25.6 -93.2  -0.9   41.7   35.6   27.9\\n', '  239   88 B N  S <  S-     0   0   95     -3,-0.5    -3,-0.1   -60,-0.1   -60,-0.1  -0.663  77.0-177.2-120.9  74.9   40.6   36.3   31.4\\n', '  240   89 B P        -     0   0    2      0, 0.0   -62,-2.5     0, 0.0     2,-0.3  -0.327   8.7-155.0 -70.9 155.0   38.3   33.4   32.5\\n', '  241   90 B S  E     -M  177   0D  26    -64,-0.2     2,-0.4     2,-0.0    15,-0.3  -0.943  19.5-165.2-135.9 157.0   36.6   33.4   35.9\\n', '  242   91 B I  E     -M  176   0D   3    -66,-2.5   -66,-3.4    -2,-0.3     2,-0.5  -0.988  12.6-169.2-137.7 122.7   35.2   31.2   38.6\\n', '  243   92 B T  E     -MO 175 254D  27     11,-2.8    11,-2.9    -2,-0.4     2,-0.5  -0.974   1.4-170.5-120.1 126.6   32.9   32.6   41.3\\n', '  244   93 B V  E     -MO 174 253D   0    -70,-2.5   -70,-3.2    -2,-0.5     2,-0.4  -0.968   2.4-170.1-118.7 130.0   32.0   30.5   44.3\\n', '  245   94 B I  E     -MO 173 252D  12      7,-2.9     7,-2.6    -2,-0.5     2,-0.8  -0.973  19.0-137.6-124.6 133.2   29.3   31.6   46.8\\n', '  246   95 B D  E     + O   0 251D   1    -74,-2.6   -75,-3.1    -2,-0.4     5,-0.2  -0.775  22.2 176.5 -89.4 110.0   28.5   30.1   50.2\\n', '  247   96 B H        -     0   0    4      3,-2.6     2,-1.7    -2,-0.8     4,-0.1   0.060  47.4-111.4-102.3  25.4   24.7   30.0   50.5\\n', '  248   97 B G  S    S+     0   0   32      1,-0.2     3,-0.1     2,-0.1    -1,-0.1  -0.470 110.2  34.9  85.3 -70.3   24.6   28.2   53.8\\n', '  249   98 B D  S    S+     0   0   81     -2,-1.7    16,-1.8     1,-0.2     2,-0.3   0.513 130.9  10.4 -96.1  -7.2   23.1   24.8   52.8\\n', '  250   99 B R  E    S- P   0 264D  13     14,-0.3    -3,-2.6    16,-0.0     2,-0.3  -0.985  70.8-114.0-162.3 164.0   24.9   24.8   49.4\\n', '  251  100 B F  E     -OP 246 263D   4     12,-2.8    12,-2.5    -2,-0.3     2,-0.5  -0.799  20.2-147.2-103.4 148.7   27.6   26.4   47.3\\n', '  252  101 B Q  E     -OP 245 262D  10     -7,-2.6    -7,-2.9    -2,-0.3     2,-0.5  -0.967  13.5-168.2-114.1 130.7   26.8   28.3   44.1\\n', '  253  102 B I  E     -OP 244 261D   0      8,-2.9     8,-3.1    -2,-0.5     2,-0.4  -0.990   4.6-175.9-122.8 125.0   29.5   28.0   41.4\\n', '  254  103 B R  E     -O  243   0D  33    -11,-2.9   -11,-2.8    -2,-0.5     2,-0.1  -0.978   2.5-175.7-122.8 133.3   29.4   30.3   38.3\\n', '  255  104 B F        -     0   0    5      2,-0.5   -20,-0.1    -2,-0.4   -13,-0.1  -0.242  49.5 -85.3-105.6-162.7   31.7   30.2   35.4\\n', '  256  105 B D  S    S+     0   0   17    -22,-1.4     2,-0.2   -15,-0.3   -21,-0.2   0.861 109.4  19.9 -77.9 -33.7   31.8   32.6   32.4\\n', '  257  106 B Y  S    S+     0   0   97    -23,-1.7    -2,-0.5     2,-0.1   -23,-0.0  -0.764  99.3  40.2-128.2 175.1   29.1   30.7   30.5\\n', '  258  107 B G  S    S-     0   0   47     -2,-0.2     2,-0.1    -4,-0.1     3,-0.1  -0.198  92.8 -52.4  78.1-177.0   26.4   28.1   31.2\\n', '  259  108 B T        -     0   0   98      1,-0.1    -4,-0.3    -5,-0.1    -2,-0.1  -0.422  65.4 -83.1 -91.1 168.5   24.1   28.2   34.2\\n', '  260  109 B S        -     0   0    7     -2,-0.1     2,-0.5    -6,-0.1    -6,-0.2  -0.385  42.4-136.8 -66.4 150.0   25.1   28.4   37.9\\n', '  261  110 B I  E     -P  253   0D  18     -8,-3.1    -8,-2.9    -3,-0.1     2,-0.5  -0.939  14.2-154.8-111.1 132.8   26.0   25.1   39.4\\n', '  262  111 B Y  E     -P  252   0D  50     -2,-0.5     2,-0.5   -10,-0.2   -10,-0.2  -0.924   9.6-173.7-112.3 132.9   24.6   24.3   42.9\\n', '  263  112 B Y  E     -P  251   0D   0    -12,-2.5   -12,-2.8    -2,-0.5     2,-0.1  -0.952  21.6-133.9-127.8 112.2   26.4   21.9   45.2\\n', '  264  113 B N  E     -P  250   0D  98     -2,-0.5   -14,-0.3   -14,-0.3   -15,-0.1  -0.410  34.0-106.1 -64.9 135.8   24.7   20.9   48.5\\n', '  265  114 B K        -     0   0   22    -16,-1.8     3,-0.1     1,-0.1    -1,-0.1  -0.354  32.5-176.1 -63.2 140.9   27.2   21.0   51.4\\n', '  266  115 B R  S    S+     0   0   74    -66,-0.4     2,-0.4     1,-0.2    -1,-0.1   0.528  71.2  47.7-112.2 -15.7   28.3   17.6   52.7\\n', '  267  116 B I  S    S-     0   0   26    -67,-0.2     2,-1.9     2,-0.1    -1,-0.2  -0.992  74.5-138.6-132.7 126.4   30.4   18.8   55.6\\n', '  268  117 B K        +     0   0  169     -2,-0.4     2,-0.3    -3,-0.1   -19,-0.0  -0.446  67.4 103.5 -83.9  66.2   29.3   21.4   58.1\\n', '  269  118 B E        -     0   0   71     -2,-1.9     2,-0.2   -75,-0.0    -2,-0.1  -0.944  68.2-117.9-138.8 158.9   32.6   23.4   58.4\\n', '  270  119 B N        -     0   0   56     -2,-0.3   -75,-0.5  -101,-0.0     2,-0.3  -0.656  35.6-106.2 -95.2 157.0   33.7   26.7   57.0\\n', '  271  120 B A  E     +Q  168   0E   1   -103,-2.2  -103,-3.3    -2,-0.2   -77,-0.3  -0.639  32.3 177.0 -84.8 140.6   36.6   26.9   54.5\\n', '  272  121 B A  E     +     0   0E   3    -79,-3.4  -108,-2.7    -2,-0.3     2,-0.3   0.421  68.3  26.0-118.5  -7.2   40.0   28.3   55.8\\n', '  273  122 B A  E     -QS 163 193E   8    -80,-1.8   -80,-3.1  -110,-0.3     2,-0.4  -0.994  60.0-145.0-155.8 154.1   42.0   27.8   52.7\\n', '  274  123 B I  E     -QS 162 192E   0   -112,-1.9  -112,-2.2    -2,-0.3     2,-0.3  -0.994  20.9-160.3-123.6 128.5   41.7   27.5   48.9\\n', '  275  124 B A  E     -QS 161 191E   6    -84,-3.2   -84,-2.5    -2,-0.4     2,-0.4  -0.829   6.1-161.0-111.6 152.3   44.0   25.2   47.0\\n', '  276  125 B Y  E     +QS 160 190E   0   -116,-2.5  -117,-2.0  -119,-0.3  -116,-1.3  -0.984  13.6 179.4-135.5 116.4   44.8   25.2   43.3\\n', '  277  126 B N  E     + S   0 189E  65    -88,-2.5   -88,-3.2    -2,-0.4     2,-0.2  -0.971  15.0 132.4-126.2 137.5   46.3   22.0   41.6\\n', '  278  127 B A  E     - S   0 188E   9     -2,-0.4     3,-0.3   -90,-0.2   -90,-0.1  -0.848  39.1-155.0-177.2 139.2   47.4   21.3   38.0\\n', '  279  128 B E  S    S+     0   0  182    -92,-0.5     2,-0.3     1,-0.3    -1,-0.1   0.823 108.1  14.5 -85.0 -37.0   50.3   19.8   36.1\\n', '  280  129 B N  S    S-     0   0   91    -93,-0.4   -94,-3.3   -98,-0.0    -1,-0.3  -0.735  96.4-146.2-137.8  82.4   49.4   21.9   33.0\\n', '  281  130 B S        -     0   0   15     -3,-0.3   -94,-0.1    -2,-0.3   -93,-0.1  -0.157  17.7-147.4 -60.0 143.7   47.0   24.5   34.2\\n', '  282  131 B L  S    S+     0   0    1    -96,-0.3  -100,-3.1   -97,-0.2     2,-0.2   0.692  87.4  46.2 -76.3 -21.7   44.1   25.9   32.1\\n', '  283  132 B F  B    S-V  181   0F   6   -102,-0.2  -102,-0.2     1,-0.1    -2,-0.1  -0.610  94.9 -82.3-119.7 177.3   44.7   29.1   34.0\\n', '  284  133 B S        -     0   0   14   -104,-0.6    -1,-0.1    -2,-0.2    -2,-0.1  -0.257  49.3 -96.9 -75.1 163.2   47.5   31.5   35.0\\n', '  285  134 B S  S    S+     0   0   61   -128,-0.1  -128,-0.2    -4,-0.1     2,-0.1  -0.987 107.8  23.1-127.4 116.9   49.7   31.1   38.1\\n', '  286  135 B P  S    S-     0   0   35      0, 0.0  -107,-0.4     0, 0.0     2,-0.4   0.634  84.0-151.8 -84.8-176.3   48.8   32.7   40.3\\n', '  287  136 B V  E     -L  156   0D   0   -131,-2.3  -131,-2.1    -2,-0.1     2,-0.4  -0.878  25.1-113.0-110.7 148.3   45.1   33.4   39.8\\n', '  288  137 B T  E     -LN 155 177D  39   -111,-3.1  -111,-3.5    -2,-0.4     2,-0.5  -0.709  35.1-169.8 -83.1 129.6   43.6   36.6   41.3\\n', '  289  138 B V  E     -LN 154 176D   0   -135,-2.6  -135,-2.8    -2,-0.4     2,-0.6  -0.964  10.9-172.4-128.4 116.9   41.2   35.8   44.1\\n', '  290  139 B D  E     -LN 153 175D  13   -115,-2.4  -115,-3.2    -2,-0.5     2,-0.5  -0.938  10.8-165.0-108.2 118.1   38.9   38.4   45.8\\n', '  291  140 B V  E     - N   0 174D  17   -139,-2.8  -139,-0.6    -2,-0.6     2,-0.3  -0.910   3.2-167.3-109.8 125.3   37.1   37.1   48.8\\n', '  292  141 B H  E     - N   0 173D  14   -119,-3.0  -119,-2.1    -2,-0.5     3,-0.1  -0.833  25.4-146.1-111.6 147.4   34.1   38.9   50.2\\n', '  293  142 B G  S    S+     0   0   50     -2,-0.3     2,-0.3     1,-0.2    -1,-0.1   0.691  92.8   4.5 -80.4 -20.5   32.3   38.4   53.6\\n', '  294  143 B L  S    S-     0   0   78   -121,-0.1    -1,-0.2  -122,-0.1  -121,-0.1  -0.973  99.7 -75.6-156.2 158.5   29.1   39.4   51.8\\n', '  295  144 B L  P>    -     0   0   25     -2,-0.3     3,-0.1  -124,-0.1  -123,-0.0  -0.431  57.0-130.9 -60.2 124.7   28.0   40.2   48.3\\n', '  296  145 B P  PP    -     0   0    7      0, 0.0    -1,-0.1     0, 0.0    -4,-0.0  -0.280  33.6 -76.5 -75.9 163.6   29.3   43.8   47.6\\n', '  297  146 B P  PP    -     0   0  105      0, 0.0  -147,-0.1     0, 0.0  -273,-0.0  -0.310  58.2-125.2 -56.8 144.4   27.2   46.6   46.2\\n', '  298  147 B L  PP    -     0   0   38     -3,-0.1  -151,-0.1  -151,-0.1  -274,-0.0  -0.755   6.7-123.4-103.7 142.4   26.8   46.1   42.4\\n', '  299  148 B P  P<    -     0   0   43      0, 0.0    -1,-0.0     0, 0.0  -160,-0.0  -0.414  35.9-108.5 -72.1 155.4   27.6   48.3   39.5\\n', '  300  149 B P              0   0   88      0, 0.0  -160,-0.1     0, 0.0  -276,-0.0   0.639 360.0 360.0 -60.5 -11.1   24.6   49.0   37.2\\n', '  301  150 B A              0   0  110   -155,-0.1  -162,-0.0  -160,-0.1  -160,-0.0   0.022 360.0 360.0-176.2 360.0   26.4   46.8   34.7\\n', '  302        !*             0   0    0      0, 0.0     0, 0.0     0, 0.0     0, 0.0   0.000 360.0 360.0 360.0 360.0    0.0    0.0    0.0\\n', '  303    1 C M              0   0  111    139,-0.8   139,-2.8   138,-0.1     2,-0.4   0.000 360.0 360.0 360.0 128.7  -34.4   24.0   29.5\\n', '  304    2 C L  E     +W  441   0G  89    137,-0.2     2,-0.3   135,-0.0   137,-0.2  -0.740 360.0 174.4 -95.6 137.3  -32.9   21.5   27.1\\n', '  305    3 C Y  E     -W  440   0G  32    135,-2.1   135,-2.4    -2,-0.4     2,-0.7  -0.964  26.8-140.2-136.2 152.0  -31.6   22.4   23.6\\n', '  306    4 C H  E     -W  439   0G 102     -2,-0.3     2,-0.8   133,-0.2   133,-0.2  -0.927  21.6-161.3-115.4 102.7  -29.7   20.4   21.0\\n', '  307    5 C L  E     -W  438   0G   0    131,-1.9   131,-2.0    -2,-0.7     2,-0.2  -0.776   6.6-154.0 -91.0 108.2  -27.1   22.7   19.5\\n', '  308    6 C F    >   -     0   0  106     -2,-0.8     3,-1.2   129,-0.3   119,-0.3  -0.492  29.1 -93.7 -80.9 149.5  -25.9   21.4   16.1\\n', '  309    7 C V  T 3  S+     0   0   32      1,-0.3   119,-0.2    -2,-0.2     3,-0.1  -0.346 110.5  14.7 -62.9 139.9  -22.4   22.3   14.7\\n', '  310    8 C N  T 3  S+     0   0   90    117,-2.3    -1,-0.3     1,-0.3     2,-0.2   0.787 120.4  77.2  62.3  32.2  -22.4   25.3   12.4\\n', '  311    9 C N  E <   -B  427   0H  98     -3,-1.2   116,-2.1   116,-1.0    -1,-0.3  -0.726  67.8-138.6-145.8-167.9  -25.9   26.3   13.6\\n', '  312   10 C Q  E     -B  426   0H  89    114,-0.2     2,-0.4    -2,-0.2   114,-0.2  -0.938   6.9-129.5-154.4 172.6  -27.4   27.9   16.7\\n', '  313   11 C V  E     -B  425   0H  18    112,-2.1   112,-2.0    -2,-0.3     2,-0.6  -0.999  19.2-135.2-131.8 133.9  -30.2   27.7   19.2\\n', '  314   12 C K  E     -B  424   0H 168     -2,-0.4   110,-0.3   110,-0.2     2,-0.2  -0.806  27.3-125.8 -90.9 124.4  -32.4   30.6   20.2\\n', '  315   13 C L  E     -     0   0H  10    108,-2.8     4,-0.1    -2,-0.6     3,-0.0  -0.448   0.7-142.2 -72.4 137.5  -33.0   30.7   23.9\\n', '  316   14 C Q  E    S+     0   0H 140     -2,-0.2     2,-0.3     1,-0.1    -1,-0.1   0.776  92.2  28.0 -67.8 -25.5  -36.6   30.8   25.2\\n', '  317   15 C N  E    S-     0   0H 109    106,-0.0    -1,-0.1     3,-0.0    -2,-0.0  -0.955  99.4 -91.5-134.6 153.2  -35.3   33.2   27.9\\n', '  318   16 C D  E     -     0   0H  82     -2,-0.3     2,-1.0     1,-0.1   105,-0.2  -0.354  34.3-125.9 -63.2 141.3  -32.4   35.7   28.0\\n', '  319   17 C F  E     +B  422   0H   1    103,-3.0   103,-1.3     4,-0.1    -1,-0.1  -0.829  57.1 143.8 -89.0 106.8  -29.2   34.3   29.4\\n', '  320   18 C K    >   -     0   0   87     -2,-1.0     3,-2.7     1,-0.2    77,-0.4  -0.631  51.4 -29.3-133.1-171.0  -28.5   36.9   32.1\\n', '  321   19 C P  T 3  S+     0   0   55      0, 0.0    77,-0.2     0, 0.0    -1,-0.2  -0.174 129.6  18.3 -48.8 136.4  -27.1   37.1   35.7\\n', '  322   20 C E  T 3  S+     0   0   22     75,-3.1   124,-0.3     1,-0.3    76,-0.1   0.125  95.6 132.4  86.3 -17.4  -27.6   33.9   37.6\\n', '  323   21 C S    <   -     0   0    3     -3,-2.7    74,-2.5   122,-0.1     2,-0.4  -0.353  39.0-162.4 -66.9 145.7  -28.2   32.0   34.4\\n', '  324   22 C V  E     -XY 396 443G   1    119,-2.0   119,-3.2    72,-0.2     2,-0.5  -0.997   8.5-167.8-134.2 131.3  -26.3   28.7   33.9\\n', '  325   23 C A  E     -XY 395 442G   0     70,-3.0    70,-2.5    -2,-0.4     2,-0.7  -0.978   2.9-166.1-121.5 122.1  -25.6   26.8   30.7\\n', '  326   24 C A  E     -XY 394 441G  11    115,-2.9   115,-2.3    -2,-0.5     2,-0.5  -0.915   6.9-167.9-109.2 108.3  -24.2   23.3   30.8\\n', '  327   25 C I  E     -XY 393 440G   3     66,-3.6    66,-2.7    -2,-0.7     2,-0.4  -0.844   8.4-176.9 -97.7 127.9  -23.0   22.1   27.4\\n', '  328   26 C R  E     -XY 392 439G 129    111,-3.4   111,-3.4    -2,-0.5     2,-0.3  -0.990  13.0-153.1-130.8 134.1  -22.2   18.4   27.1\\n', '  329   27 C S        -     0   0    6     62,-2.5   109,-0.1    -2,-0.4   107,-0.0  -0.774  13.7-148.3-101.8 146.1  -20.8   16.4   24.2\\n', '  330   28 C S  S    S+     0   0   87    107,-0.4     2,-0.3    -2,-0.3    -1,-0.1   0.408  83.9  26.8 -90.5   2.9  -21.5   12.8   23.6\\n', '  331   29 C A        -     0   0   29     60,-0.2   104,-0.5    57,-0.0     2,-0.4  -0.976  59.9-144.3-156.3 167.2  -18.0   12.5   22.1\\n', '  332   30 C F        -     0   0   50     -2,-0.3     2,-1.1   102,-0.2   102,-0.2  -0.925  13.3-159.1-140.2 110.9  -14.5   13.8   22.0\\n', '  333   31 C N    >   -     0   0   71    100,-3.1     3,-3.1    -2,-0.4   101,-0.0  -0.767   3.1-164.8 -95.7  93.4  -12.7   13.7   18.6\\n', '  334   32 C S  T 3  S+     0   0   85     -2,-1.1    -1,-0.2     1,-0.3    99,-0.1   0.806  89.2  49.9 -44.5 -40.8   -9.0   13.9   19.5\\n', '  335   33 C K  T 3  S+     0   0  177     -3,-0.1    -1,-0.3    97,-0.1    98,-0.1   0.250  87.7 122.3 -86.2  12.5   -8.1   14.8   15.9\\n', '  336   34 C G    <   -     0   0    8     -3,-3.1    96,-0.2    97,-0.1    97,-0.1  -0.217  52.0-136.5 -70.9 166.5  -10.7   17.5   15.8\\n', '  337   35 C G        -     0   0   26     94,-3.6    96,-0.3     1,-0.2    -1,-0.1  -0.198  50.6 -27.3-107.7-159.6   -9.8   21.1   15.0\\n', '  338   36 C T        -     0   0   67     19,-0.1    92,-0.5    94,-0.1     2,-0.4  -0.289  61.4-165.5 -55.9 127.3  -10.7   24.6   16.3\\n', '  339   37 C T  E     -C  356   0H   0     17,-2.4    17,-3.1    90,-0.1     2,-0.4  -0.977   2.0-160.3-123.7 127.8  -14.2   24.5   17.9\\n', '  340   38 C V  E     -CD 355 428H  17     88,-3.3    88,-2.3    -2,-0.4     2,-0.4  -0.848  14.0-175.1-116.1 145.8  -16.2   27.6   18.7\\n', '  341   39 C F  E     -CD 354 427H   0     13,-1.7    13,-2.6    -2,-0.4     2,-0.4  -0.997  12.7-169.7-131.5 134.0  -19.1   28.5   20.9\\n', '  342   40 C N  E     -CD 353 426H  12     84,-2.4    84,-2.7    -2,-0.4     2,-0.7  -0.987  18.3-150.6-130.0 137.2  -20.7   32.0   20.9\\n', '  343   41 C F  E     -CD 352 425H   0      9,-2.6     8,-1.4    -2,-0.4     9,-1.4  -0.934  29.8-160.5-103.0 118.2  -23.2   33.6   23.2\\n', '  344   42 C L  E     -CD 350 424H  21     80,-3.4    79,-3.9    -2,-0.7    80,-1.7  -0.835  13.5-134.3-109.7 140.9  -25.1   36.1   21.1\\n', '  345   43 C S  E >   - D   0 422H   4      4,-2.4     3,-2.6    -2,-0.4    77,-0.2  -0.349  39.8-100.9 -78.3 165.9  -27.2   39.1   22.0\\n', '  346   44 C A  T 3  S+     0   0   87     75,-0.6    -1,-0.1     1,-0.3    76,-0.1   0.794 129.1  57.1 -58.6 -25.7  -30.5   39.6   20.2\\n', '  347   45 C G  T 3  S-     0   0   39      2,-0.1    -1,-0.3     1,-0.0    23,-0.2   0.268 122.6-112.2 -88.2  11.6  -28.6   42.2   18.1\\n', '  348   46 C E    <   +     0   0  102     -3,-2.6     2,-0.3     1,-0.2    -2,-0.2   0.802  65.1 151.9  62.3  32.9  -26.3   39.4   17.2\\n', '  349   47 C N        -     0   0   25     20,-0.1    -4,-2.4    21,-0.1     2,-1.4  -0.740  52.9-117.6 -90.2 143.0  -23.3   40.8   19.0\\n', '  350   48 C I  E     +CE 344 368H  13     18,-2.2    18,-1.6    -2,-0.3    -6,-0.2  -0.657  38.3 172.2 -85.5  90.9  -20.8   38.3   20.3\\n', '  351   49 C L  E    S+     0   0H   3     -8,-1.4     2,-0.4    -2,-1.4    66,-0.4   0.915  74.0  14.5 -63.6 -45.5  -21.0   38.8   24.0\\n', '  352   50 C L  E     -C  343   0H   0     -9,-1.4    -9,-2.6    -3,-0.2     2,-0.5  -0.974  67.2-173.7-139.5 124.3  -18.8   35.7   24.7\\n', '  353   51 C H  E     -CE 342 366H  17     13,-2.5    13,-1.5    -2,-0.4     2,-0.5  -0.958   1.9-171.4-117.6 127.9  -16.7   33.9   22.2\\n', '  354   52 C I  E     -CE 341 365H   0    -13,-2.6   -13,-1.7    -2,-0.5     2,-0.7  -0.937   3.4-174.6-122.8 105.6  -14.9   30.6   23.2\\n', '  355   53 C S  E     -CE 340 364H   4      9,-3.1     9,-2.6    -2,-0.5     2,-0.8  -0.885   8.7-160.2-105.0 107.3  -12.5   29.4   20.5\\n', '  356   54 C I  E     -CE 339 363H   0    -17,-3.1   -17,-2.4    -2,-0.7     7,-0.2  -0.775  11.1-177.4 -89.6 109.7  -11.0   26.0   21.5\\n', '  357   55 C R  E  >> + E   0 362H  48      5,-2.8     5,-2.2    -2,-0.8     4,-0.5  -0.767  13.3 173.9-113.1  87.8   -7.8   25.6   19.5\\n', '  358   56 C P  T  45S+     0   0   30      0, 0.0     3,-0.2     0, 0.0    -1,-0.2   0.811  82.6  54.7 -61.3 -29.8   -6.2   22.2   20.1\\n', '  359   57 C G  T  45S+     0   0   80      1,-0.2    -2,-0.1     3,-0.1   -22,-0.0   0.872 114.3  37.8 -71.9 -37.1   -3.6   22.8   17.4\\n', '  360   58 C E  T  45S-     0   0   94     -3,-0.3    -1,-0.2     2,-0.2     3,-0.1   0.379 109.3-128.5 -92.1   2.1   -2.4   26.1   18.9\\n', '  361   59 C N  T  <5S+     0   0   84     -4,-0.5    21,-1.0     1,-0.2     2,-0.3   0.883  71.3 104.9  50.1  46.5   -2.9   24.5   22.3\\n', '  362   60 C V  E   < -EF 357 381H  12     -5,-2.2    -5,-2.8    19,-0.3     2,-0.5  -0.989  61.7-144.4-151.5 156.8   -5.0   27.4   23.5\\n', '  363   61 C I  E     -EF 356 380H   0     17,-2.3    17,-1.5    -2,-0.3     2,-0.4  -0.996  25.8-156.8-123.1 119.2   -8.6   28.4   24.3\\n', '  364   62 C V  E     -EF 355 379H   0     -9,-2.6    -9,-3.1    -2,-0.5     2,-0.4  -0.854   6.2-163.1-105.1 134.3   -9.3   32.0   23.4\\n', '  365   63 C F  E     +EF 354 378H   1     13,-2.7    13,-2.0    -2,-0.4     2,-0.3  -0.912  26.6 139.3-112.4 140.4  -12.0   34.1   24.9\\n', '  366   64 C N  E     -E  353   0H   1    -13,-1.5   -13,-2.5    -2,-0.4     2,-0.3  -0.959  32.6-144.5-164.7 178.2  -13.2   37.3   23.3\\n', '  367   65 C S  E     +     0   0H   5     -2,-0.3     8,-1.8   -15,-0.2     2,-0.3  -0.960  17.1 170.6-150.7 158.1  -16.1   39.5   22.4\\n', '  368   66 C R  E     -E  350   0H  68    -18,-1.6   -18,-2.2    -2,-0.3     6,-0.1  -0.918  35.0-101.6-171.3 146.6  -17.0   41.8   19.4\\n', '  369   67 C L    >   -     0   0   94      4,-0.3     3,-0.9    -2,-0.3   -20,-0.1  -0.323  52.4 -90.7 -70.3 157.1  -19.9   43.8   18.0\\n', '  370   68 C K  T 3  S+     0   0  106      1,-0.3    -1,-0.1   -23,-0.2   -21,-0.1   0.771 126.3   9.8 -36.9 -38.9  -21.8   42.4   15.1\\n', '  371   69 C N  T 3  S+     0   0  152      2,-0.1    -1,-0.3     0, 0.0   -22,-0.0   0.321 109.2 103.2-127.1   5.5  -19.5   44.2   12.8\\n', '  372   70 C G  S <  S-     0   0   35     -3,-0.9     2,-0.2     1,-0.1     3,-0.0  -0.148  74.0 -92.6 -80.8-178.7  -16.7   45.4   15.1\\n', '  373   71 C A        -     0   0   88      1,-0.1    -4,-0.3    -5,-0.0     2,-0.3  -0.490  46.8 -85.7 -92.7 165.0  -13.2   44.0   15.5\\n', '  374   72 C W        -     0   0  110     -2,-0.2    -6,-0.2    -6,-0.1     3,-0.1  -0.552  45.1-140.5 -71.5 126.3  -12.1   41.4   18.0\\n', '  375   73 C G        -     0   0   30     -8,-1.8    -1,-0.1    -2,-0.3    -7,-0.0   0.132  50.9 -42.7 -70.1-165.5  -11.1   43.0   21.3\\n', '  376   74 C P  P>    -     0   0   96      0, 0.0     2,-0.5     0, 0.0    -1,-0.2  -0.294  65.5-118.0 -62.6 149.2   -8.1   41.9   23.3\\n', '  377   75 C E  PP    -     0   0   85     -3,-0.1     2,-0.7   -11,-0.1   -11,-0.2  -0.793  18.5-160.2 -95.6 124.6   -7.7   38.1   23.6\\n', '  378   76 C E  E<    -F  365   0H  71    -13,-2.0   -13,-2.7    -2,-0.5     2,-0.3  -0.905  25.2-172.6-100.6 118.3   -7.8   36.5   27.0\\n', '  379   77 C R  E     +F  364   0H 111     -2,-0.7   -15,-0.2   -15,-0.2   -17,-0.0  -0.839  19.5 177.3-120.3 156.1   -6.1   33.1   26.7\\n', '  380   78 C I  E     -F  363   0H  34    -17,-1.5   -17,-2.3    -2,-0.3    32,-0.0  -0.966  43.0 -82.5-146.9 155.7   -5.6   30.0   28.7\\n', '  381   79 C P  E     -F  362   0H  81      0, 0.0   -19,-0.3     0, 0.0   -20,-0.1  -0.299  26.7-152.7 -62.6 145.3   -3.9   26.6   28.0\\n', '  382   80 C Y    >   +     0   0   33    -21,-1.0     3,-0.6   -25,-0.2     4,-0.3   0.877  61.8 104.8 -84.2 -43.5   -6.0   24.0   26.2\\n', '  383   81 C A  T 3  S+     0   0   42    -22,-0.2     0, 0.0     1,-0.2     0, 0.0   0.060  88.9   3.0 -42.0 145.6   -4.2   20.9   27.6\\n', '  384   82 C E  T 3  S+     0   0  169      1,-0.1    -1,-0.2     2,-0.1     3,-0.1   0.831  95.5 110.3  37.4  54.7   -5.9   18.8   30.3\\n', '  385   83 C K    <   +     0   0   30     -3,-0.6    23,-2.1     1,-0.2    22,-1.3   0.641  67.4  57.1-116.1 -37.8   -9.2   20.9   30.3\\n', '  386   84 C F        -     0   0   16     -4,-0.3    -1,-0.2    21,-0.2    -2,-0.1  -0.730  65.8-148.8-104.0 148.8  -11.5   18.3   28.8\\n', '  387   85 C R        -     0   0  115     -2,-0.3     3,-0.4     3,-0.1     0, 0.0  -0.936  33.0-111.2-118.0 111.4  -12.4   14.8   29.8\\n', '  388   86 C P  S    S+     0   0   75      0, 0.0   -57,-0.0     0, 0.0   -58,-0.0  -0.367  84.9  26.4 -74.1 161.2  -13.3   12.2   27.2\\n', '  389   87 C P  S    S+     0   0   92      0, 0.0   -56,-0.0     0, 0.0    -2,-0.0  -0.993 134.6  23.1 -89.2  -5.1  -15.5   10.6   26.1\\n', '  390   88 C N  S    S-     0   0   95     -3,-0.4   -60,-0.1   -60,-0.1   -58,-0.1  -0.611  78.2-178.0-118.8  70.3  -18.2   13.1   27.3\\n', '  391   89 C P        -     0   0    2      0, 0.0   -62,-2.5     0, 0.0     2,-0.3  -0.266   9.0-161.8 -66.6 158.0  -16.4   16.5   27.6\\n', '  392   90 C S  E     -X  328   0G  25    -64,-0.2     2,-0.4     2,-0.0    15,-0.3  -0.932  20.1-155.0-141.4 165.1  -18.3   19.5   28.9\\n', '  393   91 C I  E     -X  327   0G   3    -66,-2.7   -66,-3.6    -2,-0.3     2,-0.5  -0.987  11.2-166.7-140.6 126.8  -18.4   23.3   29.2\\n', '  394   92 C T  E     -XZ 326 405G  26     11,-3.0    11,-3.2    -2,-0.4     2,-0.4  -0.968   3.7-173.0-120.4 126.0  -20.3   25.2   31.8\\n', '  395   93 C V  E     -XZ 325 404G   0    -70,-2.5   -70,-3.0    -2,-0.5     2,-0.4  -0.954   1.4-173.7-118.2 133.3  -20.9   28.9   31.6\\n', '  396   94 C I  E     -XZ 324 403G  12      7,-2.6     7,-2.8    -2,-0.4     2,-0.7  -0.993  20.7-136.3-129.4 132.5  -22.4   31.0   34.3\\n', '  397   95 C D  E     + Z   0 402G   0    -74,-2.5   -75,-3.1    -2,-0.4     5,-0.2  -0.788  21.1 177.9 -89.2 115.7  -23.4   34.6   34.1\\n', '  398   96 C H        -     0   0    5      3,-3.0     2,-1.9    -2,-0.7     4,-0.1   0.060  47.5-113.7-104.1  23.7  -22.3   36.4   37.3\\n', '  399   97 C G  S    S+     0   0   31      1,-0.2     3,-0.1     2,-0.1    -1,-0.1  -0.398 110.5  39.4  82.9 -63.8  -23.5   39.8   36.2\\n', '  400   98 C D  S    S+     0   0   72     -2,-1.9    16,-1.6     1,-0.2     2,-0.3   0.495 130.4   9.0 -96.4  -7.2  -20.2   41.6   35.8\\n', '  401   99 C R  E    S- A   0 415G  11     14,-0.3    -3,-3.0    16,-0.0     2,-0.3  -0.985  70.8-114.9-165.1 161.2  -18.5   38.6   34.3\\n', '  402  100 C F  E     -ZA 397 414G   5     12,-2.7    12,-2.7    -2,-0.3     2,-0.5  -0.782  22.1-144.6-101.4 148.7  -19.0   35.1   32.9\\n', '  403  101 C Q  E     -ZA 396 413G   9     -7,-2.8    -7,-2.6    -2,-0.3     2,-0.5  -0.969  16.1-168.7-112.2 127.7  -17.6   32.1   34.7\\n', '  404  102 C I  E     -ZA 395 412G   0      8,-3.4     8,-3.2    -2,-0.5     2,-0.4  -0.976   6.3-179.4-122.1 122.7  -16.5   29.3   32.3\\n', '  405  103 C R  E     -Z  394   0G  20    -11,-3.2   -11,-3.0    -2,-0.5     6,-0.1  -0.971   5.7-171.8-125.4 136.1  -15.6   25.9   33.6\\n', '  406  104 C F        -     0   0    5      2,-0.4   -20,-0.2    -2,-0.4   -13,-0.1  -0.149  49.1 -78.9-102.9-159.3  -14.4   22.9   31.8\\n', '  407  105 C D  S    S+     0   0   15    -22,-1.3     2,-0.3   -15,-0.3   -21,-0.2   0.874 109.9  26.6 -75.3 -36.7  -13.9   19.3   33.0\\n', '  408  106 C Y  S    S+     0   0   97    -23,-2.1    -2,-0.4     2,-0.1   -21,-0.0  -0.785 100.1  22.9-123.2 167.3  -10.6   19.9   34.8\\n', '  409  107 C G  S    S-     0   0   47     -2,-0.3     2,-0.1    -4,-0.1    -3,-0.1  -0.019  92.9 -43.6  73.7-179.4   -8.8   22.8   36.4\\n', '  410  108 C T        -     0   0   91      1,-0.1    -4,-0.2    -5,-0.1    -2,-0.1  -0.410  68.1 -92.6 -82.4 161.5  -10.1   26.1   37.8\\n', '  411  109 C S        -     0   0   10     -6,-0.1     2,-0.5    -2,-0.1    -6,-0.2  -0.206  30.6-126.9 -69.4 163.3  -12.7   28.2   36.2\\n', '  412  110 C I  E     -A  404   0G  21     -8,-3.2    -8,-3.4     2,-0.0     2,-0.5  -0.945  19.4-152.8-111.3 132.5  -12.0   31.0   33.8\\n', '  413  111 C Y  E     -A  403   0G  33     -2,-0.5     2,-0.5   -10,-0.2   -10,-0.2  -0.924   9.9-171.3-110.2 133.3  -13.6   34.4   34.6\\n', '  414  112 C Y  E     -A  402   0G   0    -12,-2.7   -12,-2.7    -2,-0.5     3,-0.1  -0.928  19.1-136.0-129.1 107.1  -14.3   36.8   31.8\\n', '  415  113 C N  E     -A  401   0G  93     -2,-0.5   -14,-0.3   -14,-0.3   -15,-0.1  -0.348  35.8-103.5 -60.1 137.1  -15.5   40.3   32.6\\n', '  416  114 C K        -     0   0   47    -16,-1.6     3,-0.1     1,-0.1    -1,-0.1  -0.342  33.6-175.4 -63.8 142.4  -18.3   41.3   30.3\\n', '  417  115 C R  S    S+     0   0   79    -66,-0.4     2,-0.4     1,-0.3    -1,-0.1   0.513  72.3  47.4-114.3 -13.8  -17.4   43.7   27.5\\n', '  418  116 C I  S    S-     0   0   25    -67,-0.2     2,-2.4     2,-0.1    -1,-0.3  -0.970  78.5-135.5-133.9 117.3  -20.9   44.3   26.1\\n', '  419  117 C K        +     0   0  206     -2,-0.4     2,-0.3    -3,-0.1   -68,-0.0  -0.404  66.9 105.2 -74.6  73.0  -23.8   45.1   28.3\\n', '  420  118 C E        -     0   0   66     -2,-2.4     2,-0.4   -75,-0.0    -2,-0.1  -0.985  67.7-118.8-146.8 152.7  -26.4   42.8   26.8\\n', '  421  119 C N        -     0   0   59     -2,-0.3   -75,-0.6     1,-0.1  -101,-0.2  -0.788  37.6-110.5 -95.4 138.0  -28.1   39.5   27.7\\n', '  422  120 C A  E     -BD 319 345H   1   -103,-1.3  -103,-3.0    -2,-0.4   -77,-0.3  -0.364  30.2-179.7 -68.0 143.1  -27.5   36.6   25.3\\n', '  423  121 C A  E     +     0   0H  19    -79,-3.9  -108,-2.8     1,-0.3     2,-0.3   0.459  67.8  20.9-118.0 -10.9  -30.5   35.4   23.3\\n', '  424  122 C A  E     -BD 314 344H   8    -80,-1.7   -80,-3.4  -110,-0.3     2,-0.4  -0.994  59.9-140.7-157.0 157.8  -28.8   32.6   21.4\\n', '  425  123 C I  E     -BD 313 343H   0   -112,-2.0  -112,-2.1    -2,-0.3     2,-0.3  -0.992  20.0-162.7-125.0 129.5  -25.8   30.3   21.5\\n', '  426  124 C A  E     -BD 312 342H   9    -84,-2.7   -84,-2.4    -2,-0.4     2,-0.4  -0.786   4.6-160.2-111.0 157.0  -23.9   29.4   18.3\\n', '  427  125 C Y  E     +BD 311 341H   0   -116,-2.1  -117,-2.3  -119,-0.3  -116,-1.0  -0.988  14.0 178.5-140.0 120.8  -21.5   26.5   17.7\\n', '  428  126 C N  E     + D   0 340H  64    -88,-2.3   -88,-3.3    -2,-0.4     2,-0.2  -0.975  13.8 133.9-130.2 146.2  -19.0   26.5   14.8\\n', '  429  127 C A        -     0   0   11     -2,-0.4     3,-0.3   -90,-0.2   -90,-0.1  -0.816  41.4-147.8 176.7 140.6  -16.3   24.1   13.5\\n', '  430  128 C E  S    S+     0   0  207    -92,-0.5     2,-0.2     1,-0.3    -1,-0.1   0.849 109.3   9.7 -80.7 -39.1  -15.0   22.5   10.3\\n', '  431  129 C N  S    S-     0   0   95    -93,-0.4   -94,-3.6    -3,-0.1    -1,-0.3  -0.744  94.6-146.6-142.7  84.7  -14.0   19.4   12.2\\n', '  432  130 C S        -     0   0   13     -3,-0.3   -94,-0.1    -2,-0.2   -99,-0.1  -0.176  17.6-144.9 -61.5 143.5  -15.4   19.7   15.8\\n', '  433  131 C L  S    S+     0   0    0    -96,-0.3  -100,-3.1   -97,-0.1     2,-0.2   0.727  88.6  52.5 -73.7 -22.2  -13.7   18.3   18.9\\n', '  434  132 C F  S    S-     0   0    8   -102,-0.2  -102,-0.2     1,-0.1    -2,-0.1  -0.497  94.8 -85.1-112.3 177.7  -17.3   17.5   20.0\\n', '  435  133 C S        -     0   0   14   -104,-0.5    -1,-0.1    -2,-0.2    -2,-0.1  -0.319  47.9 -97.7 -78.6 162.8  -20.4   15.8   18.9\\n', '  436  134 C S  S    S+     0   0   61   -128,-0.1  -128,-0.2     1,-0.1     2,-0.1  -0.978 108.4  21.7-126.8 115.0  -23.1   17.4   16.7\\n', '  437  135 C P  S    S-     0   0   33      0, 0.0  -107,-0.4     0, 0.0     2,-0.3   0.627  83.5-154.3 -87.6-176.7  -25.3   18.3   18.4\\n', '  438  136 C V  E     -W  307   0G   1   -131,-2.0  -131,-1.9    -2,-0.1     2,-0.4  -0.836  26.1-112.5-109.2 152.5  -24.1   18.9   21.9\\n', '  439  137 C T  E     -WY 306 328G  38   -111,-3.4  -111,-3.4    -2,-0.3     2,-0.5  -0.722  34.3-169.3 -86.4 134.9  -26.6   18.8   24.7\\n', '  440  138 C V  E     -WY 305 327G   0   -135,-2.4  -135,-2.1    -2,-0.4     2,-0.6  -0.953  10.1-170.9-131.6 116.5  -27.2   22.1   26.4\\n', '  441  139 C D  E     -WY 304 326G  14   -115,-2.3  -115,-2.9    -2,-0.5     2,-0.5  -0.923  10.1-166.8-106.8 118.9  -29.2   22.6   29.6\\n', '  442  140 C V  E     - Y   0 325G  16   -139,-2.8  -139,-0.8    -2,-0.6     2,-0.3  -0.918   3.3-169.2-110.1 127.1  -29.9   26.2   30.4\\n', '  443  141 C H  E     - Y   0 324G  15   -119,-3.2  -119,-2.0    -2,-0.5     3,-0.1  -0.843  28.0-141.4-113.8 150.9  -31.1   27.3   33.9\\n', '  444  142 C G  S    S+     0   0   50     -2,-0.3     2,-0.3     1,-0.2    -1,-0.1   0.701  93.1   5.9 -79.9 -19.6  -32.5   30.6   35.1\\n', '  445  143 C L  S    S-     0   0   84   -121,-0.1    -1,-0.2  -122,-0.1  -122,-0.1  -0.985  98.0 -76.4-158.1 155.9  -30.6   30.0   38.3\\n', '  446  144 C L  P>    -     0   0   30     -2,-0.3     3,-0.1  -124,-0.3  -123,-0.0  -0.371  59.1-134.0 -56.7 122.8  -28.1   27.5   39.8\\n', '  447  145 C P  PP    -     0   0    8      0, 0.0    -1,-0.1     0, 0.0    -4,-0.0  -0.267  31.1 -72.4 -79.3 167.4  -30.2   24.5   40.7\\n', '  448  146 C P  PP    -     0   0  109      0, 0.0   155,-0.1     0, 0.0    29,-0.0  -0.312  58.8-126.1 -55.9 140.8  -30.3   22.4   43.9\\n', '  449  147 C L  PP    -     0   0   40     -3,-0.1   151,-0.1   151,-0.1   148,-0.0  -0.743   7.8-126.8-101.4 140.6  -27.1   20.3   44.0\\n', '  450  148 C P  P<    -     0   0   41      0, 0.0    -1,-0.0     0, 0.0   142,-0.0  -0.389  37.1-102.9 -73.7 160.8  -26.7   16.6   44.4\\n', '  451  149 C P              0   0   87      0, 0.0    -2,-0.0     0, 0.0   142,-0.0   0.786 360.0 360.0 -53.6 -31.4  -24.4   15.4   47.3\\n', '  452  150 C A              0   0  110    147,-0.1   140,-0.0   142,-0.1   142,-0.0   0.315 360.0 360.0-160.6 360.0  -21.8   14.6   44.6\\n', '  453        !*             0   0    0      0, 0.0     0, 0.0     0, 0.0     0, 0.0   0.000 360.0 360.0 360.0 360.0    0.0    0.0    0.0\\n', '  454    1 D M              0   0  103    140,-0.1   139,-2.9   137,-0.1     2,-0.4   0.000 360.0 360.0 360.0 123.6  -18.5   15.6   51.6\\n', '  455    2 D L  E     +G  592   0I  88    137,-0.2     2,-0.3   135,-0.0   137,-0.2  -0.750 360.0 176.8 -93.2 130.4  -21.3   15.7   54.1\\n', '  456    3 D Y  E     -G  591   0I  29    135,-2.5   135,-2.4    -2,-0.4     2,-0.7  -0.937  25.3-141.7-128.0 151.5  -20.9   17.4   57.5\\n', '  457    4 D H  E     -G  590   0I 100     -2,-0.3     2,-0.7   133,-0.2   133,-0.2  -0.947  21.6-161.8-114.3 107.9  -23.4   18.0   60.3\\n', '  458    5 D L  E     -G  589   0I   0    131,-2.2   131,-1.7    -2,-0.7     2,-0.2  -0.821   5.3-153.7 -96.3 110.3  -22.7   21.4   61.7\\n', '  459    6 D F    >   -     0   0  103     -2,-0.7     3,-0.8   129,-0.3   119,-0.4  -0.545  27.7 -95.8 -84.4 148.6  -24.2   22.0   65.2\\n', '  460    7 D V  T 3  S+     0   0   35      1,-0.3   119,-0.2    -2,-0.2     3,-0.1  -0.251 109.6  16.3 -60.6 145.7  -25.1   25.4   66.5\\n', '  461    8 D N  T 3  S+     0   0   92    117,-1.2    -1,-0.3     1,-0.3     2,-0.2   0.881 122.3  77.3  54.6  40.2  -22.5   27.2   68.6\\n', '  462    9 D N  E <   -L  578   0J  94    116,-1.7   116,-2.8    -3,-0.8     2,-0.3  -0.828  66.5-142.0-154.6-170.4  -20.0   24.7   67.2\\n', '  463   10 D Q  E     -L  577   0J  83    114,-0.3     2,-0.4    -2,-0.2   114,-0.2  -0.951   9.9-126.3-156.6 172.0  -18.0   24.0   64.1\\n', '  464   11 D V  E     -L  576   0J  18    112,-2.0   112,-1.8    -2,-0.3     2,-0.6  -0.996  19.9-133.4-129.3 134.9  -16.8   21.3   61.7\\n', '  465   12 D K  E     -L  575   0J 143     -2,-0.4   110,-0.3   110,-0.2     2,-0.1  -0.795  31.8-124.1 -86.8 122.2  -13.2   20.9   60.7\\n', '  466   13 D L  E     -     0   0J   5    108,-2.4     3,-0.0    -2,-0.6     4,-0.0  -0.419   1.9-140.9 -70.9 139.2  -13.1   20.4   56.9\\n', '  467   14 D Q  E    S+     0   0J 133     -2,-0.1     2,-0.3     1,-0.1    -1,-0.1   0.842  93.1  23.3 -65.5 -32.8  -11.4   17.3   55.5\\n', '  468   15 D N  E    S-     0   0J 113    105,-0.0    -1,-0.1     3,-0.0     2,-0.1  -0.940 100.7 -89.2-132.1 154.6  -10.0   19.5   52.8\\n', '  469   16 D D  E     -     0   0J  81     -2,-0.3     2,-0.9     1,-0.1   105,-0.2  -0.365  37.9-122.5 -62.1 137.0   -9.3   23.2   52.6\\n', '  470   17 D F  E     +L  573   0J   1    103,-3.4   103,-2.3     4,-0.1    -1,-0.1  -0.758  57.9 149.1 -82.5 110.0  -12.3   25.2   51.3\\n', '  471   18 D K    >   -     0   0   91     -2,-0.9     3,-2.7   101,-0.2    77,-0.3  -0.664  48.4 -35.1-133.1-174.5  -10.7   26.9   48.3\\n', '  472   19 D P  T 3  S+     0   0   26      0, 0.0    77,-0.2     0, 0.0    -1,-0.2  -0.190 128.7  20.6 -49.9 135.2  -11.4   28.3   44.8\\n', '  473   20 D E  T 3  S+     0   0   15     75,-3.2   124,-0.1     1,-0.3    76,-0.1   0.017  95.8 132.6  91.1 -23.3  -14.0   26.2   43.0\\n', '  474   21 D S    <   -     0   0    4     -3,-2.7    74,-2.4   122,-0.1     2,-0.4  -0.252  39.7-161.6 -62.1 146.2  -15.2   24.9   46.4\\n', '  475   22 D V  E     -HI 547 594I   1    119,-2.2   119,-3.3    72,-0.2     2,-0.6  -0.998   8.6-165.6-133.5 128.0  -18.9   24.9   47.1\\n', '  476   23 D A  E     -HI 546 593I   0     70,-2.9    70,-2.6    -2,-0.4     2,-0.8  -0.962   5.2-164.9-116.5 114.4  -20.6   24.6   50.5\\n', '  477   24 D A  E     -HI 545 592I  12    115,-2.9   115,-2.2    -2,-0.6     2,-0.5  -0.865   8.9-168.4-102.2 101.4  -24.3   23.8   50.4\\n', '  478   25 D I  E     -HI 544 591I   3     66,-3.1    66,-2.6    -2,-0.8     2,-0.4  -0.797   7.6-178.7 -92.2 126.9  -25.8   24.6   53.8\\n', '  479   26 D R  E     -HI 543 590I 131    111,-3.0   111,-2.9    -2,-0.5     2,-0.3  -0.990  14.2-152.2-133.6 130.4  -29.3   23.4   54.4\\n', '  480   27 D S        -     0   0    7     62,-2.7   109,-0.1    -2,-0.4     3,-0.0  -0.772  12.6-149.2 -99.7 142.8  -31.6   23.7   57.4\\n', '  481   28 D S  S    S+     0   0   93    107,-0.4     2,-0.3    -2,-0.3    -1,-0.1   0.332  84.4  27.1 -88.9   7.3  -34.3   21.2   58.3\\n', '  482   29 D A        -     0   0   27     60,-0.2   104,-0.6     2,-0.0     2,-0.4  -0.993  59.9-144.8-162.0 161.0  -36.3   24.1   59.8\\n', '  483   30 D F  B     -Q  585   0K  42     -2,-0.3     2,-1.5   102,-0.2   102,-0.2  -0.889  17.6-157.2-133.6  98.7  -37.0   27.8   59.7\\n', '  484   31 D N    >   -     0   0   69    100,-2.1     3,-1.2    -2,-0.4    -2,-0.0  -0.649   4.7-168.6 -82.1  91.8  -37.6   29.2   63.2\\n', '  485   32 D S  T 3  S+     0   0   79     -2,-1.5     2,-0.4     1,-0.3    -1,-0.2   0.814  85.9  55.3 -47.7 -35.9  -39.6   32.3   62.4\\n', '  486   33 D K  T 3  S+     0   0  182     -3,-0.1    -1,-0.3    97,-0.1     2,-0.2  -0.248  88.1 116.6 -96.6  46.6  -39.2   33.3   66.0\\n', '  487   34 D G    <   -     0   0   12     -3,-1.2    23,-0.2    -2,-0.4    96,-0.2  -0.717  52.6-132.7-112.5 164.2  -35.4   33.1   66.1\\n', '  488   35 D G        -     0   0   28     94,-2.4    96,-0.3    -2,-0.2    -1,-0.1  -0.027  54.2 -30.5 -98.2-158.8  -32.6   35.6   66.7\\n', '  489   36 D T        -     0   0   65     19,-0.1    92,-0.4    94,-0.1     2,-0.4  -0.335  60.2-163.3 -59.1 130.7  -29.3   36.6   65.1\\n', '  490   37 D T  E     -M  507   0J   0     17,-3.0    17,-3.4    90,-0.1     2,-0.4  -0.977   2.6-160.8-123.9 130.2  -27.7   33.6   63.4\\n', '  491   38 D V  E     -MN 506 579J  19     88,-3.5    88,-2.9    -2,-0.4     2,-0.4  -0.903  14.3-175.9-120.1 141.6  -24.0   33.4   62.4\\n', '  492   39 D F  E     -MN 505 578J   0     13,-1.5    13,-2.1    -2,-0.4     2,-0.4  -1.000  14.5-169.4-128.4 127.1  -21.9   31.2   60.1\\n', '  493   40 D N  E     -MN 504 577J  15     84,-2.5    84,-3.1    -2,-0.4     2,-0.6  -0.960  17.3-152.9-122.6 138.6  -18.2   31.7   59.9\\n', '  494   41 D F  E     -MN 503 576J   0      9,-2.3     8,-1.4    -2,-0.4     9,-1.3  -0.950  28.2-165.5-106.2 116.6  -15.6   30.4   57.5\\n', '  495   42 D L  E     -MN 501 575J  17     80,-2.5    79,-3.1    -2,-0.6    80,-1.7  -0.843  14.7-135.3-111.5 145.9  -12.3   30.2   59.4\\n', '  496   43 D S    >   -     0   0    7      4,-2.0     3,-2.0    -2,-0.4    77,-0.2  -0.287  41.0 -96.3 -83.3 174.4   -8.7   29.8   58.2\\n', '  497   44 D A  T 3  S+     0   0   88     75,-0.4    -1,-0.1     1,-0.3    76,-0.1   0.869 131.0  53.9 -60.5 -35.3   -6.3   27.4   60.0\\n', '  498   45 D G  T 3  S-     0   0   44      1,-0.1    -1,-0.3     2,-0.0    23,-0.2   0.294 122.9-112.5 -81.6  11.4   -5.1   30.4   61.9\\n', '  499   46 D E    <   +     0   0   91     -3,-2.0     2,-0.3     1,-0.2    -2,-0.1   0.771  61.9 157.3  62.9  30.0   -8.7   31.0   62.9\\n', '  500   47 D N        -     0   0   20     20,-0.1    -4,-2.0     1,-0.1     2,-1.3  -0.618  49.1-120.4 -79.3 144.7   -9.0   34.2   60.9\\n', '  501   48 D I  E     +MO 495 519J   9     18,-3.2    18,-1.8    -2,-0.3    -6,-0.2  -0.706  38.1 173.2 -91.2  89.6  -12.7   34.9   60.0\\n', '  502   49 D L  E    S+     0   0J   3     -8,-1.4     2,-0.4    -2,-1.3    66,-0.4   0.921  72.0  17.1 -60.7 -49.1  -12.5   34.9   56.2\\n', '  503   50 D L  E     -M  494   0J   0     -9,-1.3    -9,-2.3    -3,-0.2     2,-0.5  -0.985  67.5-173.6-134.1 124.1  -16.2   35.2   55.8\\n', '  504   51 D H  E     -MO 493 517J  18     13,-2.2    13,-1.4    -2,-0.4     2,-0.6  -0.969   1.8-171.5-117.9 127.0  -18.7   36.2   58.4\\n', '  505   52 D I  E     -MO 492 516J   0    -13,-2.1   -13,-1.5    -2,-0.5     2,-0.8  -0.916   2.5-171.6-122.0 102.1  -22.4   36.0   57.7\\n', '  506   53 D S  E     -MO 491 515J   5      9,-3.4     9,-2.8    -2,-0.6     2,-0.9  -0.837   5.3-163.6 -98.7 106.9  -24.5   37.6   60.4\\n', '  507   54 D I  E     -MO 490 514J   0    -17,-3.4   -17,-3.0    -2,-0.8     7,-0.2  -0.796  10.7-178.9 -93.1 103.9  -28.2   36.8   59.8\\n', '  508   55 D R  E > > + O   0 513J  46      5,-2.2     5,-1.9    -2,-0.9     3,-1.1  -0.832  12.8 175.0-110.0  97.0  -30.2   39.3   61.9\\n', '  509   56 D P  G > 5S+     0   0   21      0, 0.0     3,-3.0     0, 0.0    -1,-0.2   0.978  81.5  55.8 -61.1 -57.1  -34.0   38.8   61.6\\n', '  510   57 D G  G 3 5S+     0   0   89      1,-0.3   -22,-0.1   -23,-0.2    -2,-0.0   0.445 115.3  40.1 -57.7   0.8  -34.8   41.4   64.3\\n', '  511   58 D E  G < 5S-     0   0   91     -3,-1.1    -1,-0.3     2,-0.2     3,-0.1   0.276 104.4-130.6-128.9   2.5  -32.8   44.0   62.3\\n', '  512   59 D N  T < 5S+     0   0   77     -3,-3.0    21,-1.1    -4,-0.2     2,-0.3   0.859  71.6 103.6  44.1  48.0  -34.0   42.8   58.9\\n', '  513   60 D V  E   < -OP 508 532J  22     -5,-1.9    -5,-2.2    19,-0.3     2,-0.5  -0.993  60.3-146.4-154.8 148.9  -30.5   42.7   57.6\\n', '  514   61 D I  E     -OP 507 531J   0     17,-3.0    17,-2.1    -2,-0.3     2,-0.4  -0.981  23.8-154.2-117.6 124.4  -27.7   40.2   56.8\\n', '  515   62 D V  E     -OP 506 530J   1     -9,-2.8    -9,-3.4    -2,-0.5     2,-0.4  -0.872   7.5-163.8-108.6 134.5  -24.2   41.4   57.4\\n', '  516   63 D F  E     +OP 505 529J   2     13,-2.7    13,-1.8    -2,-0.4     2,-0.3  -0.915  26.2 139.0-113.1 138.7  -21.0   40.2   55.7\\n', '  517   64 D N  E     -O  504   0J   2    -13,-1.4   -13,-2.2    -2,-0.4     2,-0.3  -0.968  34.9-135.4-165.0 173.9  -17.6   40.9   57.2\\n', '  518   65 D S  E     -     0   0J   6     -2,-0.3     8,-1.3     9,-0.2     2,-0.3  -0.891  18.8-179.3-136.0 165.7  -14.1   39.6   57.8\\n', '  519   66 D R  E     -O  501   0J  75    -18,-1.8   -18,-3.2    -2,-0.3     6,-0.1  -0.927  28.7-115.6-169.6 139.6  -11.7   39.7   60.8\\n', '  520   67 D L    >   -     0   0   99      4,-0.4     3,-2.4    -2,-0.3   -20,-0.1  -0.435  46.3-101.2 -73.8 156.2   -8.3   38.6   62.0\\n', '  521   68 D K  T 3  S+     0   0   97      1,-0.3    -1,-0.1   -23,-0.2   -22,-0.1   0.891 123.6  19.9 -43.3 -51.4   -8.3   36.1   64.9\\n', '  522   69 D N  T 3  S+     0   0  153      2,-0.1    -1,-0.3     0, 0.0     2,-0.2  -0.261 113.2 101.4-114.6  43.7   -7.4   38.9   67.3\\n', '  523   70 D G  S <  S-     0   0   32     -3,-2.4     2,-0.2     1,-0.0     3,-0.0  -0.678  70.9 -98.4-121.4 176.7   -8.5   41.8   65.1\\n', '  524   71 D A        -     0   0   91     -2,-0.2    -4,-0.4     1,-0.1     2,-0.1  -0.597  45.2 -91.4 -95.0 157.0  -11.5   44.2   64.8\\n', '  525   72 D W        -     0   0  107     -2,-0.2    -6,-0.2    -6,-0.1     3,-0.1  -0.391  42.1-141.9 -65.1 139.0  -14.4   43.9   62.4\\n', '  526   73 D G        -     0   0   31     -8,-1.3    -1,-0.1     1,-0.2     2,-0.0  -0.023  49.1 -41.3 -86.8-163.9  -13.8   45.7   59.1\\n', '  527   74 D P        -     0   0   97      0, 0.0    -1,-0.2     0, 0.0     2,-0.2  -0.360  67.8-121.3 -62.7 140.4  -16.4   47.7   57.1\\n', '  528   75 D E        -     0   0   93     -3,-0.1     2,-0.5   -11,-0.1   -11,-0.2  -0.550  17.7-160.9 -84.7 149.0  -19.8   45.9   57.0\\n', '  529   76 D E  E     +P  516   0J  70    -13,-1.8   -13,-2.7    -2,-0.2     2,-0.3  -0.944  25.3 177.5-128.1 103.2  -21.5   44.9   53.8\\n', '  530   77 D R  E     -P  515   0J 109     -2,-0.5   -15,-0.2   -15,-0.2   -17,-0.0  -0.774  16.8-173.1-115.2 159.6  -25.2   44.3   54.6\\n', '  531   78 D I  E     -P  514   0J  33    -17,-2.1   -17,-3.0    -2,-0.3    -2,-0.0  -0.957  41.5 -80.2-141.2 156.4  -28.3   43.4   52.7\\n', '  532   79 D P  E     -P  513   0J  76      0, 0.0   -19,-0.3     0, 0.0   -20,-0.1  -0.290  31.8-148.5 -60.5 144.6  -32.0   43.2   53.6\\n', '  533   80 D Y    >   +     0   0   32    -21,-1.1     3,-0.6   -25,-0.2     4,-0.4   0.888  67.0  90.7 -80.4 -45.6  -33.1   40.0   55.4\\n', '  534   81 D A  T 3  S+     0   0   46      1,-0.2     0, 0.0   -22,-0.2     0, 0.0  -0.101  89.0  20.7 -54.8 151.1  -36.6   39.6   54.2\\n', '  535   82 D E  T 3  S+     0   0  168      2,-0.2    -1,-0.2     1,-0.1     3,-0.1   0.812  97.7  98.1  58.6  35.8  -37.3   37.5   51.0\\n', '  536   83 D K  S <  S+     0   0   26     -3,-0.6    22,-1.8     1,-0.2    23,-1.6   0.701  72.5  53.1-115.4 -40.7  -34.0   35.6   51.2\\n', '  537   84 D F        -     0   0   14     -4,-0.4    -1,-0.2    21,-0.2    -2,-0.2  -0.781  64.5-149.7-105.9 145.0  -35.0   32.4   52.9\\n', '  538   85 D R    >   -     0   0  130     -2,-0.3     3,-0.6     3,-0.1     0, 0.0  -0.922  33.8-109.9-113.8 110.0  -37.7   29.9   52.0\\n', '  539   86 D P  T 3  S+     0   0   70      0, 0.0   -57,-0.0     0, 0.0   -55,-0.0  -0.323  87.0  27.6 -70.0 161.3  -39.3   27.9   54.8\\n', '  540   87 D P  T 3  S+     0   0   90      0, 0.0   -56,-0.0     0, 0.0    -2,-0.0  -0.987 134.4  23.8 -89.9  -1.2  -39.4   25.1   55.9\\n', '  541   88 D N  S <  S-     0   0   90     -3,-0.6    -3,-0.1   -60,-0.1   -60,-0.1  -0.616  76.1-177.4-123.0  70.8  -35.9   24.2   54.6\\n', '  542   89 D P        -     0   0    1      0, 0.0   -62,-2.7     0, 0.0     2,-0.3  -0.268   9.9-164.7 -65.5 158.5  -34.0   27.5   54.1\\n', '  543   90 D S  E     -H  479   0I  24    -64,-0.2     2,-0.4     2,-0.0    15,-0.3  -0.954  20.6-154.4-144.3 163.0  -30.5   27.2   52.6\\n', '  544   91 D I  E     -H  478   0I   2    -66,-2.6   -66,-3.1    -2,-0.3     2,-0.5  -0.989  11.4-164.1-138.3 126.3  -27.2   29.1   52.1\\n', '  545   92 D T  E     -HJ 477 556I  27     11,-3.1    11,-2.8    -2,-0.4     2,-0.5  -0.959   4.3-170.1-117.3 123.0  -24.8   28.3   49.3\\n', '  546   93 D V  E     -HJ 476 555I   0    -70,-2.6   -70,-2.9    -2,-0.5     2,-0.5  -0.949   2.5-174.3-116.1 123.4  -21.3   29.6   49.4\\n', '  547   94 D I  E     -HJ 475 554I  16      7,-2.7     7,-2.7    -2,-0.5     2,-0.8  -0.969  19.4-140.4-119.7 129.3  -19.0   29.3   46.4\\n', '  548   95 D D  E     + J   0 553I   1    -74,-2.4   -75,-3.2    -2,-0.5     5,-0.2  -0.797  21.6 175.4 -88.9 111.0  -15.3   30.3   46.5\\n', '  549   96 D H        -     0   0    3      3,-2.7     2,-1.9    -2,-0.8     4,-0.1   0.039  47.6-115.6-102.2  23.7  -14.5   32.1   43.2\\n', '  550   97 D G  S    S+     0   0    4      1,-0.2    -1,-0.1     2,-0.1     3,-0.1  -0.484 108.8  41.2  82.3 -74.5  -10.9   32.9   44.3\\n', '  551   98 D D  S    S+     0   0   46     -2,-1.9    16,-2.2     1,-0.1     2,-0.3   0.602 130.4   6.7 -83.5 -12.4  -11.1   36.7   44.3\\n', '  552   99 D R  E    S- K   0 566I  18     14,-0.3    -3,-2.7  -483,-0.1     2,-0.3  -0.989  70.4-114.3-161.7 164.3  -14.5   36.6   46.0\\n', '  553  100 D F  E     -JK 548 565I   5     12,-2.7    12,-2.5    -2,-0.3     2,-0.5  -0.807  20.0-147.3-103.9 147.3  -17.2   34.5   47.5\\n', '  554  101 D Q  E     -JK 547 564I   9     -7,-2.7    -7,-2.7    -2,-0.3     2,-0.5  -0.971  14.6-166.5-112.0 126.6  -20.6   34.1   46.0\\n', '  555  102 D I  E     -JK 546 563I   0      8,-3.4     8,-3.1    -2,-0.5     2,-0.4  -0.972   4.9-174.5-118.6 123.5  -23.4   33.6   48.5\\n', '  556  103 D R  E     -J  545   0I  27    -11,-2.8   -11,-3.1    -2,-0.5     6,-0.1  -0.958   2.5-176.0-117.9 132.8  -26.8   32.4   47.4\\n', '  557  104 D F        -     0   0    5      2,-0.5   -20,-0.2    -2,-0.4   -13,-0.1  -0.155  51.3 -79.1-104.3-156.3  -29.8   32.2   49.6\\n', '  558  105 D D  S    S+     0   0   17    -22,-1.8     2,-0.2   -15,-0.3   -21,-0.2   0.797 109.4  19.0 -80.5 -27.3  -33.2   30.8   48.5\\n', '  559  106 D Y  S    S+     0   0  106    -23,-1.6    -2,-0.5     2,-0.1   -21,-0.0  -0.762  97.3  37.2-132.0 179.2  -34.4   33.9   46.7\\n', '  560  107 D G  S    S-     0   0   46     -2,-0.2    -3,-0.1    -4,-0.1     3,-0.1   0.110  92.5 -48.9  66.4 175.2  -33.0   37.1   45.1\\n', '  561  108 D T  S    S-     0   0  100      1,-0.1    -4,-0.3    -5,-0.1     2,-0.1  -0.153  70.3 -84.6 -73.1 174.0  -29.7   37.5   43.2\\n', '  562  109 D S        -     0   0    8     -6,-0.1     2,-0.4     1,-0.0    -6,-0.2  -0.425  37.9-131.6 -77.0 156.2  -26.4   36.1   44.5\\n', '  563  110 D I  E     -K  555   0I  23     -8,-3.1    -8,-3.4    -2,-0.1     2,-0.5  -0.915  17.8-152.9-108.3 136.9  -24.4   38.3   46.9\\n', '  564  111 D Y  E     -K  554   0I  48     -2,-0.4     2,-0.5   -10,-0.2   -10,-0.2  -0.951   9.3-170.9-115.2 132.2  -20.7   38.7   46.1\\n', '  565  112 D Y  E     -K  553   0I   1    -12,-2.5   -12,-2.7    -2,-0.5     3,-0.1  -0.949  19.7-134.9-125.2 111.8  -18.1   39.4   48.8\\n', '  566  113 D N  E     -K  552   0I  97     -2,-0.5   -14,-0.3   -14,-0.2     2,-0.2  -0.365  34.2-103.2 -63.4 139.0  -14.6   40.3   47.8\\n', '  567  114 D K        +     0   0   15    -16,-2.2     3,-0.1     1,-0.2    -1,-0.1  -0.460  38.4 175.6 -69.4 130.7  -11.9   38.5   49.7\\n', '  568  115 D R  S    S+     0   0   79    -66,-0.4     2,-0.5     1,-0.3    -1,-0.2   0.663  75.6  43.8-104.1 -28.5  -10.3   40.5   52.5\\n', '  569  116 D I  S    S-     0   0   23    -67,-0.2     2,-1.9     2,-0.1    -1,-0.3  -0.980  78.4-146.1-122.0 115.9   -8.1   37.7   53.9\\n', '  570  117 D K        +     0   0   61     -2,-0.5  -498,-3.2    -3,-0.1     2,-0.3  -0.461  58.5 109.5 -83.2  67.8   -6.4   35.6   51.2\\n', '  571  118 D E        -     0   0   54     -2,-1.9     2,-0.1  -500,-0.2    -2,-0.1  -0.966  67.6-112.4-139.0 153.5   -6.5   32.2   53.0\\n', '  572  119 D N        -     0   0   38     -2,-0.3   -75,-0.4  -500,-0.3     2,-0.4  -0.478  38.6-105.7 -81.5 158.9   -8.4   29.0   52.5\\n', '  573  120 D A  E     +L  470   0J   1   -103,-2.3  -103,-3.4   -77,-0.2   -77,-0.3  -0.729  31.1 179.2 -90.2 134.3  -11.0   27.9   55.0\\n', '  574  121 D A  E    S+     0   0J   9    -79,-3.1  -108,-2.4    -2,-0.4     2,-0.3   0.564  70.4  14.3-105.0 -16.9  -10.2   25.0   57.4\\n', '  575  122 D A  E     -LN 465 495J   6    -80,-1.7   -80,-2.5  -110,-0.3     2,-0.4  -0.981  60.0-139.1-154.3 161.5  -13.4   25.0   59.3\\n', '  576  123 D I  E     -LN 464 494J   0   -112,-1.8  -112,-2.0    -2,-0.3     2,-0.3  -0.998  22.9-161.5-127.1 124.5  -17.0   26.4   59.2\\n', '  577  124 D A  E     -LN 463 493J   4    -84,-3.1   -84,-2.5    -2,-0.4     2,-0.4  -0.769   5.2-160.9-107.8 156.5  -18.5   27.6   62.5\\n', '  578  125 D Y  E     +LN 462 492J   0   -116,-2.8  -116,-1.7  -119,-0.4  -117,-1.2  -0.994  15.0 175.2-138.6 123.2  -22.2   28.2   63.4\\n', '  579  126 D N  E     + N   0 491J  59    -88,-2.9   -88,-3.5    -2,-0.4     2,-0.2  -0.992  13.2 130.2-137.7 144.4  -23.3   30.4   66.3\\n', '  580  127 D A        -     0   0   10     -2,-0.4     3,-0.5   -90,-0.2   -90,-0.1  -0.782  42.2-141.9 175.7 138.4  -26.6   31.6   67.7\\n', '  581  128 D E  S    S+     0   0  205    -92,-0.4     2,-0.3     1,-0.3    -1,-0.1   0.928 109.3  12.3 -73.0 -47.3  -28.4   31.9   71.0\\n', '  582  129 D N  S    S-     0   0   94    -93,-0.3   -94,-2.4   -98,-0.0    -1,-0.3  -0.671  97.7-141.3-131.2  75.7  -31.8   31.1   69.5\\n', '  583  130 D S        -     0   0   13     -3,-0.5   -93,-0.1    -2,-0.3   -94,-0.1   0.024  13.9-148.1 -45.7 136.5  -31.0   29.9   66.0\\n', '  584  131 D L  S    S+     0   0    0    -96,-0.3  -100,-2.1   -97,-0.1     2,-0.2   0.633  86.7  45.7 -76.7 -15.9  -33.1   30.9   63.1\\n', '  585  132 D F  B    S-Q  483   0K   8   -102,-0.2  -102,-0.2     1,-0.1     4,-0.1  -0.672  97.6 -77.8-123.6 176.3  -32.1   27.5   61.6\\n', '  586  133 D S        -     0   0   12   -104,-0.6    -1,-0.1    -2,-0.2    -2,-0.1  -0.216  47.4-102.3 -72.7 162.5  -31.9   23.8   62.6\\n', '  587  134 D S  S    S+     0   0   61   -128,-0.1  -128,-0.2    -4,-0.1     2,-0.2  -0.968 106.1  21.7-132.6 115.9  -29.1   22.3   64.7\\n', '  588  135 D P  S    S-     0   0   33      0, 0.0  -107,-0.4     0, 0.0     2,-0.4   0.616  84.8-154.7 -82.6-175.5  -27.2   20.8   63.1\\n', '  589  136 D V  E     -G  458   0I   1   -131,-1.7  -131,-2.2    -2,-0.2     2,-0.4  -0.884  25.6-112.1-113.4 149.7  -27.6   22.1   59.5\\n', '  590  137 D T  E     -GI 457 479I  38   -111,-2.9  -111,-3.0    -2,-0.4     2,-0.4  -0.694  34.3-170.5 -83.3 134.4  -26.6   19.8   56.7\\n', '  591  138 D V  E     -GI 456 478I   0   -135,-2.4  -135,-2.5    -2,-0.4     2,-0.7  -0.942  10.7-171.7-132.2 113.3  -23.5   20.9   54.8\\n', '  592  139 D D  E     -GI 455 477I  15   -115,-2.2  -115,-2.9    -2,-0.4     2,-0.5  -0.913  12.4-164.5-103.1 116.2  -22.3   19.3   51.6\\n', '  593  140 D V  E     - I   0 476I  17   -139,-2.9     2,-0.3    -2,-0.7  -117,-0.2  -0.906   4.4-168.3-106.7 123.9  -18.9   20.6   50.7\\n', '  594  141 D H  E     - I   0 475I  16   -119,-3.3  -119,-2.2    -2,-0.5  -140,-0.1  -0.820  25.8-145.8-110.3 149.7  -17.7   20.1   47.1\\n', '  595  142 D G  S    S+     0   0   42     -2,-0.3     2,-0.3     1,-0.2    -1,-0.1   0.643  92.2   5.8 -84.7 -17.5  -14.2   20.6   45.7\\n', '  596  143 D L  S    S-     0   0   76   -121,-0.1    -1,-0.2  -186,-0.0  -121,-0.1  -0.975 100.3 -75.4-158.4 156.6  -15.8   21.7   42.4\\n', '  597  144 D L  P>    -     0   0   27     -2,-0.3     3,-0.1  -124,-0.1  -123,-0.0  -0.350  58.2-131.1 -56.1 130.9  -19.3   22.4   41.1\\n', '  598  145 D P  PP    -     0   0   10      0, 0.0    -1,-0.1     0, 0.0    -4,-0.0  -0.284  34.4 -69.3 -82.7 172.2  -20.9   18.9   40.6\\n', '  599  146 D P  PP    -     0   0  104      0, 0.0  -147,-0.1     0, 0.0  -273,-0.0  -0.332  58.7-128.9 -58.0 141.4  -22.8   17.6   37.6\\n', '  600  147 D L  PP    -     0   0   37   -151,-0.1  -151,-0.1    -3,-0.1  -159,-0.0  -0.807   6.7-127.6-103.3 135.6  -26.1   19.5   37.3\\n', '  601  148 D P  P<    -     0   0   46      0, 0.0    -1,-0.0     0, 0.0  -160,-0.0  -0.381  33.0-109.9 -69.9 156.9  -29.5   17.9   37.0\\n', '  602  149 D P              0   0   84      0, 0.0  -160,-0.1     0, 0.0  -276,-0.1   0.718 360.0 360.0 -61.3 -18.3  -31.5   19.3   34.0\\n', '  603  150 D A              0   0  114   -155,-0.1  -300,-0.1  -160,-0.1  -160,-0.0   0.069 360.0 360.0-170.5 360.0  -33.8   20.9   36.7\\n']\n"
     ]
    },
    {
     "ename": "IndexError",
     "evalue": "list index out of range",
     "output_type": "error",
     "traceback": [
      "\u001b[31m---------------------------------------------------------------------------\u001b[39m",
      "\u001b[31mIndexError\u001b[39m                                Traceback (most recent call last)",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[260]\u001b[39m\u001b[32m, line 42\u001b[39m\n\u001b[32m     39\u001b[39m     \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[32m     40\u001b[39m         box.append(line)\n\u001b[32m---> \u001b[39m\u001b[32m42\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[43mbox\u001b[49m\u001b[43m[\u001b[49m\u001b[43m-\u001b[49m\u001b[32;43m1\u001b[39;49m\u001b[43m]\u001b[49m.startswith(\u001b[33m'\u001b[39m\u001b[33m  #  RESIDUE AA\u001b[39m\u001b[33m'\u001b[39m):\n\u001b[32m     43\u001b[39m     box = box[:-\u001b[32m1\u001b[39m]\n\u001b[32m     44\u001b[39m     \u001b[38;5;28;01mwith\u001b[39;00m \u001b[38;5;28mopen\u001b[39m(csar_pocket_dssp / \u001b[33mf\u001b[39m\u001b[33m'\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mpdb\u001b[38;5;132;01m}\u001b[39;00m\u001b[33m.dssp\u001b[39m\u001b[33m'\u001b[39m, \u001b[33m'\u001b[39m\u001b[33mw\u001b[39m\u001b[33m'\u001b[39m) \u001b[38;5;28;01mas\u001b[39;00m f:\n",
      "\u001b[31mIndexError\u001b[39m: list index out of range"
     ]
    }
   ],
   "source": [
    "success = []\n",
    "csar_pocket_dssp = Path('data/preprocessed/CSAR_DSSP_pocket')\n",
    "csar_pocket_dssp.mkdir(parents=True, exist_ok=True)\n",
    "dssps = Path('data/preprocessed/CSAR_DSSP').glob('*.dssp')\n",
    "for dssp in dssps:\n",
    "    pdb = dssp.stem\n",
    "    # if not pdb in pocket_inform:\n",
    "    #     continue\n",
    "    if pdb != '1uld':\n",
    "        continue\n",
    "    \n",
    "    box = []\n",
    "    new_num = 1\n",
    "    with open(dssp, 'r') as f:\n",
    "        parsing = False\n",
    "        for line in f:\n",
    "            if line.startswith('  #  RESIDUE AA'):\n",
    "                box.append(line)\n",
    "                parsing = True\n",
    "                continue\n",
    "            if parsing:\n",
    "                if line[13] == '!':\n",
    "                    continue\n",
    "                num = int(line[:5])\n",
    "                resid = int(line[6:10])\n",
    "                chain = line[11]\n",
    "                \n",
    "                key = f\"{chain}-{resid}\"\n",
    "                if key in pocket_inform[pdb]:\n",
    "                    box.append(f\"{new_num:>5}\" + line[5:])\n",
    "                    new_num += 1\n",
    "                else:\n",
    "                    if box[-1].startswith('  #  RESIDUE AA'):\n",
    "                        continue\n",
    "                    if box and box[-1][13] != '!':\n",
    "                        box.append(f\"{new_num:>5}{'!':>9}\\n\")\n",
    "                        new_num += 1\n",
    "            else:\n",
    "                box.append(line)\n",
    "        \n",
    "        if not box[-1].startswith('  #  RESIDUE AA'):\n",
    "            box = box[:-1]\n",
    "            with open(csar_pocket_dssp / f'{pdb}.dssp', 'w') as f:\n",
    "                for line in box:\n",
    "                    f.write(line)\n",
    "            success.append(pdb)\n",
    "print('done')\n",
    "print(len(success))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 259,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['==== Secondary Structure Definition by the program DSSP, NKI version 4.5.1                         ==== DATE=2025-05-13        .\\n',\n",
       " 'REFERENCE W. KABSCH AND C.SANDER, BIOPOLYMERS 22 (1983) 2577-2637                                                              .\\n',\n",
       " 'HEADER    SUGAR BINDING PROTEIN                   12-SEP-03   1ULD                                                             .\\n',\n",
       " 'COMPND    MOL_ID: 1; MOLECULE: galectin-2; CHAIN: A, B, C, D; SYNONYM: CGL2; ENGINEERED: YES                                   .\\n',\n",
       " 'SOURCE    MOL_ID: 1; GENE: cgl2; ORGANISM_SCIENTIFIC: Coprinopsis cinerea; ORGANISM_TAXID: 5346; EXPRESSION_SYSTEM: Sacchar... .\\n',\n",
       " 'AUTHOR    P.J.Walser; P.W.Haebel; M.Kuenzler; U.Kues; M.Aebi; N.Ban                                                            .\\n',\n",
       " '  600  4  0  0  0 TOTAL NUMBER OF RESIDUES, NUMBER OF CHAINS, NUMBER OF SS-BRIDGES(TOTAL,INTRACHAIN,INTERCHAIN)                .\\n',\n",
       " ' 27366.9   ACCESSIBLE SURFACE OF PROTEIN (ANGSTROM**2)                                                                         .\\n',\n",
       " '  396 66.0   TOTAL NUMBER OF HYDROGEN BONDS OF TYPE O(I)-->H-N(J)  , SAME NUMBER PER 100 RESIDUES                              .\\n',\n",
       " '    0  0.0   TOTAL NUMBER OF HYDROGEN BONDS IN     PARALLEL BRIDGES, SAME NUMBER PER 100 RESIDUES                              .\\n',\n",
       " '  265 44.2   TOTAL NUMBER OF HYDROGEN BONDS IN ANTIPARALLEL BRIDGES, SAME NUMBER PER 100 RESIDUES                              .\\n',\n",
       " '    4  0.7   TOTAL NUMBER OF HYDROGEN BONDS OF TYPE O(I)-->H-N(I-5), SAME NUMBER PER 100 RESIDUES                              .\\n',\n",
       " '    4  0.7   TOTAL NUMBER OF HYDROGEN BONDS OF TYPE O(I)-->H-N(I-4), SAME NUMBER PER 100 RESIDUES                              .\\n',\n",
       " '    4  0.7   TOTAL NUMBER OF HYDROGEN BONDS OF TYPE O(I)-->H-N(I-3), SAME NUMBER PER 100 RESIDUES                              .\\n',\n",
       " '    1  0.2   TOTAL NUMBER OF HYDROGEN BONDS OF TYPE O(I)-->H-N(I-2), SAME NUMBER PER 100 RESIDUES                              .\\n',\n",
       " '    0  0.0   TOTAL NUMBER OF HYDROGEN BONDS OF TYPE O(I)-->H-N(I-1), SAME NUMBER PER 100 RESIDUES                              .\\n',\n",
       " '    0  0.0   TOTAL NUMBER OF HYDROGEN BONDS OF TYPE O(I)-->H-N(I+0), SAME NUMBER PER 100 RESIDUES                              .\\n',\n",
       " '    0  0.0   TOTAL NUMBER OF HYDROGEN BONDS OF TYPE O(I)-->H-N(I+1), SAME NUMBER PER 100 RESIDUES                              .\\n',\n",
       " '   79 13.2   TOTAL NUMBER OF HYDROGEN BONDS OF TYPE O(I)-->H-N(I+2), SAME NUMBER PER 100 RESIDUES                              .\\n',\n",
       " '   28  4.7   TOTAL NUMBER OF HYDROGEN BONDS OF TYPE O(I)-->H-N(I+3), SAME NUMBER PER 100 RESIDUES                              .\\n',\n",
       " '    3  0.5   TOTAL NUMBER OF HYDROGEN BONDS OF TYPE O(I)-->H-N(I+4), SAME NUMBER PER 100 RESIDUES                              .\\n',\n",
       " '    4  0.7   TOTAL NUMBER OF HYDROGEN BONDS OF TYPE O(I)-->H-N(I+5), SAME NUMBER PER 100 RESIDUES                              .\\n',\n",
       " '  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30     *** HISTOGRAMS OF ***           .\\n',\n",
       " '  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0    RESIDUES PER ALPHA HELIX         .\\n',\n",
       " '  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0    PARALLEL BRIDGES PER LADDER      .\\n',\n",
       " '  3  0  0 16 13 11  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0    ANTIPARALLEL BRIDGES PER LADDER  .\\n',\n",
       " '  0  0  0  0  8  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0    LADDERS PER SHEET                .\\n',\n",
       " '  #  RESIDUE AA STRUCTURE BP1 BP2  ACC     N-H-->O    O-->H-N    N-H-->O    O-->H-N    TCO  KAPPA ALPHA  PHI   PSI    X-CA   Y-CA   Z-CA\\n']"
      ]
     },
     "execution_count": 259,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "box"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 220,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import glob\n",
    "import re\n",
    "import os\n",
    "import sys\n",
    "\n",
    "pred_sse = Path('data/preprocessed/PredSSE_pocket')\n",
    "pred_sse.mkdir(parents=True, exist_ok=True)\n",
    "\n",
    "f=open(r\"data/preprocessed/csar.out.ss8\", \"r\")   \n",
    "for line in f.readlines():\n",
    "    if '>' in line:\n",
    "        pdbname=line.strip().split('>')[1]    \n",
    "    else:\n",
    "        num = -1\n",
    "        i= -1\n",
    "        sse =''\n",
    "        for s in line:\n",
    "            num +=1 \n",
    "#             print(num)\n",
    "            sse= sse+s\n",
    "        try:\n",
    "            with open(r'data/preprocessed/CSAR_DSSP_pocket/%s.dssp'%(pdbname), \"r\") as fg:\n",
    "                for _ in range(28):\n",
    "                    next(fg)\n",
    "                for line1 in fg:\n",
    "                    if line1[13]=='!':\n",
    "                        continue\n",
    "                    else:\n",
    "                        i +=1\n",
    "                        if i <= num:\n",
    "                            line1=line1[:16] + sse[i] + line1[17:]\n",
    "                            with open(r'data/preprocessed/PredSSE_pocket/%s_pocket.dssp'%(pdbname),\"a+\") as fw:\n",
    "                                fw.write(line1)\n",
    "        except:\n",
    "            continue\n",
    "f.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import glob\n",
    "import re\n",
    "import os\n",
    "import sys\n",
    "\n",
    "pred_sse = Path('data/preprocessed/PredSSE')\n",
    "pred_sse.mkdir(parents=True, exist_ok=True)\n",
    "\n",
    "f=open(r\"data/preprocessed/csar.out.ss8\", \"r\")   \n",
    "for line in f.readlines():\n",
    "    if '>' in line:\n",
    "        pdbname=line.strip().split('>')[1]    \n",
    "    else:\n",
    "        num = -1\n",
    "        i= -1\n",
    "        sse =''\n",
    "        for s in line:\n",
    "            num +=1 \n",
    "#             print(num)\n",
    "            sse= sse+s\n",
    "        with open(r'data/preprocessed/CSAR_DSSP/%s.dssp'%(pdbname), \"r\") as fg:\n",
    "            for _ in range(28):\n",
    "                next(fg)\n",
    "            for line1 in fg:\n",
    "                if line1[13]=='!':\n",
    "                    continue\n",
    "                else:\n",
    "                    i +=1\n",
    "                    if i <= num:\n",
    "                        line1=line1[:16] + sse[i] + line1[17:]\n",
    "                        with open(r'data/preprocessed/PredSSE/%s_protein.dssp'%(pdbname),\"a+\") as fw:\n",
    "                            fw.write(line1)\n",
    "f.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 233,
   "metadata": {},
   "outputs": [],
   "source": [
    "from tqdm import tqdm\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import sys\n",
    "import glob\n",
    "import re\n",
    "import os\n",
    "\n",
    "columns_g = ['residue','chain','aa','structure']\n",
    "columns_l = ['residue','chain','aa','structure']\n",
    "structure = ('B', 'C', 'E', 'G', 'H', 'I', 'S', 'T')\n",
    "aa = ('G', 'A', 'V', 'L', 'I', 'M', 'F', 'P', 'W', 'S', 'T', 'Y', 'C', 'Q', 'N', 'D', 'E', 'K', 'R', 'H', 'X')\n",
    "\n",
    "c1 = {\n",
    "    'non_polar' : ('G','A','V','L','I','M','F','P','W'),\n",
    "    'polar' : ('S','T','Y','C','Q','N'),\n",
    "    'acidic' : ('D','E'),\n",
    "    'basic' : ('K','R','H'),\n",
    "}\n",
    "\n",
    "def f1(aa, key):\n",
    "    if aa=='X':\n",
    "        return 1/len(c1)\n",
    "    return 1 if aa in c1[key] else 0\n",
    "\n",
    "\n",
    "c2 = {\n",
    "    1 : ('A','G','V'),\n",
    "    2 : ('I','L','F','P'),\n",
    "    3 : ('Y','M','T','S'),\n",
    "    4 : ('H','N','Q','W'),\n",
    "    5 : ('R','K'),\n",
    "    6 : ('D','E'),\n",
    "    7 : ('C',)\n",
    "}\n",
    "\n",
    "def f2(aa, key):\n",
    "    if aa=='X':\n",
    "        return 1/len(c2)\n",
    "    return 1 if aa in c2[key] else 0\n",
    "\n",
    "\n",
    "def idx_df_l_init(df_g):\n",
    "    chains = df_g.chain.unique()\n",
    "    \n",
    "    if len(chains) == 1:\n",
    "        return lambda row: chains[0]+row.residue\n",
    "    \n",
    "    i = -1\n",
    "    cidx = 0\n",
    "    \n",
    "    def idx_df_l(row):\n",
    "        nonlocal i\n",
    "        nonlocal df_g\n",
    "\n",
    "        while True:\n",
    "            i+=1\n",
    "            row_g = df_g.iloc[i]\n",
    "            \n",
    "            if row_g.residue==row.residue:\n",
    "                if row.aa != row_g.aa:\n",
    "                    continue\n",
    "                if row.chain and row.chain!=row_g.chain:\n",
    "                    continue\n",
    "                break\n",
    "            \n",
    "        return row_g.chain+row.residue\n",
    "    return idx_df_l"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 245,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 125/125 [00:02<00:00, 58.16it/s]\n"
     ]
    }
   ],
   "source": [
    "f_g = glob.glob('data/preprocessed/PredSSE/*')\n",
    "f_l = glob.glob('data/preprocessed/PredSSE_pocket/*')\n",
    "\n",
    "demo = set(pd.read_csv('data/preprocessed/csar_proteins.csv').id)\n",
    "demo = set(i for i in demo if i in success)\n",
    "dataset = {\n",
    "    'demo': demo.copy(),\n",
    "}\n",
    "columns_s = ['id','seq']\n",
    "seq_dfs = {\n",
    "    'demo': pd.DataFrame(columns=columns_s),\n",
    "}\n",
    "pocket_dfs = {\n",
    "    'demo': pd.DataFrame(columns=columns_s),\n",
    "}\n",
    "other_seq = set()\n",
    "\n",
    "\n",
    "for fi_g in tqdm(f_g):\n",
    "    seq_name = os.path.basename(fi_g)[:4]\n",
    "    \n",
    "    flag = True\n",
    "    phase = None\n",
    "    for k in dataset:\n",
    "        if seq_name in dataset[k]:\n",
    "            dataset[k].remove(seq_name)\n",
    "            phase = k\n",
    "            flag = False\n",
    "            break\n",
    "    if flag:\n",
    "        other_seq.add(seq_name)\n",
    "        continue\n",
    "    fi_l = os.path.join(os.path.dirname(fi_g),os.path.basename(fi_g)).replace('protein','pocket')\n",
    "    if not os.path.exists(fi_l):\n",
    "        print(f'{fi_l} not exists')\n",
    "        continue\n",
    "   \n",
    "    df_g = pd.DataFrame(columns=columns_g)\n",
    "    with open(fi_g) as f:\n",
    "        for line in f:\n",
    "            if line[13]=='!':\n",
    "                continue\n",
    "            else:\n",
    "                row = pd.Series([line[5:11].strip(),line[11:12].strip(),line[12:14].strip(),line[16:17].strip()],\n",
    "                           index=columns_g)\n",
    "            df_g = pd.concat([df_g, row.to_frame().T], ignore_index=True)\n",
    "            # df_g = df_g.append(row, ignore_index=True)\n",
    "     \n",
    "    df_g.aa = df_g.aa.apply(lambda x: 'C' if re.match('[A-Z]',x) is None else x)\n",
    "    df_g.structure = df_g.structure.apply(lambda x: 'C' if re.match('[A-Z]',x) is None else x)\n",
    "\n",
    "    assert len(df_g.structure.unique())<= 8\n",
    "    assert len(df_g.aa.unique())<= 21\n",
    "    \n",
    "    for key in c1:\n",
    "        new_feature = df_g.aa.apply(lambda x:f1(x,key))\n",
    "        df_g = eval(f'df_g.assign({key}=new_feature)')\n",
    "    \n",
    "    for key in c2:\n",
    "        new_feature = df_g.aa.apply(lambda x:f2(x,key))\n",
    "        df_g = eval(f'df_g.assign(c2_{key}=new_feature)')\n",
    "    \n",
    "    for si in structure:\n",
    "        new_feature = df_g.structure.apply(lambda x:1 if x==si else 0)\n",
    "        df_g = eval(f'df_g.assign(s2_{si}=new_feature)')\n",
    "    \n",
    "    for ai in aa:\n",
    "        new_feature = df_g.aa.apply(lambda x:1 if x==ai else 0)\n",
    "        df_g = eval(f'df_g.assign(a_{ai}=new_feature)')\n",
    "    \n",
    "    # seq_dfs[phase] = seq_dfs[phase].append(pd.Series([seq_name,''.join(df_g.aa)],index=columns_s),ignore_index=True)\n",
    "    seq_dfs[phase] = pd.concat([seq_dfs[phase], pd.Series([seq_name,''.join(df_g.aa)],index=columns_s)], ignore_index=True)\n",
    "    \n",
    "    df_g = df_g.assign(idx=df_g.chain+df_g.residue)\n",
    "    \n",
    "    \n",
    "    df_l = pd.DataFrame(columns=columns_l)\n",
    "    with open(fi_l) as f:\n",
    "        for line in f:\n",
    "            if line[13]=='!':\n",
    "                continue\n",
    "            else:\n",
    "                row = pd.Series([line[5:11].strip(),line[11:12].strip(),line[12:14].strip(),line[16:17].strip()],\n",
    "                           index=columns_l)\n",
    "            \n",
    "            # df_l = df_l.append(row, ignore_index=True)\n",
    "            df_l = pd.concat([df_l, row.to_frame().T], ignore_index=True)\n",
    "    \n",
    "    df_l.aa = df_l.aa.apply(lambda x: 'C' if re.match('[A-Z]',x) is None else x)\n",
    "    \n",
    "    # pocket_dfs[phase] = pocket_dfs[phase].append(pd.Series([seq_name,''.join(df_l.aa)],index=columns_s),ignore_index=True)\n",
    "    pocket_dfs[phase] = pd.concat([pocket_dfs[phase], pd.Series([seq_name,''.join(df_l.aa)],index=columns_s)], ignore_index=True)\n",
    "    \n",
    "    try:\n",
    "        df_l = df_l.assign(idx=df_l.apply(idx_df_l_init(df_g),axis=1))\n",
    "    except:\n",
    "        print(f\"{seq_name} error: {sys.exc_info()[0]}\")\n",
    "        continue\n",
    "    \n",
    "    df_g.drop('aa',axis=1,inplace=True)\n",
    "    df_g.drop('structure',axis=1,inplace=True)\n",
    "    df_g.drop('residue',axis=1,inplace=True)\n",
    "    df_g.drop('chain',axis=1,inplace=True)\n",
    "    \n",
    "    df_g.to_csv(f'{phase}/global/{seq_name}.csv')\n",
    "    \n",
    "    df_l.drop('aa',axis=1,inplace=True)\n",
    "    df_l.drop('structure',axis=1,inplace=True)\n",
    "    df_l.drop('residue',axis=1,inplace=True)\n",
    "    df_l.drop('chain',axis=1,inplace=True)\n",
    "    \n",
    "    df_l = df_l.merge(df_g)\n",
    "    \n",
    "    df_l.to_csv(f'{phase}/pocket/{seq_name}.csv')\n",
    "    \n",
    "    del df_g\n",
    "    del df_l\n",
    "\n",
    "for k in pocket_dfs:\n",
    "    pocket_dfs[k].to_csv(f'{k}_pocket_.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 248,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[102, 92, 85, 83, 79, 79, 78, 77, 77, 75, 74, 73, 72, 70, 70, 68, 68, 67, 67, 67, 66, 65, 65, 65, 62, 61, 60, 59, 58, 57, 57, 49, 49, 48, 47, 45, 45, 44, 42, 41, 39, 38, 37, 36, 36, 20, 19, 13, 11, 6, 1]\n"
     ]
    }
   ],
   "source": [
    "lens = []\n",
    "for fi in Path('demo/pocket').glob('*.csv'):\n",
    "    df = pd.read_csv(fi)\n",
    "    lens.append(len(df))\n",
    "print(sorted(lens, reverse=True))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 249,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "24"
      ]
     },
     "execution_count": 249,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len([102, 92, 85, 83, 79, 79, 78, 77, 77, 75, 74, 73, 72, 70, 70, 68, 68, 67, 67, 67, 66, 65, 65, 65])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 254,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1q6e\n",
      "1q6g\n",
      "1s7y\n",
      "1txf\n",
      "1y93\n",
      "1zhx\n",
      "2dm5\n",
      "4ubp\n",
      "[102, 78, 77, 75, 68, 67, 20, 11]\n"
     ]
    }
   ],
   "source": [
    "lens = []\n",
    "for fi in Path('demo/pocket').glob('*.csv'):\n",
    "    if fi.stem.split('_')[0] in no_sse_pdb:\n",
    "        print(fi.stem.split('_')[0])\n",
    "        df = pd.read_csv(fi)\n",
    "        lens.append(len(df))\n",
    "print(sorted(lens, reverse=True))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 255,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[137, 125, 124, 124, 123, 121, 120, 119, 119, 119, 118, 116, 116, 116, 116, 115, 115, 115, 114, 114, 114, 114, 114, 113, 112, 112, 112, 112, 111, 111, 111, 111, 111, 110, 110, 110, 109, 109, 109, 108, 108, 107, 107, 107, 107, 106, 106, 106, 106, 106, 106, 106, 105, 104, 103, 103, 103, 101, 100, 100, 99, 99, 98, 98, 98, 98, 98, 96, 95, 95, 95, 94, 94, 94, 94, 94, 94, 94, 93, 93, 93, 93, 92, 92, 92, 92, 92, 92, 91, 91, 91, 90, 90, 90, 90, 90, 90, 90, 90, 89, 89, 89, 89, 89, 89, 89, 88, 88, 88, 88, 88, 88, 88, 87, 87, 87, 87, 87, 87, 87, 87, 86, 86, 86, 86, 86, 86, 86, 86, 86, 85, 85, 85, 85, 85, 84, 84, 84, 84, 84, 84, 84, 84, 84, 84, 84, 84, 84, 84, 84, 83, 83, 83, 83, 83, 83, 83, 83, 83, 83, 83, 83, 83, 83, 83, 83, 83, 83, 82, 82, 82, 82, 82, 82, 82, 82, 82, 82, 82, 81, 81, 81, 81, 81, 81, 81, 81, 81, 81, 81, 81, 81, 81, 81, 81, 81, 81, 81, 81, 81, 81, 81, 81, 80, 80, 80, 80, 80, 80, 80, 80, 80, 80, 80, 80, 80, 80, 80, 80, 79, 79, 79, 79, 79, 79, 79, 79, 79, 79, 79, 79, 79, 79, 79, 79, 79, 78, 78, 78, 78, 78, 78, 78, 78, 78, 78, 78, 78, 78, 78, 78, 78, 78, 77, 77, 77, 77, 77, 77, 77, 77, 77, 77, 77, 77, 77, 77, 77, 77, 77, 77, 77, 77, 77, 77, 77, 77, 77, 77, 77, 77, 77, 77, 76, 76, 76, 76, 76, 76, 76, 76, 76, 76, 76, 76, 76, 76, 76, 76, 76, 76, 76, 76, 76, 76, 76, 76, 76, 76, 76, 76, 76, 76, 76, 76, 76, 76, 75, 75, 75, 75, 75, 75, 75, 75, 75, 75, 75, 75, 75, 75, 75, 75, 75, 75, 75, 75, 75, 75, 75, 75, 75, 75, 75, 75, 75, 74, 74, 74, 74, 74, 74, 74, 74, 74, 74, 74, 74, 74, 74, 74, 74, 74, 74, 74, 74, 74, 74, 74, 74, 74, 74, 74, 74, 74, 74, 74, 74, 74, 74, 74, 74, 74, 74, 74, 74, 74, 73, 73, 73, 73, 73, 73, 73, 73, 73, 73, 73, 73, 73, 73, 73, 73, 73, 73, 73, 73, 73, 73, 73, 73, 73, 73, 73, 73, 73, 73, 73, 73, 73, 73, 73, 73, 73, 73, 73, 73, 73, 73, 73, 73, 73, 73, 72, 72, 72, 72, 72, 72, 72, 72, 72, 72, 72, 72, 72, 72, 72, 72, 72, 72, 72, 72, 72, 72, 72, 72, 72, 72, 72, 72, 72, 72, 72, 72, 72, 72, 72, 72, 72, 72, 72, 72, 72, 72, 72, 72, 72, 72, 72, 71, 71, 71, 71, 71, 71, 71, 71, 71, 71, 71, 71, 71, 71, 71, 71, 71, 71, 71, 71, 71, 71, 71, 71, 71, 71, 71, 71, 71, 71, 71, 71, 71, 71, 71, 71, 71, 71, 71, 71, 71, 71, 71, 71, 71, 71, 71, 71, 70, 70, 70, 70, 70, 70, 70, 70, 70, 70, 70, 70, 70, 70, 70, 70, 70, 70, 70, 70, 70, 70, 70, 70, 70, 70, 70, 70, 70, 70, 70, 70, 70, 70, 70, 70, 70, 70, 70, 70, 70, 70, 70, 70, 70, 70, 70, 70, 70, 70, 70, 70, 70, 70, 70, 70, 70, 70, 70, 70, 70, 70, 70, 70, 70, 70, 70, 70, 70, 70, 70, 70, 70, 70, 70, 69, 69, 69, 69, 69, 69, 69, 69, 69, 69, 69, 69, 69, 69, 69, 69, 69, 69, 69, 69, 69, 69, 69, 69, 69, 69, 69, 69, 69, 69, 69, 69, 69, 69, 69, 69, 69, 69, 69, 69, 69, 69, 69, 69, 69, 69, 69, 69, 69, 69, 69, 69, 69, 69, 69, 69, 69, 69, 69, 69, 69, 69, 69, 69, 69, 69, 69, 69, 69, 69, 69, 69, 69, 68, 68, 68, 68, 68, 68, 68, 68, 68, 68, 68, 68, 68, 68, 68, 68, 68, 68, 68, 68, 68, 68, 68, 68, 68, 68, 68, 68, 68, 68, 68, 68, 68, 68, 68, 68, 68, 68, 68, 68, 68, 68, 68, 68, 68, 68, 68, 68, 68, 68, 68, 68, 68, 68, 68, 68, 68, 68, 68, 68, 68, 68, 68, 68, 68, 68, 68, 68, 68, 68, 68, 68, 68, 68, 68, 68, 68, 68, 68, 68, 68, 68, 68, 68, 68, 68, 68, 68, 68, 68, 68, 68, 68, 68, 68, 68, 68, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 61, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 58, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 55, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 54, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 49, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 47, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 45, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 29, 29, 29, 29, 29, 29, 29, 29, 29, 29, 29, 29, 29, 29, 29, 29, 29, 29, 29, 29, 29, 29, 29, 29, 29, 29, 29, 29, 29, 29, 29, 29, 29, 29, 29, 29, 29, 29, 29, 29, 29, 29, 29, 29, 29, 29, 29, 29, 29, 29, 29, 29, 29, 29, 29, 29, 29, 29, 29, 29, 29, 29, 29, 29, 29, 29, 29, 29, 29, 29, 29, 29, 29, 29, 29, 29, 29, 29, 29, 29, 29, 29, 29, 29, 29, 29, 29, 29, 29, 29, 29, 29, 29, 29, 29, 29, 29, 29, 29, 29, 29, 29, 29, 29, 29, 28, 28, 28, 28, 28, 28, 28, 28, 28, 28, 28, 28, 28, 28, 28, 28, 28, 28, 28, 28, 28, 28, 28, 28, 28, 28, 28, 28, 28, 28, 28, 28, 28, 28, 28, 28, 28, 28, 28, 28, 28, 28, 28, 28, 28, 28, 28, 28, 28, 28, 28, 28, 28, 28, 28, 28, 28, 28, 28, 28, 28, 28, 28, 28, 28, 28, 28, 28, 28, 28, 28, 28, 28, 28, 28, 28, 28, 28, 28, 28, 28, 28, 28, 28, 28, 28, 28, 28, 28, 28, 28, 28, 28, 27, 27, 27, 27, 27, 27, 27, 27, 27, 27, 27, 27, 27, 27, 27, 27, 27, 27, 27, 27, 27, 27, 27, 27, 27, 27, 27, 27, 27, 27, 27, 27, 27, 27, 27, 27, 27, 27, 27, 27, 27, 27, 27, 27, 27, 27, 27, 27, 27, 27, 27, 27, 27, 27, 27, 27, 27, 27, 27, 27, 27, 27, 27, 27, 27, 27, 27, 27, 27, 27, 27, 27, 27, 27, 27, 27, 27, 27, 26, 26, 26, 26, 26, 26, 26, 26, 26, 26, 26, 26, 26, 26, 26, 26, 26, 26, 26, 26, 26, 26, 26, 26, 26, 26, 26, 26, 26, 26, 26, 26, 26, 26, 26, 26, 26, 26, 26, 26, 26, 26, 26, 26, 26, 26, 26, 26, 26, 26, 26, 26, 26, 26, 26, 26, 26, 26, 26, 26, 26, 26, 26, 26, 26, 26, 26, 26, 26, 26, 26, 26, 26, 26, 26, 26, 26, 26, 26, 26, 26, 26, 26, 26, 26, 26, 26, 26, 25, 25, 25, 25, 25, 25, 25, 25, 25, 25, 25, 25, 25, 25, 25, 25, 25, 25, 25, 25, 25, 25, 25, 25, 25, 25, 25, 25, 25, 25, 25, 25, 25, 25, 25, 25, 25, 25, 25, 25, 25, 25, 25, 25, 25, 25, 25, 25, 25, 25, 25, 25, 25, 25, 25, 25, 25, 25, 25, 25, 25, 25, 25, 25, 25, 25, 25, 25, 25, 25, 25, 25, 25, 25, 25, 25, 25, 25, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 23, 23, 23, 23, 23, 23, 23, 23, 23, 23, 23, 23, 23, 23, 23, 23, 23, 23, 23, 23, 23, 23, 23, 23, 23, 23, 23, 23, 23, 23, 23, 23, 23, 23, 23, 23, 23, 23, 23, 23, 23, 23, 23, 23, 23, 23, 22, 22, 22, 22, 22, 22, 22, 22, 22, 22, 22, 22, 22, 22, 22, 22, 22, 22, 22, 22, 22, 22, 22, 22, 22, 22, 22, 22, 22, 22, 22, 22, 22, 22, 22, 22, 22, 22, 22, 22, 22, 22, 22, 22, 22, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 20, 20, 20, 20, 20, 20, 20, 20, 20, 20, 20, 20, 20, 20, 19, 19, 19, 19, 19, 19, 19, 19, 19, 19, 19, 19, 19, 19, 19, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 17, 17, 17, 17, 17, 17, 17, 17, 17, 17, 17, 17, 16, 16, 16, 16, 16, 16, 16, 16, 16, 15, 15, 15, 15, 15, 15, 15, 14, 14, 14, 14, 14, 14, 14, 14, 13, 13, 13, 13, 13, 13, 12, 12, 12, 12, 11, 11, 10, 10, 10, 9, 9, 6, 6, 6, 4, 0]\n"
     ]
    }
   ],
   "source": [
    "lens = []\n",
    "for fi in Path('data/preprocessed/sse/pocket').glob('*.csv'):\n",
    "    df = pd.read_csv(fi)\n",
    "    lens.append(len(df))\n",
    "print(sorted(lens, reverse=True))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 263,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "SET\n",
       "General    7412\n",
       "Refined    2732\n",
       "CORE        229\n",
       "CSAR         62\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 263,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fff = ff[~ff['PDB'].isin(no_sse_pdb)].reset_index(drop=True)\n",
    "fff['SET'].value_counts()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 264,
   "metadata": {},
   "outputs": [],
   "source": [
    "fff.to_csv('data/new_data_info.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 271,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "python main.py --model DeepDTA --project DeepDTA_rmse_mse_mean_ap-True_sh-True_epoch20 --n_epoch 20 --loss mse_mean --apex --scheduler --aim rmse\n",
      "python main.py --model DeepDTA --project DeepDTA_rmse_mse_mean_ap-True_sh-True_epoch50 --n_epoch 50 --loss mse_mean --apex --scheduler --aim rmse\n",
      "python main.py --model DeepDTA --project DeepDTA_rmse_mse_mean_ap-True_sh-False_epoch20 --n_epoch 20 --loss mse_mean --apex --scheduler --aim rmse\n",
      "python main.py --model DeepDTA --project DeepDTA_rmse_mse_mean_ap-True_sh-False_epoch50 --n_epoch 50 --loss mse_mean --apex --scheduler --aim rmse\n",
      "python main.py --model DeepDTA --project DeepDTA_rmse_mse_mean_ap-False_sh-True_epoch20 --n_epoch 20 --loss mse_mean --apex --scheduler --aim rmse\n",
      "python main.py --model DeepDTA --project DeepDTA_rmse_mse_mean_ap-False_sh-True_epoch50 --n_epoch 50 --loss mse_mean --apex --scheduler --aim rmse\n",
      "python main.py --model DeepDTA --project DeepDTA_rmse_mse_mean_ap-False_sh-False_epoch20 --n_epoch 20 --loss mse_mean --apex --scheduler --aim rmse\n",
      "python main.py --model DeepDTA --project DeepDTA_rmse_mse_mean_ap-False_sh-False_epoch50 --n_epoch 50 --loss mse_mean --apex --scheduler --aim rmse\n",
      "python main.py --model DeepDTA --project DeepDTA_rmse_mse_sum_ap-True_sh-True_epoch20 --n_epoch 20 --loss mse_sum --apex --scheduler --aim rmse\n",
      "python main.py --model DeepDTA --project DeepDTA_rmse_mse_sum_ap-True_sh-True_epoch50 --n_epoch 50 --loss mse_sum --apex --scheduler --aim rmse\n",
      "python main.py --model DeepDTA --project DeepDTA_rmse_mse_sum_ap-True_sh-False_epoch20 --n_epoch 20 --loss mse_sum --apex --scheduler --aim rmse\n",
      "python main.py --model DeepDTA --project DeepDTA_rmse_mse_sum_ap-True_sh-False_epoch50 --n_epoch 50 --loss mse_sum --apex --scheduler --aim rmse\n",
      "python main.py --model DeepDTA --project DeepDTA_rmse_mse_sum_ap-False_sh-True_epoch20 --n_epoch 20 --loss mse_sum --apex --scheduler --aim rmse\n",
      "python main.py --model DeepDTA --project DeepDTA_rmse_mse_sum_ap-False_sh-True_epoch50 --n_epoch 50 --loss mse_sum --apex --scheduler --aim rmse\n",
      "python main.py --model DeepDTA --project DeepDTA_rmse_mse_sum_ap-False_sh-False_epoch20 --n_epoch 20 --loss mse_sum --apex --scheduler --aim rmse\n",
      "python main.py --model DeepDTA --project DeepDTA_rmse_mse_sum_ap-False_sh-False_epoch50 --n_epoch 50 --loss mse_sum --apex --scheduler --aim rmse\n",
      "python main.py --model DeepDTAF --project DeepDTAF_rmse_mse_mean_ap-True_sh-True_epoch20 --n_epoch 20 --loss mse_mean --apex --scheduler --aim rmse\n",
      "python main.py --model DeepDTAF --project DeepDTAF_rmse_mse_mean_ap-True_sh-True_epoch50 --n_epoch 50 --loss mse_mean --apex --scheduler --aim rmse\n",
      "python main.py --model DeepDTAF --project DeepDTAF_rmse_mse_mean_ap-True_sh-False_epoch20 --n_epoch 20 --loss mse_mean --apex --scheduler --aim rmse\n",
      "python main.py --model DeepDTAF --project DeepDTAF_rmse_mse_mean_ap-True_sh-False_epoch50 --n_epoch 50 --loss mse_mean --apex --scheduler --aim rmse\n",
      "python main.py --model DeepDTAF --project DeepDTAF_rmse_mse_mean_ap-False_sh-True_epoch20 --n_epoch 20 --loss mse_mean --apex --scheduler --aim rmse\n",
      "python main.py --model DeepDTAF --project DeepDTAF_rmse_mse_mean_ap-False_sh-True_epoch50 --n_epoch 50 --loss mse_mean --apex --scheduler --aim rmse\n",
      "python main.py --model DeepDTAF --project DeepDTAF_rmse_mse_mean_ap-False_sh-False_epoch20 --n_epoch 20 --loss mse_mean --apex --scheduler --aim rmse\n",
      "python main.py --model DeepDTAF --project DeepDTAF_rmse_mse_mean_ap-False_sh-False_epoch50 --n_epoch 50 --loss mse_mean --apex --scheduler --aim rmse\n",
      "python main.py --model DeepDTAF --project DeepDTAF_rmse_mse_sum_ap-True_sh-True_epoch20 --n_epoch 20 --loss mse_sum --apex --scheduler --aim rmse\n",
      "python main.py --model DeepDTAF --project DeepDTAF_rmse_mse_sum_ap-True_sh-True_epoch50 --n_epoch 50 --loss mse_sum --apex --scheduler --aim rmse\n",
      "python main.py --model DeepDTAF --project DeepDTAF_rmse_mse_sum_ap-True_sh-False_epoch20 --n_epoch 20 --loss mse_sum --apex --scheduler --aim rmse\n",
      "python main.py --model DeepDTAF --project DeepDTAF_rmse_mse_sum_ap-True_sh-False_epoch50 --n_epoch 50 --loss mse_sum --apex --scheduler --aim rmse\n",
      "python main.py --model DeepDTAF --project DeepDTAF_rmse_mse_sum_ap-False_sh-True_epoch20 --n_epoch 20 --loss mse_sum --apex --scheduler --aim rmse\n",
      "python main.py --model DeepDTAF --project DeepDTAF_rmse_mse_sum_ap-False_sh-True_epoch50 --n_epoch 50 --loss mse_sum --apex --scheduler --aim rmse\n",
      "python main.py --model DeepDTAF --project DeepDTAF_rmse_mse_sum_ap-False_sh-False_epoch20 --n_epoch 20 --loss mse_sum --apex --scheduler --aim rmse\n",
      "python main.py --model DeepDTAF --project DeepDTAF_rmse_mse_sum_ap-False_sh-False_epoch50 --n_epoch 50 --loss mse_sum --apex --scheduler --aim rmse\n",
      "python main.py --model CAPLA --project CAPLA_rmse_mse_mean_ap-True_sh-True_epoch20 --n_epoch 20 --loss mse_mean --apex --scheduler --aim rmse\n",
      "python main.py --model CAPLA --project CAPLA_rmse_mse_mean_ap-True_sh-True_epoch50 --n_epoch 50 --loss mse_mean --apex --scheduler --aim rmse\n",
      "python main.py --model CAPLA --project CAPLA_rmse_mse_mean_ap-True_sh-False_epoch20 --n_epoch 20 --loss mse_mean --apex --scheduler --aim rmse\n",
      "python main.py --model CAPLA --project CAPLA_rmse_mse_mean_ap-True_sh-False_epoch50 --n_epoch 50 --loss mse_mean --apex --scheduler --aim rmse\n",
      "python main.py --model CAPLA --project CAPLA_rmse_mse_mean_ap-False_sh-True_epoch20 --n_epoch 20 --loss mse_mean --apex --scheduler --aim rmse\n",
      "python main.py --model CAPLA --project CAPLA_rmse_mse_mean_ap-False_sh-True_epoch50 --n_epoch 50 --loss mse_mean --apex --scheduler --aim rmse\n",
      "python main.py --model CAPLA --project CAPLA_rmse_mse_mean_ap-False_sh-False_epoch20 --n_epoch 20 --loss mse_mean --apex --scheduler --aim rmse\n",
      "python main.py --model CAPLA --project CAPLA_rmse_mse_mean_ap-False_sh-False_epoch50 --n_epoch 50 --loss mse_mean --apex --scheduler --aim rmse\n",
      "python main.py --model CAPLA --project CAPLA_rmse_mse_sum_ap-True_sh-True_epoch20 --n_epoch 20 --loss mse_sum --apex --scheduler --aim rmse\n",
      "python main.py --model CAPLA --project CAPLA_rmse_mse_sum_ap-True_sh-True_epoch50 --n_epoch 50 --loss mse_sum --apex --scheduler --aim rmse\n",
      "python main.py --model CAPLA --project CAPLA_rmse_mse_sum_ap-True_sh-False_epoch20 --n_epoch 20 --loss mse_sum --apex --scheduler --aim rmse\n",
      "python main.py --model CAPLA --project CAPLA_rmse_mse_sum_ap-True_sh-False_epoch50 --n_epoch 50 --loss mse_sum --apex --scheduler --aim rmse\n",
      "python main.py --model CAPLA --project CAPLA_rmse_mse_sum_ap-False_sh-True_epoch20 --n_epoch 20 --loss mse_sum --apex --scheduler --aim rmse\n",
      "python main.py --model CAPLA --project CAPLA_rmse_mse_sum_ap-False_sh-True_epoch50 --n_epoch 50 --loss mse_sum --apex --scheduler --aim rmse\n",
      "python main.py --model CAPLA --project CAPLA_rmse_mse_sum_ap-False_sh-False_epoch20 --n_epoch 20 --loss mse_sum --apex --scheduler --aim rmse\n",
      "python main.py --model CAPLA --project CAPLA_rmse_mse_sum_ap-False_sh-False_epoch50 --n_epoch 50 --loss mse_sum --apex --scheduler --aim rmse\n",
      "python main.py --model DeepDTA --project DeepDTA_loss_mse_mean_ap-True_sh-True_epoch20 --n_epoch 20 --loss mse_mean --apex --scheduler --aim loss\n",
      "python main.py --model DeepDTA --project DeepDTA_loss_mse_mean_ap-True_sh-True_epoch50 --n_epoch 50 --loss mse_mean --apex --scheduler --aim loss\n",
      "python main.py --model DeepDTA --project DeepDTA_loss_mse_mean_ap-True_sh-False_epoch20 --n_epoch 20 --loss mse_mean --apex --scheduler --aim loss\n",
      "python main.py --model DeepDTA --project DeepDTA_loss_mse_mean_ap-True_sh-False_epoch50 --n_epoch 50 --loss mse_mean --apex --scheduler --aim loss\n",
      "python main.py --model DeepDTA --project DeepDTA_loss_mse_mean_ap-False_sh-True_epoch20 --n_epoch 20 --loss mse_mean --apex --scheduler --aim loss\n",
      "python main.py --model DeepDTA --project DeepDTA_loss_mse_mean_ap-False_sh-True_epoch50 --n_epoch 50 --loss mse_mean --apex --scheduler --aim loss\n",
      "python main.py --model DeepDTA --project DeepDTA_loss_mse_mean_ap-False_sh-False_epoch20 --n_epoch 20 --loss mse_mean --apex --scheduler --aim loss\n",
      "python main.py --model DeepDTA --project DeepDTA_loss_mse_mean_ap-False_sh-False_epoch50 --n_epoch 50 --loss mse_mean --apex --scheduler --aim loss\n",
      "python main.py --model DeepDTA --project DeepDTA_loss_mse_sum_ap-True_sh-True_epoch20 --n_epoch 20 --loss mse_sum --apex --scheduler --aim loss\n",
      "python main.py --model DeepDTA --project DeepDTA_loss_mse_sum_ap-True_sh-True_epoch50 --n_epoch 50 --loss mse_sum --apex --scheduler --aim loss\n",
      "python main.py --model DeepDTA --project DeepDTA_loss_mse_sum_ap-True_sh-False_epoch20 --n_epoch 20 --loss mse_sum --apex --scheduler --aim loss\n",
      "python main.py --model DeepDTA --project DeepDTA_loss_mse_sum_ap-True_sh-False_epoch50 --n_epoch 50 --loss mse_sum --apex --scheduler --aim loss\n",
      "python main.py --model DeepDTA --project DeepDTA_loss_mse_sum_ap-False_sh-True_epoch20 --n_epoch 20 --loss mse_sum --apex --scheduler --aim loss\n",
      "python main.py --model DeepDTA --project DeepDTA_loss_mse_sum_ap-False_sh-True_epoch50 --n_epoch 50 --loss mse_sum --apex --scheduler --aim loss\n",
      "python main.py --model DeepDTA --project DeepDTA_loss_mse_sum_ap-False_sh-False_epoch20 --n_epoch 20 --loss mse_sum --apex --scheduler --aim loss\n",
      "python main.py --model DeepDTA --project DeepDTA_loss_mse_sum_ap-False_sh-False_epoch50 --n_epoch 50 --loss mse_sum --apex --scheduler --aim loss\n",
      "python main.py --model DeepDTAF --project DeepDTAF_loss_mse_mean_ap-True_sh-True_epoch20 --n_epoch 20 --loss mse_mean --apex --scheduler --aim loss\n",
      "python main.py --model DeepDTAF --project DeepDTAF_loss_mse_mean_ap-True_sh-True_epoch50 --n_epoch 50 --loss mse_mean --apex --scheduler --aim loss\n",
      "python main.py --model DeepDTAF --project DeepDTAF_loss_mse_mean_ap-True_sh-False_epoch20 --n_epoch 20 --loss mse_mean --apex --scheduler --aim loss\n",
      "python main.py --model DeepDTAF --project DeepDTAF_loss_mse_mean_ap-True_sh-False_epoch50 --n_epoch 50 --loss mse_mean --apex --scheduler --aim loss\n",
      "python main.py --model DeepDTAF --project DeepDTAF_loss_mse_mean_ap-False_sh-True_epoch20 --n_epoch 20 --loss mse_mean --apex --scheduler --aim loss\n",
      "python main.py --model DeepDTAF --project DeepDTAF_loss_mse_mean_ap-False_sh-True_epoch50 --n_epoch 50 --loss mse_mean --apex --scheduler --aim loss\n",
      "python main.py --model DeepDTAF --project DeepDTAF_loss_mse_mean_ap-False_sh-False_epoch20 --n_epoch 20 --loss mse_mean --apex --scheduler --aim loss\n",
      "python main.py --model DeepDTAF --project DeepDTAF_loss_mse_mean_ap-False_sh-False_epoch50 --n_epoch 50 --loss mse_mean --apex --scheduler --aim loss\n",
      "python main.py --model DeepDTAF --project DeepDTAF_loss_mse_sum_ap-True_sh-True_epoch20 --n_epoch 20 --loss mse_sum --apex --scheduler --aim loss\n",
      "python main.py --model DeepDTAF --project DeepDTAF_loss_mse_sum_ap-True_sh-True_epoch50 --n_epoch 50 --loss mse_sum --apex --scheduler --aim loss\n",
      "python main.py --model DeepDTAF --project DeepDTAF_loss_mse_sum_ap-True_sh-False_epoch20 --n_epoch 20 --loss mse_sum --apex --scheduler --aim loss\n",
      "python main.py --model DeepDTAF --project DeepDTAF_loss_mse_sum_ap-True_sh-False_epoch50 --n_epoch 50 --loss mse_sum --apex --scheduler --aim loss\n",
      "python main.py --model DeepDTAF --project DeepDTAF_loss_mse_sum_ap-False_sh-True_epoch20 --n_epoch 20 --loss mse_sum --apex --scheduler --aim loss\n",
      "python main.py --model DeepDTAF --project DeepDTAF_loss_mse_sum_ap-False_sh-True_epoch50 --n_epoch 50 --loss mse_sum --apex --scheduler --aim loss\n",
      "python main.py --model DeepDTAF --project DeepDTAF_loss_mse_sum_ap-False_sh-False_epoch20 --n_epoch 20 --loss mse_sum --apex --scheduler --aim loss\n",
      "python main.py --model DeepDTAF --project DeepDTAF_loss_mse_sum_ap-False_sh-False_epoch50 --n_epoch 50 --loss mse_sum --apex --scheduler --aim loss\n",
      "python main.py --model CAPLA --project CAPLA_loss_mse_mean_ap-True_sh-True_epoch20 --n_epoch 20 --loss mse_mean --apex --scheduler --aim loss\n",
      "python main.py --model CAPLA --project CAPLA_loss_mse_mean_ap-True_sh-True_epoch50 --n_epoch 50 --loss mse_mean --apex --scheduler --aim loss\n",
      "python main.py --model CAPLA --project CAPLA_loss_mse_mean_ap-True_sh-False_epoch20 --n_epoch 20 --loss mse_mean --apex --scheduler --aim loss\n",
      "python main.py --model CAPLA --project CAPLA_loss_mse_mean_ap-True_sh-False_epoch50 --n_epoch 50 --loss mse_mean --apex --scheduler --aim loss\n",
      "python main.py --model CAPLA --project CAPLA_loss_mse_mean_ap-False_sh-True_epoch20 --n_epoch 20 --loss mse_mean --apex --scheduler --aim loss\n",
      "python main.py --model CAPLA --project CAPLA_loss_mse_mean_ap-False_sh-True_epoch50 --n_epoch 50 --loss mse_mean --apex --scheduler --aim loss\n",
      "python main.py --model CAPLA --project CAPLA_loss_mse_mean_ap-False_sh-False_epoch20 --n_epoch 20 --loss mse_mean --apex --scheduler --aim loss\n",
      "python main.py --model CAPLA --project CAPLA_loss_mse_mean_ap-False_sh-False_epoch50 --n_epoch 50 --loss mse_mean --apex --scheduler --aim loss\n",
      "python main.py --model CAPLA --project CAPLA_loss_mse_sum_ap-True_sh-True_epoch20 --n_epoch 20 --loss mse_sum --apex --scheduler --aim loss\n",
      "python main.py --model CAPLA --project CAPLA_loss_mse_sum_ap-True_sh-True_epoch50 --n_epoch 50 --loss mse_sum --apex --scheduler --aim loss\n",
      "python main.py --model CAPLA --project CAPLA_loss_mse_sum_ap-True_sh-False_epoch20 --n_epoch 20 --loss mse_sum --apex --scheduler --aim loss\n",
      "python main.py --model CAPLA --project CAPLA_loss_mse_sum_ap-True_sh-False_epoch50 --n_epoch 50 --loss mse_sum --apex --scheduler --aim loss\n",
      "python main.py --model CAPLA --project CAPLA_loss_mse_sum_ap-False_sh-True_epoch20 --n_epoch 20 --loss mse_sum --apex --scheduler --aim loss\n",
      "python main.py --model CAPLA --project CAPLA_loss_mse_sum_ap-False_sh-True_epoch50 --n_epoch 50 --loss mse_sum --apex --scheduler --aim loss\n",
      "python main.py --model CAPLA --project CAPLA_loss_mse_sum_ap-False_sh-False_epoch20 --n_epoch 20 --loss mse_sum --apex --scheduler --aim loss\n",
      "python main.py --model CAPLA --project CAPLA_loss_mse_sum_ap-False_sh-False_epoch50 --n_epoch 50 --loss mse_sum --apex --scheduler --aim loss\n"
     ]
    }
   ],
   "source": [
    "aims = ['rmse', 'loss']\n",
    "models = ['DeepDTA', 'DeepDTAF', 'CAPLA']\n",
    "loss = ['mse_mean', 'mse_sum']\n",
    "apex = [True, False]\n",
    "scheduler = [True, False]\n",
    "epoch = [20, 50]\n",
    "\n",
    "for aim in aims:\n",
    "    for model in models:\n",
    "        for l in loss:\n",
    "            for a in apex:\n",
    "                for s in scheduler:\n",
    "                    for e in epoch:\n",
    "                        if apex:\n",
    "                            if scheduler:\n",
    "                                command = f\"python main.py --model {model} --project {model}_{aim}_{l}_ap-{a}_sh-{s}_epoch{e} --n_epoch {e} --loss {l} --apex --scheduler --aim {aim}\"\n",
    "                            else:\n",
    "                                command = f\"python main.py --model {model} --project {model}_{aim}_{l}_ap-{a}_sh-{s}_epoch{e} --n_epoch {e} --loss {l} --apex --aim {aim}\"\n",
    "                        else:\n",
    "                            if scheduler:\n",
    "                                command = f\"python main.py --model {model} --project {model}_{aim}_{l}_ap-{a}_sh-{s}_epoch{e} --n_epoch {e} --loss {l} --scheduler --aim {aim}\"\n",
    "                            else:\n",
    "                                command = f\"python main.py --model {model} --project {model}_{aim}_{l}_ap-{a}_sh-{s}_epoch{e} --n_epoch {e} --loss {l} --aim {aim}\"\n",
    "                        print(command)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 285,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "python main.py --model DeepDTA --n_epoch 20 --loss mse_mean --aim rmse --project DeepDTA_rmse_mean_epoch20\n",
      "python main.py --model DeepDTA --n_epoch 40 --loss mse_mean --aim rmse --project DeepDTA_rmse_mean_epoch40\n",
      "python main.py --model DeepDTA --n_epoch 60 --loss mse_mean --aim rmse --project DeepDTA_rmse_mean_epoch60\n",
      "python main.py --model DeepDTA --n_epoch 80 --loss mse_mean --aim rmse --project DeepDTA_rmse_mean_epoch80\n",
      "python main.py --model DeepDTA --n_epoch 100 --loss mse_mean --aim rmse --project DeepDTA_rmse_mean_epoch100\n",
      "python main.py --model DeepDTA --n_epoch 20 --loss mse_sum --aim rmse --project DeepDTA_rmse_sum_epoch20\n",
      "python main.py --model DeepDTA --n_epoch 40 --loss mse_sum --aim rmse --project DeepDTA_rmse_sum_epoch40\n",
      "python main.py --model DeepDTA --n_epoch 60 --loss mse_sum --aim rmse --project DeepDTA_rmse_sum_epoch60\n",
      "python main.py --model DeepDTA --n_epoch 80 --loss mse_sum --aim rmse --project DeepDTA_rmse_sum_epoch80\n",
      "python main.py --model DeepDTA --n_epoch 100 --loss mse_sum --aim rmse --project DeepDTA_rmse_sum_epoch100\n",
      "python main.py --model DeepDTA --n_epoch 20 --loss mse_mean --aim mse --project DeepDTA_mse_mean_epoch20\n",
      "python main.py --model DeepDTA --n_epoch 40 --loss mse_mean --aim mse --project DeepDTA_mse_mean_epoch40\n",
      "python main.py --model DeepDTA --n_epoch 60 --loss mse_mean --aim mse --project DeepDTA_mse_mean_epoch60\n",
      "python main.py --model DeepDTA --n_epoch 80 --loss mse_mean --aim mse --project DeepDTA_mse_mean_epoch80\n",
      "python main.py --model DeepDTA --n_epoch 100 --loss mse_mean --aim mse --project DeepDTA_mse_mean_epoch100\n",
      "python main.py --model DeepDTA --n_epoch 20 --loss mse_sum --aim mse --project DeepDTA_mse_sum_epoch20\n",
      "python main.py --model DeepDTA --n_epoch 40 --loss mse_sum --aim mse --project DeepDTA_mse_sum_epoch40\n",
      "python main.py --model DeepDTA --n_epoch 60 --loss mse_sum --aim mse --project DeepDTA_mse_sum_epoch60\n",
      "python main.py --model DeepDTA --n_epoch 80 --loss mse_sum --aim mse --project DeepDTA_mse_sum_epoch80\n",
      "python main.py --model DeepDTA --n_epoch 100 --loss mse_sum --aim mse --project DeepDTA_mse_sum_epoch100\n",
      "python main.py --model DeepDTAF --n_epoch 20 --loss mse_mean --aim rmse --project DeepDTAF_rmse_mean_epoch20\n",
      "python main.py --model DeepDTAF --n_epoch 40 --loss mse_mean --aim rmse --project DeepDTAF_rmse_mean_epoch40\n",
      "python main.py --model DeepDTAF --n_epoch 60 --loss mse_mean --aim rmse --project DeepDTAF_rmse_mean_epoch60\n",
      "python main.py --model DeepDTAF --n_epoch 80 --loss mse_mean --aim rmse --project DeepDTAF_rmse_mean_epoch80\n",
      "python main.py --model DeepDTAF --n_epoch 100 --loss mse_mean --aim rmse --project DeepDTAF_rmse_mean_epoch100\n",
      "python main.py --model DeepDTAF --n_epoch 20 --loss mse_sum --aim rmse --project DeepDTAF_rmse_sum_epoch20\n",
      "python main.py --model DeepDTAF --n_epoch 40 --loss mse_sum --aim rmse --project DeepDTAF_rmse_sum_epoch40\n",
      "python main.py --model DeepDTAF --n_epoch 60 --loss mse_sum --aim rmse --project DeepDTAF_rmse_sum_epoch60\n",
      "python main.py --model DeepDTAF --n_epoch 80 --loss mse_sum --aim rmse --project DeepDTAF_rmse_sum_epoch80\n",
      "python main.py --model DeepDTAF --n_epoch 100 --loss mse_sum --aim rmse --project DeepDTAF_rmse_sum_epoch100\n",
      "python main.py --model DeepDTAF --n_epoch 20 --loss mse_mean --aim mse --project DeepDTAF_mse_mean_epoch20\n",
      "python main.py --model DeepDTAF --n_epoch 40 --loss mse_mean --aim mse --project DeepDTAF_mse_mean_epoch40\n",
      "python main.py --model DeepDTAF --n_epoch 60 --loss mse_mean --aim mse --project DeepDTAF_mse_mean_epoch60\n",
      "python main.py --model DeepDTAF --n_epoch 80 --loss mse_mean --aim mse --project DeepDTAF_mse_mean_epoch80\n",
      "python main.py --model DeepDTAF --n_epoch 100 --loss mse_mean --aim mse --project DeepDTAF_mse_mean_epoch100\n",
      "python main.py --model DeepDTAF --n_epoch 20 --loss mse_sum --aim mse --project DeepDTAF_mse_sum_epoch20\n",
      "python main.py --model DeepDTAF --n_epoch 40 --loss mse_sum --aim mse --project DeepDTAF_mse_sum_epoch40\n",
      "python main.py --model DeepDTAF --n_epoch 60 --loss mse_sum --aim mse --project DeepDTAF_mse_sum_epoch60\n",
      "python main.py --model DeepDTAF --n_epoch 80 --loss mse_sum --aim mse --project DeepDTAF_mse_sum_epoch80\n",
      "python main.py --model DeepDTAF --n_epoch 100 --loss mse_sum --aim mse --project DeepDTAF_mse_sum_epoch100\n",
      "python main.py --model CAPLA --n_epoch 20 --loss mse_mean --aim rmse --project CAPLA_rmse_mean_epoch20\n",
      "python main.py --model CAPLA --n_epoch 40 --loss mse_mean --aim rmse --project CAPLA_rmse_mean_epoch40\n",
      "python main.py --model CAPLA --n_epoch 60 --loss mse_mean --aim rmse --project CAPLA_rmse_mean_epoch60\n",
      "python main.py --model CAPLA --n_epoch 80 --loss mse_mean --aim rmse --project CAPLA_rmse_mean_epoch80\n",
      "python main.py --model CAPLA --n_epoch 100 --loss mse_mean --aim rmse --project CAPLA_rmse_mean_epoch100\n",
      "python main.py --model CAPLA --n_epoch 20 --loss mse_sum --aim rmse --project CAPLA_rmse_sum_epoch20\n",
      "python main.py --model CAPLA --n_epoch 40 --loss mse_sum --aim rmse --project CAPLA_rmse_sum_epoch40\n",
      "python main.py --model CAPLA --n_epoch 60 --loss mse_sum --aim rmse --project CAPLA_rmse_sum_epoch60\n",
      "python main.py --model CAPLA --n_epoch 80 --loss mse_sum --aim rmse --project CAPLA_rmse_sum_epoch80\n",
      "python main.py --model CAPLA --n_epoch 100 --loss mse_sum --aim rmse --project CAPLA_rmse_sum_epoch100\n",
      "python main.py --model CAPLA --n_epoch 20 --loss mse_mean --aim mse --project CAPLA_mse_mean_epoch20\n",
      "python main.py --model CAPLA --n_epoch 40 --loss mse_mean --aim mse --project CAPLA_mse_mean_epoch40\n",
      "python main.py --model CAPLA --n_epoch 60 --loss mse_mean --aim mse --project CAPLA_mse_mean_epoch60\n",
      "python main.py --model CAPLA --n_epoch 80 --loss mse_mean --aim mse --project CAPLA_mse_mean_epoch80\n",
      "python main.py --model CAPLA --n_epoch 100 --loss mse_mean --aim mse --project CAPLA_mse_mean_epoch100\n",
      "python main.py --model CAPLA --n_epoch 20 --loss mse_sum --aim mse --project CAPLA_mse_sum_epoch20\n",
      "python main.py --model CAPLA --n_epoch 40 --loss mse_sum --aim mse --project CAPLA_mse_sum_epoch40\n",
      "python main.py --model CAPLA --n_epoch 60 --loss mse_sum --aim mse --project CAPLA_mse_sum_epoch60\n",
      "python main.py --model CAPLA --n_epoch 80 --loss mse_sum --aim mse --project CAPLA_mse_sum_epoch80\n",
      "python main.py --model CAPLA --n_epoch 100 --loss mse_sum --aim mse --project CAPLA_mse_sum_epoch100\n"
     ]
    }
   ],
   "source": [
    "models = ['DeepDTA', 'DeepDTAF', 'CAPLA']\n",
    "aims = ['rmse', 'mse']\n",
    "reduction = ['mse_mean', 'mse_sum']\n",
    "epoch = [20, 40, 60, 80, 100]\n",
    "\n",
    "for m in models:\n",
    "    for a in aims:\n",
    "        for r in reduction:\n",
    "            for e in epoch:\n",
    "                command = f\"python main.py --model {m} --n_epoch {e} --loss {r} --aim {a} --project {m}_{a}_{r.split('_')[1]}_epoch{e}\"\n",
    "                print(command)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 284,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>model</th>\n",
       "      <th>aim</th>\n",
       "      <th>mse_reduction</th>\n",
       "      <th>apex</th>\n",
       "      <th>scheduler</th>\n",
       "      <th>epoch</th>\n",
       "      <th>CORE_rmse</th>\n",
       "      <th>CORE_mae</th>\n",
       "      <th>CORE_pcc</th>\n",
       "      <th>CORE_sd</th>\n",
       "      <th>CORE_ci</th>\n",
       "      <th>CSAR_rmse</th>\n",
       "      <th>CSAR_mae</th>\n",
       "      <th>CSAR_pcc</th>\n",
       "      <th>CSAR_sd</th>\n",
       "      <th>CSAR_ci</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CAPLA</td>\n",
       "      <td>rmse</td>\n",
       "      <td>mean</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>20</td>\n",
       "      <td>1.510282</td>\n",
       "      <td>1.226740</td>\n",
       "      <td>0.708645</td>\n",
       "      <td>1.487428</td>\n",
       "      <td>0.756714</td>\n",
       "      <td>2.222766</td>\n",
       "      <td>1.721082</td>\n",
       "      <td>0.571250</td>\n",
       "      <td>2.165147</td>\n",
       "      <td>0.704666</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CAPLA</td>\n",
       "      <td>rmse</td>\n",
       "      <td>mean</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>50</td>\n",
       "      <td>1.527745</td>\n",
       "      <td>1.223725</td>\n",
       "      <td>0.700007</td>\n",
       "      <td>1.505596</td>\n",
       "      <td>0.752728</td>\n",
       "      <td>2.144618</td>\n",
       "      <td>1.693175</td>\n",
       "      <td>0.602664</td>\n",
       "      <td>2.107476</td>\n",
       "      <td>0.710923</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CAPLA</td>\n",
       "      <td>rmse</td>\n",
       "      <td>mean</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>20</td>\n",
       "      <td>1.510282</td>\n",
       "      <td>1.226740</td>\n",
       "      <td>0.708645</td>\n",
       "      <td>1.487428</td>\n",
       "      <td>0.756714</td>\n",
       "      <td>2.222766</td>\n",
       "      <td>1.721082</td>\n",
       "      <td>0.571250</td>\n",
       "      <td>2.165147</td>\n",
       "      <td>0.704666</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CAPLA</td>\n",
       "      <td>rmse</td>\n",
       "      <td>mean</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>50</td>\n",
       "      <td>1.527745</td>\n",
       "      <td>1.223725</td>\n",
       "      <td>0.700007</td>\n",
       "      <td>1.505596</td>\n",
       "      <td>0.752728</td>\n",
       "      <td>2.144618</td>\n",
       "      <td>1.693175</td>\n",
       "      <td>0.602664</td>\n",
       "      <td>2.107476</td>\n",
       "      <td>0.710923</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>DeepDTAF</td>\n",
       "      <td>loss</td>\n",
       "      <td>mean</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>20</td>\n",
       "      <td>1.537258</td>\n",
       "      <td>1.238472</td>\n",
       "      <td>0.700638</td>\n",
       "      <td>1.503907</td>\n",
       "      <td>0.750762</td>\n",
       "      <td>2.212589</td>\n",
       "      <td>1.779897</td>\n",
       "      <td>0.537089</td>\n",
       "      <td>2.224508</td>\n",
       "      <td>0.702969</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>DeepDTA</td>\n",
       "      <td>rmse</td>\n",
       "      <td>sum</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>50</td>\n",
       "      <td>1.483163</td>\n",
       "      <td>1.210515</td>\n",
       "      <td>0.735650</td>\n",
       "      <td>1.426866</td>\n",
       "      <td>0.767094</td>\n",
       "      <td>1.831741</td>\n",
       "      <td>1.433866</td>\n",
       "      <td>0.747953</td>\n",
       "      <td>1.752715</td>\n",
       "      <td>0.786373</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>DeepDTA</td>\n",
       "      <td>rmse</td>\n",
       "      <td>sum</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>20</td>\n",
       "      <td>1.441149</td>\n",
       "      <td>1.167233</td>\n",
       "      <td>0.739143</td>\n",
       "      <td>1.413604</td>\n",
       "      <td>0.771426</td>\n",
       "      <td>1.741605</td>\n",
       "      <td>1.343187</td>\n",
       "      <td>0.759746</td>\n",
       "      <td>1.715038</td>\n",
       "      <td>0.789873</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>DeepDTA</td>\n",
       "      <td>rmse</td>\n",
       "      <td>sum</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>50</td>\n",
       "      <td>1.483163</td>\n",
       "      <td>1.210515</td>\n",
       "      <td>0.735650</td>\n",
       "      <td>1.426866</td>\n",
       "      <td>0.767094</td>\n",
       "      <td>1.831741</td>\n",
       "      <td>1.433866</td>\n",
       "      <td>0.747953</td>\n",
       "      <td>1.752715</td>\n",
       "      <td>0.786373</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>DeepDTA</td>\n",
       "      <td>rmse</td>\n",
       "      <td>sum</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>20</td>\n",
       "      <td>1.441149</td>\n",
       "      <td>1.167233</td>\n",
       "      <td>0.739143</td>\n",
       "      <td>1.413604</td>\n",
       "      <td>0.771426</td>\n",
       "      <td>1.741605</td>\n",
       "      <td>1.343187</td>\n",
       "      <td>0.759746</td>\n",
       "      <td>1.715038</td>\n",
       "      <td>0.789873</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67</th>\n",
       "      <td>DeepDTA</td>\n",
       "      <td>rmse</td>\n",
       "      <td>sum</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>50</td>\n",
       "      <td>1.483163</td>\n",
       "      <td>1.210515</td>\n",
       "      <td>0.735650</td>\n",
       "      <td>1.426866</td>\n",
       "      <td>0.767094</td>\n",
       "      <td>1.831741</td>\n",
       "      <td>1.433866</td>\n",
       "      <td>0.747953</td>\n",
       "      <td>1.752715</td>\n",
       "      <td>0.786373</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>68 rows × 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       model   aim mse_reduction   apex scheduler epoch  CORE_rmse  CORE_mae  \\\n",
       "0      CAPLA  rmse          mean   True     False    20   1.510282  1.226740   \n",
       "1      CAPLA  rmse          mean   True     False    50   1.527745  1.223725   \n",
       "2      CAPLA  rmse          mean   True      True    20   1.510282  1.226740   \n",
       "3      CAPLA  rmse          mean   True      True    50   1.527745  1.223725   \n",
       "4   DeepDTAF  loss          mean  False     False    20   1.537258  1.238472   \n",
       "..       ...   ...           ...    ...       ...   ...        ...       ...   \n",
       "63   DeepDTA  rmse           sum  False      True    50   1.483163  1.210515   \n",
       "64   DeepDTA  rmse           sum   True     False    20   1.441149  1.167233   \n",
       "65   DeepDTA  rmse           sum   True     False    50   1.483163  1.210515   \n",
       "66   DeepDTA  rmse           sum   True      True    20   1.441149  1.167233   \n",
       "67   DeepDTA  rmse           sum   True      True    50   1.483163  1.210515   \n",
       "\n",
       "    CORE_pcc   CORE_sd   CORE_ci  CSAR_rmse  CSAR_mae  CSAR_pcc   CSAR_sd  \\\n",
       "0   0.708645  1.487428  0.756714   2.222766  1.721082  0.571250  2.165147   \n",
       "1   0.700007  1.505596  0.752728   2.144618  1.693175  0.602664  2.107476   \n",
       "2   0.708645  1.487428  0.756714   2.222766  1.721082  0.571250  2.165147   \n",
       "3   0.700007  1.505596  0.752728   2.144618  1.693175  0.602664  2.107476   \n",
       "4   0.700638  1.503907  0.750762   2.212589  1.779897  0.537089  2.224508   \n",
       "..       ...       ...       ...        ...       ...       ...       ...   \n",
       "63  0.735650  1.426866  0.767094   1.831741  1.433866  0.747953  1.752715   \n",
       "64  0.739143  1.413604  0.771426   1.741605  1.343187  0.759746  1.715038   \n",
       "65  0.735650  1.426866  0.767094   1.831741  1.433866  0.747953  1.752715   \n",
       "66  0.739143  1.413604  0.771426   1.741605  1.343187  0.759746  1.715038   \n",
       "67  0.735650  1.426866  0.767094   1.831741  1.433866  0.747953  1.752715   \n",
       "\n",
       "     CSAR_ci  \n",
       "0   0.704666  \n",
       "1   0.710923  \n",
       "2   0.704666  \n",
       "3   0.710923  \n",
       "4   0.702969  \n",
       "..       ...  \n",
       "63  0.786373  \n",
       "64  0.789873  \n",
       "65  0.786373  \n",
       "66  0.789873  \n",
       "67  0.786373  \n",
       "\n",
       "[68 rows x 16 columns]"
      ]
     },
     "execution_count": 284,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import json\n",
    "summaries = []\n",
    "results = [fd for fd in Path('logs').glob('*') if fd.stem != 'DeepDTAF_log']\n",
    "for r in results:\n",
    "    name = r.stem \n",
    "    md = name.split('_')[0]\n",
    "    aim = name.split('_')[1]\n",
    "    mse_reduction = name.split('_')[3]\n",
    "    apex = name.split('_')[4].split('-')[1]\n",
    "    scheduler = name.split('_')[5].split('-')[1]\n",
    "    epoch = name.split('_')[6][5:]\n",
    "\n",
    "    summary = {'model': md, \n",
    "               'aim': aim, \n",
    "               'mse_reduction': mse_reduction, \n",
    "               'apex': apex, 'scheduler': scheduler, \n",
    "               'epoch': epoch}\n",
    "    \n",
    "    file = r / 'average_test_metrics.json'\n",
    "    if not file.exists():\n",
    "        continue\n",
    "\n",
    "    with open(file, 'r') as f:\n",
    "        tst_result = json.load(f)\n",
    "    for dt in ['CORE', 'CSAR']:\n",
    "        rmse = tst_result[dt]['rmse']\n",
    "        mae = tst_result[dt]['mae']\n",
    "        pcc = tst_result[dt]['pcc']\n",
    "        sd = tst_result[dt]['sd']\n",
    "        ci = tst_result[dt]['ci']\n",
    "        \n",
    "        summary[f'{dt}_rmse'] = rmse\n",
    "        summary[f'{dt}_mae'] = mae\n",
    "        summary[f'{dt}_pcc'] = pcc\n",
    "        summary[f'{dt}_sd'] = sd\n",
    "        summary[f'{dt}_ci'] = ci\n",
    "    \n",
    "    summaries.append(summary)\n",
    "    \n",
    "summaries = pd.DataFrame(summaries)\n",
    "summaries.to_csv('summaries.csv', index=False)\n",
    "summaries\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 281,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>PDB</th>\n",
       "      <th>Ligand</th>\n",
       "      <th>Ligand_Len</th>\n",
       "      <th>Pocket</th>\n",
       "      <th>Pocket_Len</th>\n",
       "      <th>Global</th>\n",
       "      <th>Global_Len</th>\n",
       "      <th>Set</th>\n",
       "      <th>Affinity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>13322</th>\n",
       "      <td>13322</td>\n",
       "      <td>5lz4</td>\n",
       "      <td>CC1(C)CN(CCOc2ccc(c(C#N)c2)F)C(=O)C1</td>\n",
       "      <td>36</td>\n",
       "      <td>WIYLFLSHGLGAFTLYSSSRRHSFGGATLDAWMFPFYNHQFADFTFAL</td>\n",
       "      <td>48</td>\n",
       "      <td>KIPRGNGPYSVGCTDLMFDHTNKGTFLRLYYPSQDNDRLDTLWIPN...</td>\n",
       "      <td>371</td>\n",
       "      <td>TST71</td>\n",
       "      <td>5.7</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       Unnamed: 0   PDB                                Ligand  Ligand_Len  \\\n",
       "13322       13322  5lz4  CC1(C)CN(CCOc2ccc(c(C#N)c2)F)C(=O)C1          36   \n",
       "\n",
       "                                                 Pocket  Pocket_Len  \\\n",
       "13322  WIYLFLSHGLGAFTLYSSSRRHSFGGATLDAWMFPFYNHQFADFTFAL          48   \n",
       "\n",
       "                                                  Global  Global_Len    Set  \\\n",
       "13322  KIPRGNGPYSVGCTDLMFDHTNKGTFLRLYYPSQDNDRLDTLWIPN...         371  TST71   \n",
       "\n",
       "       Affinity  \n",
       "13322       5.7  "
      ]
     },
     "execution_count": 281,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "base_df = pd.read_csv('data/deepdtaf_data.csv')\n",
    "base_df[base_df['PDB'] == '5lz4'] # TST71은 데이터 제공 X"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
